<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-0.9.629">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Introduction to Machine Learning for Economists (Under Construction) - 13&nbsp; Forest-based CATE Estimators</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./C04-cf-extension.html" rel="next">
<link href="./C02-xstr-learner.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
<link href="style.css" rel="stylesheet" type="text/css">

<!-- Global site tag (gtag.js) - Google Analytics -->
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-45VKT2HZSL"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag() { dataLayer.push(arguments); }
  gtag('js', new Date());

  gtag('config', 'G-45VKT2HZSL');
</script>

  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="nav-sidebar floating slimcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Forest-based CATE Estimators</span></h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Introduction to Machine Learning for Economists (Under Construction)</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/tmieno2/ML-Economist" title="Source Code" class="sidebar-tool px-1"><i class="bi bi-github"></i></a>
    <a href="" title="Share" id="sidebar-tool-dropdown-0" class="sidebar-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi bi-share"></i></a>
    <ul class="dropdown-menu" aria-labelledby="sidebar-tool-dropdown-0">
        <li>
          <a class="dropdown-item sidebar-tools-main-item" href="https://twitter.com/intent/tweet?url=|url|">
            <i class="bi bi-bi-twitter pe-1"></i>
          Twitter
          </a>
        </li>
        <li>
          <a class="dropdown-item sidebar-tools-main-item" href="https://www.facebook.com/sharer/sharer.php?u=|url|">
            <i class="bi bi-bi-facebook pe-1"></i>
          Facebook
          </a>
        </li>
        <li>
          <a class="dropdown-item sidebar-tools-main-item" href="https://www.linkedin.com/sharing/share-offsite/?url=|url|">
            <i class="bi bi-bi-linkedin pe-1"></i>
          LinkedIn
          </a>
        </li>
    </ul>
</div>
    </div>
      </div>
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">Welcome</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./H00-preface.html" class="sidebar-item-text sidebar-link">Preface</a>
  </div>
</li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
    <div class="sidebar-item-container"> 
        <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">Basics</a>
      <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
        <i class="bi bi-chevron-right ms-2"></i>
      </a>
    </div>
    <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./B01-nonlinear.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Non-linear function estimation</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./B02-bias-variance-tradeoff.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Bias-variance Trade-off</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./B03-cross-validation.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Cross-validation</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./B04-regularization.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Regression Shrinkage Methods</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./B05-bootstrap.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Bootstrap</span></a>
  </div>
</li>
    </ul>
  </li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
    <div class="sidebar-item-container"> 
        <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">Prediction ML</a>
      <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
        <i class="bi bi-chevron-right ms-2"></i>
      </a>
    </div>
    <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./P01-random-forest.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Random Forest</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./P02-boosted-regression-forest.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Boosted Regression Forest</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./P03-xgb.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Extreme Gradient Boosting</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./P04-local-linear-forest.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Local Linear Forest</span></a>
  </div>
</li>
    </ul>
  </li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
    <div class="sidebar-item-container"> 
        <a href="./C0P-causal-ml.html" class="sidebar-item-text sidebar-link">Causal Machine Learning (CML) Methods</a>
      <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
        <i class="bi bi-chevron-right ms-2"></i>
      </a>
    </div>
    <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./C00-why-not-this.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Why can’t we just do this?</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./C01-dml.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Double Machine Learning</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./C02-xstr-learner.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">S-, X-, T-, and R-learner</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./C03-cf-orf.html" class="sidebar-item-text sidebar-link active"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Forest-based CATE Estimators</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./C04-cf-extension.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Extensions of Causal Forest</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./C05-causal-model-selection.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Causal Model Selection</span></a>
  </div>
</li>
    </ul>
  </li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
    <div class="sidebar-item-container"> 
        <a href="./E0P-extensions.html" class="sidebar-item-text sidebar-link">Extensions</a>
      <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
        <i class="bi bi-chevron-right ms-2"></i>
      </a>
    </div>
    <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./E01-spatial-cv.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Spatial Cross-validation</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./E02-grf.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Generalized Random Forest</span></a>
  </div>
</li>
    </ul>
  </li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
    <div class="sidebar-item-container"> 
        <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">Programming: R</a>
      <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">
        <i class="bi bi-chevron-right ms-2"></i>
      </a>
    </div>
    <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./PROG-R-01-mlr3.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Machine Learning with <code>mlr3</code></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./PROG-R-02-reticulate.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Running Python from R</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./PROG-R-03-model-selection-prediction.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Model Selection (Prediction)</span></a>
  </div>
</li>
    </ul>
  </li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
    <div class="sidebar-item-container"> 
        <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true">Programming: Python</a>
      <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true">
        <i class="bi bi-chevron-right ms-2"></i>
      </a>
    </div>
    <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./PROG-P-01-scikitlearn.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Prediction</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./PROG-P-02-CATE-econml.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">22</span>&nbsp; <span class="chapter-title">Treatment Effect Estimation</span></a>
  </div>
</li>
    </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
    <div class="sidebar-item-container"> 
        <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" aria-expanded="true">Appendices</a>
      <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" aria-expanded="true">
        <i class="bi bi-chevron-right ms-2"></i>
      </a>
    </div>
    <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth1 show">  
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./A01-mc-simulation.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Monte Carlo (MC) Simulation</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./A02-method-of-moment.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">Primer on method of moment</span></a>
  </div>
</li>
    </ul>
  </li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#model" id="toc-model" class="nav-link active" data-scroll-target="#model"> <span class="header-section-number">13.1</span> Model</a></li>
  <li><a href="#causal-forest" id="toc-causal-forest" class="nav-link" data-scroll-target="#causal-forest"> <span class="header-section-number">13.2</span> Causal Forest</a>
  <ul class="collapse">
  <li><a href="#brief-description-of-how-cf-works" id="toc-brief-description-of-how-cf-works" class="nav-link" data-scroll-target="#brief-description-of-how-cf-works"> <span class="header-section-number">13.2.1</span> Brief description of how CF works</a></li>
  <li><a href="#training-a-causal-forest" id="toc-training-a-causal-forest" class="nav-link" data-scroll-target="#training-a-causal-forest"> <span class="header-section-number">13.2.2</span> Training a causal forest</a></li>
  <li><a href="#predict-and-interpret-cate" id="toc-predict-and-interpret-cate" class="nav-link" data-scroll-target="#predict-and-interpret-cate"> <span class="header-section-number">13.2.3</span> Predict and interpret CATE</a></li>
  </ul></li>
  <li><a href="#orthogonal-random-forest" id="toc-orthogonal-random-forest" class="nav-link" data-scroll-target="#orthogonal-random-forest"> <span class="header-section-number">13.3</span> Orthogonal Random Forest</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/tmieno2/ML-Economist/edit/master/C03-cf-orf.qmd" class="toc-action">Edit this page</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title"><span id="sec-cf-orf" class="quarto-section-identifier d-none d-lg-block"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Forest-based CATE Estimators</span></span></h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>



<div class="quarto-title-meta">

    
    
  </div>
  

</header>

<p>In <a href="C02-xstr-learner.html"><span>Chapter&nbsp;12</span></a>, we saw some special cases of R-learner/DML where the final model is estimate by parametrically using a linear-in-parameter model and non-parametrically using extreme gradient boosting. Here, we learn two methods that estimate CATE non-parametrically: causal forest <span class="citation" data-cites="athey2019generalized">(<a href="#ref-athey2019generalized" role="doc-biblioref">Athey, Tibshirani, and Wager 2019</a>)</span> and orthogonal forest <span class="citation" data-cites="oprescu2019orthogonal">(<a href="#ref-oprescu2019orthogonal" role="doc-biblioref">Oprescu, Syrgkanis, and Wu 2019</a>)</span>.</p>
<div class="callout-important callout callout-style-default no-icon callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
What you will learn
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Causal forest
<ul>
<li>Mechanics</li>
<li>Hyper-parameter</li>
<li>Predict heterogeneous treatment effects (CATE)</li>
<li>Predict average treatment effects (ATE)</li>
<li>Interpretation</li>
</ul></li>
<li>Orthogonal forest
<ul>
<li>Differences from causal forest</li>
</ul></li>
</ul>
</div>
</div>
<div class="callout-tip callout callout-style-default no-icon callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Background knowledge
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>(preferable) DML/R-learner (<a href="C02-xstr-learner.html"><span>Chapter&nbsp;12</span></a>)</li>
<li>(preferable) GRF (<a href="E02-grf.html"><span>Chapter&nbsp;17</span></a>)</li>
</ul>
</div>
</div>
<div class="callout-note callout callout-style-default no-icon callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Packages to load for replication
</div>
</div>
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(grf)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<p><strong>Left to be added</strong> + how variance is estimated + balance + ATE + linear correction with locally weighted linear regression</p>
<section id="model" class="level2" data-number="13.1">
<h2 data-number="13.1" class="anchored" data-anchor-id="model"><span class="header-section-number">13.1</span> Model</h2>
<p>The heterogeneous treatment effect model of interest in this chapter is the same as the one in <a href="C02-xstr-learner.html"><span>Chapter&nbsp;12</span></a>.</p>
<p><span id="eq-model-het-te"><span class="math display">\[
\begin{aligned}
Y &amp; = \theta(X)\cdot T + g(X) + \varepsilon \\
T &amp; = f(X) + \eta
\end{aligned}
\tag{13.1}\]</span></span></p>
<ul>
<li><span class="math inline">\(Y\)</span>: dependent variable</li>
<li><span class="math inline">\(T\)</span>: treatment variable (can be either binary dummy or continuous)</li>
<li><span class="math inline">\(X\)</span>: features</li>
</ul>
<p>Causal forest and orthogonal random forest is consistent only if the following conditions fold.</p>
<ul>
<li><span class="math inline">\(E[\varepsilon|X] = 0\)</span></li>
<li><span class="math inline">\(E[\eta|X] = 0\)</span></li>
<li><span class="math inline">\(E[\eta\cdot\varepsilon|X] = 0\)</span></li>
</ul>
</section>
<section id="causal-forest" class="level2 page-columns page-full" data-number="13.2">
<h2 data-number="13.2" class="anchored" data-anchor-id="causal-forest"><span class="header-section-number">13.2</span> Causal Forest</h2>
<section id="brief-description-of-how-cf-works" class="level3 page-columns page-full" data-number="13.2.1">
<h3 data-number="13.2.1" class="anchored" data-anchor-id="brief-description-of-how-cf-works"><span class="header-section-number">13.2.1</span> Brief description of how CF works</h3>
<p>This section provides a brief and cursory description of how CF works. A much more detailed treatment of how CF (and also GRF) works along with the explanations of hyper-parameters is provided in <span class="quarto-unresolved-ref">?sec-understand-grf-example</span>.</p>
<p>Causal Forest (CF) (as implemented by the R <code>grf</code> package or python <code>econml</code> package) is a special type of R-learner (also a DML) and also a special case of generalized random forest (GRF).</p>

<div class="no-row-height column-margin column-container"><div class="">
<p>CF can be implemented using the R <code>grf</code> package or python <code>econml</code> package. Both of them implements CF as an R-learner. However, the original causal forest proposed in <span class="citation" data-cites="wager_estimation_2018">Wager and Athey (<a href="#ref-wager_estimation_2018" role="doc-biblioref">2018</a>)</span> does not follow an R-learner procedure.</p>
</div></div><p>Let <span class="math inline">\(\hat{f}(X_i)\)</span> and <span class="math inline">\(\hat{g}(X_i)\)</span> denote the estimation of <span class="math inline">\(E[Y|X]\)</span> and <span class="math inline">\(E[T|X]\)</span>, respectively. Further, let <span class="math inline">\(\hat{\tilde{Y_i}}\)</span> and <span class="math inline">\(\hat{\tilde{T_i}}\)</span> denote <span class="math inline">\(Y_i - \hat{f}(X_i)\)</span> and <span class="math inline">\(T_i - \hat{g}(X_i)\)</span>, respectively.</p>
<p>Then, CF estimates <span class="math inline">\(\theta(X)\)</span> at <span class="math inline">\(X = X_0\)</span> by solving the following equation:</p>
<p><span id="eq-solve-cf"><span class="math display">\[
\begin{aligned}
\hat{\theta}(X_0) = argmin_{\theta}\;\;\sum_{i=1}^N \alpha_i(X_i, X_0)[\hat{\tilde{Y_i}} - \theta\cdot \hat{\tilde{T_i}}]^2
\end{aligned}
\tag{13.2}\]</span></span></p>
<p>where <span class="math inline">\(\alpha_i(X_i, X_0)\)</span> is the weight given to each <span class="math inline">\(i\)</span>. The F.O.C is</p>
<p><span class="math display">\[
\begin{aligned}
2 \cdot \sum_{i=1}^N \alpha_i(X_i, X_0)[\hat{\tilde{Y_i}} - \theta\cdot \hat{\tilde{T_i}}]\hat{\tilde{T_i}} = 0
\end{aligned}
\]</span></p>
<p><span id="eq-theta-solution"><span class="math display">\[
\begin{aligned}
\hat{\theta}(X_0) = \frac{\sum_{i=1}^N\alpha_i(X_i, X_0)\cdot\hat{\tilde{Y}}_i\cdot \hat{\tilde{T}}_i}{\sum_{i=1}^N\alpha_i(X_i, X_0)\cdot\hat{\tilde{T}}_i\cdot \hat{\tilde{T}}_i}
\end{aligned}
\tag{13.3}\]</span></span></p>
<p>Unlike the DML approaches we saw in <a href="C02-xstr-learner.html"><span>Chapter&nbsp;12</span></a> that uses a linear model as the final model, CF does not assume any functional form for how <span class="math inline">\(X\)</span> affects <span class="math inline">\(\theta\)</span> as you can see from the above minimization problem. As mentioned earlier, CF is a special case of GRF (discussed in <a href="E02-grf.html"><span>Chapter&nbsp;17</span></a>), so CF is also a local constant regression (see <a href="B01-nonlinear.html#sec-local-reg"><span>Section&nbsp;1.3.1</span></a> for a brief discussion on local regression).</p>
<p><span class="math inline">\(\alpha_i(X_i, X_0)\)</span> is determined based on the trees trained based on the pseudo outcomes that are defined specifically for causal forest estimation. Suppose <span class="math inline">\(T\)</span> trees have been built and let <span class="math inline">\(\eta_{i,t}(X_i, X_0)\)</span> be 1 if observation <span class="math inline">\(i\)</span> belongs to the same leaf as <span class="math inline">\(X_0\)</span> in tree <span class="math inline">\(t\)</span>. Then,</p>
<p><span id="eq-weight-cf"><span class="math display">\[
\begin{aligned}
\alpha_i(X_i, X_0) = \frac{1}{T}\sum_{t=1}^T\frac{\eta_{i,t}(X_i, X_0)}{\sum_{i=1}^N\eta_{i,t}(X_i, X_0)}
\end{aligned}
\tag{13.4}\]</span></span></p>
<p>So, the weight given to observation <span class="math inline">\(i\)</span> is higher if observation <span class="math inline">\(i\)</span> belongs to the same leaf as the evaluation point <span class="math inline">\(X_0\)</span> in more trees.</p>
</section>
<section id="training-a-causal-forest" class="level3" data-number="13.2.2">
<h3 data-number="13.2.2" class="anchored" data-anchor-id="training-a-causal-forest"><span class="header-section-number">13.2.2</span> Training a causal forest</h3>
<p>We can use the <code>causal_forest()</code> function from the <code>grf</code> package to train a CF model in R. In Python, you can use <code>CausalForestDML()</code> from the <code>econml</code> package or <code>GRFForestRegressor</code> from the <code>skgrf</code> package.</p>
<p>As of now, there are some notable differences between <code>grf</code> and <code>econml</code>.</p>
<div class="callout-tip callout callout-style-default no-icon callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Some differences between the <code>grf</code> and <code>econml</code> implementation of causal forest
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li><p>While <code>grf</code> support clustering, <code>econml</code> does not.</p></li>
<li><p>It is easy to modify the first stage estimations in <code>econml</code>. <code>grf</code> uses random forest by default for them. If you would like to try other ML methods, then you need to write a code to predict <span class="math inline">\(Y - E[Y|X, W]\)</span> and <span class="math inline">\(T - E[T|X, W]\)</span> with your choice of ML methods yourself (see <a href="PROG-R-01-mlr3.html#sec-mlr3-in-action"><span>Section&nbsp;18.6</span></a> for an example code to do this).</p></li>
<li><p><code>grf</code> does not cross-fit in the first-stage estimation. Rather, it uses out-of-bag prediction from the trained random forest models, which avoids over-fitting (Note that the motivation behind cross-fitting in DML is to avoid over-fitting bias in the second stage estimation. See <a href="C01-dml.html"><span>Chapter&nbsp;11</span></a>).</p></li>
<li><p><code>grf</code> does not make distinctions between <span class="math inline">\(X\)</span> and <span class="math inline">\(W\)</span>.</p></li>
</ul>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co">#=== load the Treatment dataset ===#</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"Treatment"</span>, <span class="at">package =</span> <span class="st">"Ecdat"</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co">#=== convert to a data.table ===#</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> </span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.table</span>(Treatment) <span class="sc">%&gt;%</span> </span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>  <span class="co">#=== create an id variable ===#</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>  .[, <span class="at">id :=</span> <span class="dv">1</span><span class="sc">:</span>.N]</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here are the variables in this dataset that we use.</p>
<ul>
<li><code>re78</code> (<span class="math inline">\(Y\)</span>): real annual earnings in 1978 (after the treatment)</li>
<li><code>treat</code> (<span class="math inline">\(T\)</span>): <code>TRUE</code> if a person had gone through a training, <code>FALSE</code> otherwise.</li>
</ul>
<p><span class="math inline">\(X\)</span> includes</p>
<ul>
<li><code>re74</code>: real annual earnings in 1978 (after the treatment)</li>
<li><code>age</code>: age</li>
<li><code>educ</code>: education in years</li>
<li><code>ethn</code>: one of “other”, “black”, “hispanic”</li>
<li><code>married</code>: married or not</li>
</ul>
<p><code>grf::causal_forest()</code> takes only numeric values for <span class="math inline">\(X\)</span>. So, we will one-hot encode <code>ethn</code>, which is a factor variable at the moment.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>data_trt <span class="ot">&lt;-</span> mltools<span class="sc">::</span><span class="fu">one_hot</span>(data)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      treat age educ ethn_other ethn_black ethn_hispanic married    re74
   1:  TRUE  37   11          0          1             0    TRUE     0.0
   2:  TRUE  30   12          0          1             0   FALSE     0.0
   3:  TRUE  27   11          0          1             0   FALSE     0.0
   4:  TRUE  33    8          0          1             0   FALSE     0.0
   5:  TRUE  22    9          0          1             0   FALSE     0.0
  ---                                                                   
2671: FALSE  47    8          1          0             0    TRUE 44667.4
2672: FALSE  32    8          1          0             0    TRUE 47022.4
2673: FALSE  47   10          1          0             0    TRUE 48198.0
2674: FALSE  54    0          0          0             1    TRUE 49228.5
2675: FALSE  40    8          1          0             0    TRUE 50940.9
         re75     re78   u74   u75   id
   1:     0.0  9930.05  TRUE  TRUE    1
   2:     0.0 24909.50  TRUE  TRUE    2
   3:     0.0  7506.15  TRUE  TRUE    3
   4:     0.0   289.79  TRUE  TRUE    4
   5:     0.0  4056.49  TRUE  TRUE    5
  ---                                  
2671: 33837.1 38568.70 FALSE FALSE 2671
2672: 67137.1 59109.10 FALSE FALSE 2672
2673: 47968.1 55710.30 FALSE FALSE 2673
2674: 44221.0 20540.40 FALSE FALSE 2674
2675: 55500.0 53198.20 FALSE FALSE 2675</code></pre>
</div>
</div>
<p>We now have <code>ethn_black</code>, <code>ethn_hispanic</code>, and <code>ethn_other</code> from <code>ethn</code>. The model we are estimating is as follows:</p>
<p><span class="math display">\[
\begin{aligned}
re78 &amp; = \theta(age, re74, educ, ethn\_hipanic, ethn\_black, married)\cdot treat + g(age, re74, educ, ethn\_hipanic, ethn\_black, married) + \varepsilon \\
treat &amp; = f(age, re74, educ, ethn\_hipanic, ethn\_black, married) + \eta
\end{aligned}
\]</span></p>
<div class="callout-warning callout callout-style-default no-icon callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Warning
</div>
</div>
<div class="callout-body-container callout-body">
<p>Note that we are paying no attention to the potential endogeneity problem here. This is just a demonstration and the results are likely to be biased. We will look at instrumental forest later in <span class="quarto-unresolved-ref">?sec-cf-extension</span>, which is consistent under confoundedness as long as you can find appropriate external instruments.</p>
</div>
</div>
<p>In running <code>causal_forest()</code>, there are many hyper-parameters and options that we need to be aware of.</p>
<p>Since CF is a GRF and GRF uses random forest algorithm on appropriate pseudo outcome, it is natural that some of the CF hyper-parameters are the same as the ones for RF (by <code>ranger()</code>).</p>
<ul>
<li><code>num.trees</code>: number of trees</li>
<li><code>mtry</code>: number of variables tried in each split (default is <span class="math inline">\(\sqrt{K}\)</span>)</li>
<li><code>min.node.size</code>: minimum number of observations in each leaf (default is 5)</li>
</ul>
<p>A hyper-parameter that certainly affects tree building process is <code>sample.fraction</code>, which we saw earlier.</p>
<ul>
<li><code>sample.fraction</code>: fraction of the data used to build each tree (default is 0.5)</li>
</ul>
<p>A higher value of <code>sample.fraction</code> means that the trees are more correlated as they share more of the same observations.</p>
<p>There are three honesty-related options.</p>
<ul>
<li><code>honesty</code>: <code>TRUE</code> if using honest-sampling, 0 otherwise (default is <code>TRUE</code>)</li>
<li><code>honesty.fraction</code>: fraction of the data (after <code>sample.fraction</code> is applied) that is grouped into <span class="math inline">\(J_1\)</span>, which is used to determine splitting rules</li>
<li><code>honesty.prune.leaves</code>: <code>TRUE</code> if the leaves with no samples are pruned (default is <code>TRUE</code>)</li>
</ul>
<p>Let’s now train CF using <code>data_trt</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>cf_trained <span class="ot">&lt;-</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>  grf<span class="sc">::</span><span class="fu">causal_forest</span>(</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">X =</span> data_trt[, .(age, re74, educ, ethn_hispanic, ethn_black, married)] <span class="sc">%&gt;%</span> <span class="fu">as.matrix</span>(),</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">Y =</span> data_trt[, re78],</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">W =</span> data_trt[, treat]</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here is what <code>cf_trained</code> has as its attributes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(cf_trained)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "_ci_group_size"           "_num_variables"          
 [3] "_num_trees"               "_root_nodes"             
 [5] "_child_nodes"             "_leaf_samples"           
 [7] "_split_vars"              "_split_values"           
 [9] "_drawn_samples"           "_send_missing_left"      
[11] "_pv_values"               "_pv_num_types"           
[13] "predictions"              "variance.estimates"      
[15] "debiased.error"           "excess.error"            
[17] "seed"                     "ci.group.size"           
[19] "X.orig"                   "Y.orig"                  
[21] "W.orig"                   "Y.hat"                   
[23] "W.hat"                    "clusters"                
[25] "equalize.cluster.weights" "tunable.params"          
[27] "has.missing.values"      </code></pre>
</div>
</div>
<p>For example, you can get <span class="math inline">\(\theta(X_i)\)</span> by accessing the <code>predictions</code> attribute.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>cf_trained<span class="sc">$</span>predictions <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          [,1]
[1,]  449.5073
[2,]  337.7088
[3,] -484.2860
[4,] 1326.4434
[5,]  604.1448
[6,] 1086.4492</code></pre>
</div>
</div>
</section>
<section id="predict-and-interpret-cate" class="level3 page-columns page-full" data-number="13.2.3">
<h3 data-number="13.2.3" class="anchored" data-anchor-id="predict-and-interpret-cate"><span class="header-section-number">13.2.3</span> Predict and interpret CATE</h3>
<p>Before looking at how to predict <span class="math inline">\(\theta(X)\)</span>, let’s look at which variables are used to split tree nodes. You can get such information using <code>split_frequencies()</code> on a trained causal forest.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">split_frequencies</span>(cf_trained)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     [,1] [,2] [,3] [,4] [,5] [,6]
[1,]  258 1464  119   31   38   85
[2,] 1362 1125  665   39  151  266
[3,] 1984  491 1056   35  219  200
[4,] 1437  321  786    8  118   71</code></pre>
</div>
</div>
<p>In this table, rows represent the depth of the nodes and columns represent covariates. For example, the second variable (<code>educ</code>) was used to do the first split 1464 times and split a node at the second depth 1125 times. A variable with higher numbers of splits at earlier stages is more influential in driving treatment effect heterogeneity. <code>variable_importance()</code> returns a measure of how important each variable is in explaining treatment effect heterogeneity based on the split information.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">variable_importance</span>(cf_trained)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           [,1]
[1,] 0.21900771
[2,] 0.58498841
[3,] 0.10753872
[4,] 0.01362695
[5,] 0.02690857
[6,] 0.04792965</code></pre>
</div>
</div>
<p>So, according to this measure, the second variable (<code>educ</code>) is the most important variable. While variable of importance measure is informative, it does not tell us <span style="color:blue">how </span> the variables are affecting treatment effects. For that, we need to look at <span class="math inline">\(\theta(X)\)</span> at different values of <span class="math inline">\(X\)</span>.</p>
<p>You can use <code>predict()</code> to predict the treatment effect at <span class="math inline">\(X\)</span>. For example, consider the following evaluation point.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>X_0 <span class="ot">&lt;-</span> </span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.table</span>(</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">age =</span> <span class="dv">30</span>,</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">re74 =</span> <span class="dv">40000</span>, </span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">educ =</span> <span class="dv">10</span>,</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">ethn_hispanic =</span> <span class="dv">0</span>,</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">ethn_black =</span> <span class="dv">1</span>,</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">married =</span> <span class="cn">TRUE</span></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that the order of columns of the evaluation data must be the same as that of the <code>X</code> in <code>causal_forest()</code> when you trained a CF model. You can set <code>estimate.variance</code> to <code>TRUE</code> to get <span class="math inline">\(var(\hat{\theta}(X))\)</span> along with the point estimate.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">predict</span>(</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>  cf_trained, </span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">newdata =</span> X_0, </span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">estimate.variance =</span> <span class="cn">TRUE</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  predictions variance.estimates
1   -7351.959            2521470</code></pre>
</div>
</div>
<p>Unlike linear-in-parameter model, there are no coefficients that can immediately tell us how influential each of <span class="math inline">\(X\)</span> is in driving the treatment effect heterogeneity. One way to see the impact of a variable is to change its value while the value of the rest of <span class="math inline">\(X\)</span> is fixed. For example, for the given observed value of <span class="math inline">\(X\)</span> except <code>educ</code>, you can vary the value of <code>educ</code> to see how <code>educ</code> affects the treatment effect. We can do this for all the observations and then can get a good picture of how the treatment effect varies across individuals at different values of <code>educ</code>.</p>
<p>Let’s first create a sequence of <code>educ</code> values at which <span class="math inline">\(\hat{\theta}\)</span> is predicted.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>age_seq <span class="ot">&lt;-</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.table</span>(</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">age =</span> data_trt[, <span class="fu">seq</span>(<span class="fu">min</span>(age), <span class="fu">max</span>(age), <span class="at">length =</span> <span class="dv">30</span>)]</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We then create a dataset where every single individual (observation) in the original data <code>data_trt</code> to have all the age values in <code>age_seq</code> while the value of the rest of <span class="math inline">\(X\)</span> fixed at their own values.</p>

<div class="no-row-height column-margin column-container"><div class="">
<p>Confirm what <code>reshape::expand.grid.df</code> does with this simple example.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>reshape<span class="sc">::</span><span class="fu">expand.grid.df</span>(</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.table</span>(<span class="at">a =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>)), <span class="co"># first data set</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.table</span>(<span class="at">b =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">c =</span> <span class="fu">c</span>(<span class="st">"a"</span>, <span class="st">"b"</span>)) <span class="co"># second data set</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  a b c
1 1 1 a
2 2 1 a
3 3 1 a
4 1 2 b
5 2 2 b
6 3 2 b</code></pre>
</div>
</div>
</div></div><div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>data_te <span class="ot">&lt;-</span> </span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>  reshape<span class="sc">::</span><span class="fu">expand.grid.df</span>(</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>    age_seq, </span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>    data_trt[, .(re74, educ, ethn_hispanic, ethn_black, married, id)]</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span> </span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.table</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s now predict <span class="math inline">\(\hat{\theta}\)</span> with their standard error estimates.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>theta_hat_with_se <span class="ot">&lt;-</span> </span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">predict</span>(cf_trained, <span class="at">newdata =</span> dplyr<span class="sc">::</span><span class="fu">select</span>(data_te, <span class="sc">-</span>id), <span class="at">estimate.variance =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.table</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>  .[, <span class="at">se :=</span> <span class="fu">sqrt</span>(variance.estimates)] <span class="sc">%&gt;%</span> </span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">setnames</span>(<span class="st">"predictions"</span>, <span class="st">"theta_hat"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>  .[, .(theta_hat, se)]</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        theta_hat       se
    1:   196.6412 1643.043
    2:   192.1963 1643.774
    3:   227.6136 1663.190
    4:   316.3815 1659.691
    5:   255.4224 1631.758
   ---                    
80246: -5501.8036 2370.190
80247: -5469.5475 2383.729
80248: -5466.8952 2413.930
80249: -5429.8050 2404.134
80250: -5429.8050 2404.134</code></pre>
</div>
</div>
<p><a href="#fig-ind-3">Figure&nbsp;<span>13.1</span></a> shows the impact of <code>age</code> on treatment effect for the first three individuals of <code>data_trt</code>. For example, if an individual that has the identical values for <span class="math inline">\(X\)</span> except <code>age</code> and also this person is 40 years old, then the treatment effect of the training program would be about $1,000. Standard errors are fairly large and treatment effects are not statistically significantly different from 0 at any value of <code>age</code> for all three individuals. The impact of <code>age</code> seems to be very similar for all the individuals. However, you can see shifts in <span class="math inline">\(\hat{\theta}\)</span> among them. Those shifts are due to the differences in other covariates.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>plot_data <span class="ot">&lt;-</span> <span class="fu">cbind</span>(data_te, theta_hat_with_se)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(plot_data[id <span class="sc">%in%</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, ]) <span class="sc">+</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">y =</span> theta_hat, <span class="at">x =</span> age)) <span class="sc">+</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_ribbon</span>(</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>      <span class="at">ymin =</span> theta_hat <span class="sc">-</span> <span class="fl">1.96</span> <span class="sc">*</span> se, </span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>      <span class="at">ymax =</span> theta_hat <span class="sc">+</span> <span class="fl">1.96</span> <span class="sc">*</span> se, </span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> age</span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">fill =</span> <span class="st">"blue"</span>,</span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">alpha =</span> <span class="fl">0.4</span></span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(. <span class="sc">~</span> id) <span class="sc">+</span></span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-ind-3" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="C03-cf-orf_files/figure-html/fig-ind-3-1.png" class="img-fluid figure-img" width="672"></p>
<p></p><figcaption aria-hidden="true" class="figure-caption">Figure 13.1: Estimated treatment effect for the first three individuals at different values of the age variable</figcaption><p></p>
</figure>
</div>
</div>
</div>
<p><a href="#fig-all-box">Figure&nbsp;<span>13.2</span></a> shows the box-plot of treatment effects for all the individuals. Note that variations observed at each <code>age</code> value is due to heterogeneity in treatment effect driven by covariates other than <code>age</code>. It looks like the three individuals looked at are exceptions. For the majority of individuals, the estimated treatment effects are negative at any value of <code>age</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(plot_data) <span class="sc">+</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="at">y =</span> theta_hat, <span class="at">x =</span> <span class="fu">factor</span>(<span class="fu">round</span>(age, <span class="at">digits =</span> <span class="dv">2</span>)))) <span class="sc">+</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Age"</span>) <span class="sc">+</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Estimated treatment effect"</span>) <span class="sc">+</span></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>)</span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-all-box" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="C03-cf-orf_files/figure-html/fig-all-box-1.png" class="img-fluid figure-img" width="672"></p>
<p></p><figcaption aria-hidden="true" class="figure-caption">Figure 13.2: Box-plot of the estimated treatment effects from all the individuals at different values of the age variable</figcaption><p></p>
</figure>
</div>
</div>
</div>
<p>You can easily repeat this analysis for other covariates to see their impacts as well.</p>
</section>
</section>
<section id="orthogonal-random-forest" class="level2" data-number="13.3">
<h2 data-number="13.3" class="anchored" data-anchor-id="orthogonal-random-forest"><span class="header-section-number">13.3</span> Orthogonal Random Forest</h2>
</section>
<section id="references" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="references">References</h2>
<div id="refs" class="references csl-bib-body hanging-indent" role="doc-bibliography">
<div id="ref-athey2019generalized" class="csl-entry" role="doc-biblioentry">
Athey, Susan, Julie Tibshirani, and Stefan Wager. 2019. <span>“Generalized Random Forests.”</span> <em>The Annals of Statistics</em> 47 (2): 1148–78.
</div>
<div id="ref-oprescu2019orthogonal" class="csl-entry" role="doc-biblioentry">
Oprescu, Miruna, Vasilis Syrgkanis, and Zhiwei Steven Wu. 2019. <span>“Orthogonal Random Forest for Causal Inference.”</span> In <em>International Conference on Machine Learning</em>, 4932–41. PMLR.
</div>
<div id="ref-wager_estimation_2018" class="csl-entry" role="doc-biblioentry">
Wager, Stefan, and Susan Athey. 2018. <span>“Estimation and <span>Inference</span> of <span>Heterogeneous</span> <span>Treatment</span> <span>Effects</span> Using <span>Random</span> <span>Forests</span>.”</span> <em>Journal of the American Statistical Association</em> 113 (523): 1228–42. <a href="https://doi.org/10.1080/01621459.2017.1319839">https://doi.org/10.1080/01621459.2017.1319839</a>.
</div>
</div>


<!-- -->

</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      let href = ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./C02-xstr-learner.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">S-, X-, T-, and R-learner</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./C04-cf-extension.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Extensions of Causal Forest</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb25" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Forest-based CATE Estimators {#sec-cf-orf}</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>In @sec-het-dml, we saw some special cases of R-learner/DML where the final model is estimate by parametrically using a linear-in-parameter model and non-parametrically using extreme gradient boosting. Here, we learn two methods that estimate CATE non-parametrically: causal forest <span class="co">[</span><span class="ot">@athey2019generalized</span><span class="co">]</span> and orthogonal forest <span class="co">[</span><span class="ot">@oprescu2019orthogonal</span><span class="co">]</span>.</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>:::{.callout-important}</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a><span class="fu">## What you will learn</span></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>Causal forest </span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a><span class="ss">  * </span>Mechanics</span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a><span class="ss">  * </span>Hyper-parameter</span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a><span class="ss">  * </span>Predict heterogeneous treatment effects (CATE)</span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a><span class="ss">  * </span>Predict average treatment effects (ATE)</span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a><span class="ss">  * </span>Interpretation</span>
<span id="cb25-15"><a href="#cb25-15" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>Orthogonal forest </span>
<span id="cb25-16"><a href="#cb25-16" aria-hidden="true" tabindex="-1"></a><span class="ss">  * </span>Differences from causal forest</span>
<span id="cb25-17"><a href="#cb25-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-18"><a href="#cb25-18" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb25-19"><a href="#cb25-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-20"><a href="#cb25-20" aria-hidden="true" tabindex="-1"></a>:::{.callout-tip}</span>
<span id="cb25-21"><a href="#cb25-21" aria-hidden="true" tabindex="-1"></a><span class="fu">## Background knowledge</span></span>
<span id="cb25-22"><a href="#cb25-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-23"><a href="#cb25-23" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>(preferable) DML/R-learner (@sec-het-dml)</span>
<span id="cb25-24"><a href="#cb25-24" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>(preferable) GRF (@sec-grf)</span>
<span id="cb25-25"><a href="#cb25-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-26"><a href="#cb25-26" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb25-27"><a href="#cb25-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-28"><a href="#cb25-28" aria-hidden="true" tabindex="-1"></a>:::{.callout-note}</span>
<span id="cb25-29"><a href="#cb25-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-30"><a href="#cb25-30" aria-hidden="true" tabindex="-1"></a><span class="fu">## Packages to load for replication</span></span>
<span id="cb25-31"><a href="#cb25-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-34"><a href="#cb25-34" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-35"><a href="#cb25-35" aria-hidden="true" tabindex="-1"></a><span class="co">#| include: false</span></span>
<span id="cb25-36"><a href="#cb25-36" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb25-37"><a href="#cb25-37" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb25-38"><a href="#cb25-38" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(grf)</span>
<span id="cb25-39"><a href="#cb25-39" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-40"><a href="#cb25-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-43"><a href="#cb25-43" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-44"><a href="#cb25-44" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb25-45"><a href="#cb25-45" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb25-46"><a href="#cb25-46" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb25-47"><a href="#cb25-47" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(grf)</span>
<span id="cb25-48"><a href="#cb25-48" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-49"><a href="#cb25-49" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb25-50"><a href="#cb25-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-51"><a href="#cb25-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-52"><a href="#cb25-52" aria-hidden="true" tabindex="-1"></a>**Left to be added**</span>
<span id="cb25-53"><a href="#cb25-53" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>how variance is estimated</span>
<span id="cb25-54"><a href="#cb25-54" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>balance</span>
<span id="cb25-55"><a href="#cb25-55" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>ATE</span>
<span id="cb25-56"><a href="#cb25-56" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>linear correction with locally weighted linear regression</span>
<span id="cb25-57"><a href="#cb25-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-58"><a href="#cb25-58" aria-hidden="true" tabindex="-1"></a><span class="fu">## Model</span></span>
<span id="cb25-59"><a href="#cb25-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-60"><a href="#cb25-60" aria-hidden="true" tabindex="-1"></a>The heterogeneous treatment effect model of interest in this chapter is the same as the one in @sec-het-dml.</span>
<span id="cb25-61"><a href="#cb25-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-62"><a href="#cb25-62" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb25-63"><a href="#cb25-63" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb25-64"><a href="#cb25-64" aria-hidden="true" tabindex="-1"></a>Y &amp; = \theta(X)\cdot T + g(X) + \varepsilon <span class="sc">\\</span></span>
<span id="cb25-65"><a href="#cb25-65" aria-hidden="true" tabindex="-1"></a>T &amp; = f(X) + \eta </span>
<span id="cb25-66"><a href="#cb25-66" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb25-67"><a href="#cb25-67" aria-hidden="true" tabindex="-1"></a>$$ {#eq-model-het-te}</span>
<span id="cb25-68"><a href="#cb25-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-69"><a href="#cb25-69" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>$Y$: dependent variable</span>
<span id="cb25-70"><a href="#cb25-70" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>$T$: treatment variable (can be either binary dummy or continuous)</span>
<span id="cb25-71"><a href="#cb25-71" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>$X$: features</span>
<span id="cb25-72"><a href="#cb25-72" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-73"><a href="#cb25-73" aria-hidden="true" tabindex="-1"></a>Causal forest and orthogonal random forest is consistent only if the following conditions fold.</span>
<span id="cb25-74"><a href="#cb25-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-75"><a href="#cb25-75" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>$E<span class="co">[</span><span class="ot">\varepsilon|X</span><span class="co">]</span> = 0$</span>
<span id="cb25-76"><a href="#cb25-76" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>$E<span class="co">[</span><span class="ot">\eta|X</span><span class="co">]</span> = 0$</span>
<span id="cb25-77"><a href="#cb25-77" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>$E<span class="co">[</span><span class="ot">\eta\cdot\varepsilon|X</span><span class="co">]</span> = 0$</span>
<span id="cb25-78"><a href="#cb25-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-79"><a href="#cb25-79" aria-hidden="true" tabindex="-1"></a><span class="fu">## Causal Forest</span></span>
<span id="cb25-80"><a href="#cb25-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-81"><a href="#cb25-81" aria-hidden="true" tabindex="-1"></a><span class="fu">### Brief description of how CF works</span></span>
<span id="cb25-82"><a href="#cb25-82" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-83"><a href="#cb25-83" aria-hidden="true" tabindex="-1"></a>This section provides a brief and cursory description of how CF works. A much more detailed treatment of how CF (and also GRF) works along with the explanations of hyper-parameters is provided in @sec-understand-grf-example.</span>
<span id="cb25-84"><a href="#cb25-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-85"><a href="#cb25-85" aria-hidden="true" tabindex="-1"></a>Causal Forest (CF) (as implemented by the R <span class="in">`grf`</span> package or python <span class="in">`econml`</span> package) is a special type of R-learner (also a DML) and also a special case of generalized random forest (GRF). </span>
<span id="cb25-86"><a href="#cb25-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-87"><a href="#cb25-87" aria-hidden="true" tabindex="-1"></a>::: {.column-margin}</span>
<span id="cb25-88"><a href="#cb25-88" aria-hidden="true" tabindex="-1"></a>CF can be implemented using the R <span class="in">`grf`</span> package or python <span class="in">`econml`</span> package. Both of them implements CF as an R-learner. However, the original causal forest proposed in @wager_estimation_2018 does not follow an R-learner procedure.</span>
<span id="cb25-89"><a href="#cb25-89" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb25-90"><a href="#cb25-90" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-91"><a href="#cb25-91" aria-hidden="true" tabindex="-1"></a>Let $\hat{f}(X_i)$ and $\hat{g}(X_i)$ denote the estimation of $E<span class="co">[</span><span class="ot">Y|X</span><span class="co">]</span>$ and $E<span class="co">[</span><span class="ot">T|X</span><span class="co">]</span>$, respectively. Further, let $\hat{\tilde{Y_i}}$ and $\hat{\tilde{T_i}}$ denote $Y_i - \hat{f}(X_i)$ and $T_i - \hat{g}(X_i)$, respectively.</span>
<span id="cb25-92"><a href="#cb25-92" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-93"><a href="#cb25-93" aria-hidden="true" tabindex="-1"></a>Then, CF estimates $\theta(X)$ at $X = X_0$ by solving the following equation:</span>
<span id="cb25-94"><a href="#cb25-94" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-95"><a href="#cb25-95" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb25-96"><a href="#cb25-96" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb25-97"><a href="#cb25-97" aria-hidden="true" tabindex="-1"></a>\hat{\theta}(X_0) = argmin_{\theta}\;\;\sum_{i=1}^N \alpha_i(X_i, X_0)<span class="co">[</span><span class="ot">\hat{\tilde{Y_i}} - \theta\cdot \hat{\tilde{T_i}}</span><span class="co">]</span>^2</span>
<span id="cb25-98"><a href="#cb25-98" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb25-99"><a href="#cb25-99" aria-hidden="true" tabindex="-1"></a>$$ {#eq-solve-cf}</span>
<span id="cb25-100"><a href="#cb25-100" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-101"><a href="#cb25-101" aria-hidden="true" tabindex="-1"></a>where $\alpha_i(X_i, X_0)$ is the weight given to each $i$. The F.O.C is </span>
<span id="cb25-102"><a href="#cb25-102" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-103"><a href="#cb25-103" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb25-104"><a href="#cb25-104" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb25-105"><a href="#cb25-105" aria-hidden="true" tabindex="-1"></a>2 \cdot \sum_{i=1}^N \alpha_i(X_i, X_0)<span class="co">[</span><span class="ot">\hat{\tilde{Y_i}} - \theta\cdot \hat{\tilde{T_i}}</span><span class="co">]</span>\hat{\tilde{T_i}} = 0 </span>
<span id="cb25-106"><a href="#cb25-106" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb25-107"><a href="#cb25-107" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb25-108"><a href="#cb25-108" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-109"><a href="#cb25-109" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb25-110"><a href="#cb25-110" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb25-111"><a href="#cb25-111" aria-hidden="true" tabindex="-1"></a>\hat{\theta}(X_0) = \frac{\sum_{i=1}^N\alpha_i(X_i, X_0)\cdot\hat{\tilde{Y}}_i\cdot \hat{\tilde{T}}_i}{\sum_{i=1}^N\alpha_i(X_i, X_0)\cdot\hat{\tilde{T}}_i\cdot \hat{\tilde{T}}_i}</span>
<span id="cb25-112"><a href="#cb25-112" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb25-113"><a href="#cb25-113" aria-hidden="true" tabindex="-1"></a>$$ {#eq-theta-solution}</span>
<span id="cb25-114"><a href="#cb25-114" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-115"><a href="#cb25-115" aria-hidden="true" tabindex="-1"></a>Unlike the DML approaches we saw in @sec-het-dml that uses a linear model as the final model, CF does not assume any functional form for how $X$ affects $\theta$ as you can see from the above minimization problem. As mentioned earlier, CF is a special case of GRF (discussed in @sec-grf), so CF is also a local constant regression (see @sec-local-reg for a brief discussion on local regression).</span>
<span id="cb25-116"><a href="#cb25-116" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-117"><a href="#cb25-117" aria-hidden="true" tabindex="-1"></a>$\alpha_i(X_i, X_0)$ is determined based on the trees trained based on the pseudo outcomes that are defined specifically for causal forest estimation. Suppose $T$ trees have been built and let $\eta_{i,t}(X_i, X_0)$ be 1 if observation $i$ belongs to the same leaf as $X_0$ in tree $t$. Then, </span>
<span id="cb25-118"><a href="#cb25-118" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-119"><a href="#cb25-119" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb25-120"><a href="#cb25-120" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb25-121"><a href="#cb25-121" aria-hidden="true" tabindex="-1"></a>\alpha_i(X_i, X_0) = \frac{1}{T}\sum_{t=1}^T\frac{\eta_{i,t}(X_i, X_0)}{\sum_{i=1}^N\eta_{i,t}(X_i, X_0)}</span>
<span id="cb25-122"><a href="#cb25-122" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb25-123"><a href="#cb25-123" aria-hidden="true" tabindex="-1"></a>$$ {#eq-weight-cf}</span>
<span id="cb25-124"><a href="#cb25-124" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-125"><a href="#cb25-125" aria-hidden="true" tabindex="-1"></a>So, the weight given to observation $i$ is higher if observation $i$ belongs to the same leaf as the evaluation point $X_0$ in more trees. </span>
<span id="cb25-126"><a href="#cb25-126" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-127"><a href="#cb25-127" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-128"><a href="#cb25-128" aria-hidden="true" tabindex="-1"></a><span class="fu">### Training a causal forest</span></span>
<span id="cb25-129"><a href="#cb25-129" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-130"><a href="#cb25-130" aria-hidden="true" tabindex="-1"></a>We can use the <span class="in">`causal_forest()`</span> function from the <span class="in">`grf`</span> package to train a CF model in R. In Python, you can use <span class="in">`CausalForestDML()`</span> from the <span class="in">`econml`</span> package or <span class="in">`GRFForestRegressor`</span> from the <span class="in">`skgrf`</span> package.</span>
<span id="cb25-131"><a href="#cb25-131" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-132"><a href="#cb25-132" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-133"><a href="#cb25-133" aria-hidden="true" tabindex="-1"></a>As of now, there are some notable differences between <span class="in">`grf`</span> and <span class="in">`econml`</span>. </span>
<span id="cb25-134"><a href="#cb25-134" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-135"><a href="#cb25-135" aria-hidden="true" tabindex="-1"></a>:::{.callout-tip}</span>
<span id="cb25-136"><a href="#cb25-136" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-137"><a href="#cb25-137" aria-hidden="true" tabindex="-1"></a><span class="fu">## Some differences between the `grf` and `econml` implementation of causal forest</span></span>
<span id="cb25-138"><a href="#cb25-138" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-139"><a href="#cb25-139" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>While <span class="in">`grf`</span> support clustering, <span class="in">`econml`</span> does not.</span>
<span id="cb25-140"><a href="#cb25-140" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-141"><a href="#cb25-141" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>It is easy to modify the first stage estimations in <span class="in">`econml`</span>. <span class="in">`grf`</span> uses random forest by default for them. If you would like to try other ML methods, then you need to write a code to predict $Y - E<span class="co">[</span><span class="ot">Y|X, W</span><span class="co">]</span>$ and $T - E<span class="co">[</span><span class="ot">T|X, W</span><span class="co">]</span>$ with your choice of ML methods yourself (see @sec-mlr3-in-action for an example code to do this).</span>
<span id="cb25-142"><a href="#cb25-142" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-143"><a href="#cb25-143" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span><span class="in">`grf`</span> does not cross-fit in the first-stage estimation. Rather, it uses out-of-bag prediction from the trained random forest models, which avoids over-fitting (Note that the motivation behind cross-fitting in DML is to avoid over-fitting bias in the second stage estimation. See @sec-dml).</span>
<span id="cb25-144"><a href="#cb25-144" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-145"><a href="#cb25-145" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span><span class="in">`grf`</span> does not make distinctions between $X$ and $W$.</span>
<span id="cb25-146"><a href="#cb25-146" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-147"><a href="#cb25-147" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb25-148"><a href="#cb25-148" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-149"><a href="#cb25-149" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-152"><a href="#cb25-152" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-153"><a href="#cb25-153" aria-hidden="true" tabindex="-1"></a><span class="co">#=== load the Treatment dataset ===#</span></span>
<span id="cb25-154"><a href="#cb25-154" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"Treatment"</span>, <span class="at">package =</span> <span class="st">"Ecdat"</span>)</span>
<span id="cb25-155"><a href="#cb25-155" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-156"><a href="#cb25-156" aria-hidden="true" tabindex="-1"></a><span class="co">#=== convert to a data.table ===#</span></span>
<span id="cb25-157"><a href="#cb25-157" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb25-158"><a href="#cb25-158" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> </span>
<span id="cb25-159"><a href="#cb25-159" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.table</span>(Treatment) <span class="sc">%&gt;%</span> </span>
<span id="cb25-160"><a href="#cb25-160" aria-hidden="true" tabindex="-1"></a>  <span class="co">#=== create an id variable ===#</span></span>
<span id="cb25-161"><a href="#cb25-161" aria-hidden="true" tabindex="-1"></a>  .[, <span class="at">id :=</span> <span class="dv">1</span><span class="sc">:</span>.N]</span>
<span id="cb25-162"><a href="#cb25-162" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb25-163"><a href="#cb25-163" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-164"><a href="#cb25-164" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-165"><a href="#cb25-165" aria-hidden="true" tabindex="-1"></a>Here are the variables in this dataset that we use.</span>
<span id="cb25-166"><a href="#cb25-166" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-167"><a href="#cb25-167" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span><span class="in">`re78`</span> ($Y$): real annual earnings in 1978 (after the treatment)</span>
<span id="cb25-168"><a href="#cb25-168" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span><span class="in">`treat`</span> ($T$): <span class="in">`TRUE`</span> if a person had gone through a training, <span class="in">`FALSE`</span> otherwise. </span>
<span id="cb25-169"><a href="#cb25-169" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-170"><a href="#cb25-170" aria-hidden="true" tabindex="-1"></a>$X$ includes</span>
<span id="cb25-171"><a href="#cb25-171" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-172"><a href="#cb25-172" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span><span class="in">`re74`</span>: real annual earnings in 1978 (after the treatment)</span>
<span id="cb25-173"><a href="#cb25-173" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span><span class="in">`age`</span>: age</span>
<span id="cb25-174"><a href="#cb25-174" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span><span class="in">`educ`</span>: education in years</span>
<span id="cb25-175"><a href="#cb25-175" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span><span class="in">`ethn`</span>: one of "other", "black", "hispanic"</span>
<span id="cb25-176"><a href="#cb25-176" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span><span class="in">`married`</span>: married or not</span>
<span id="cb25-177"><a href="#cb25-177" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-178"><a href="#cb25-178" aria-hidden="true" tabindex="-1"></a><span class="in">`grf::causal_forest()`</span> takes only numeric values for $X$. So, we will one-hot encode <span class="in">`ethn`</span>, which is a factor variable at the moment.</span>
<span id="cb25-179"><a href="#cb25-179" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-182"><a href="#cb25-182" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-183"><a href="#cb25-183" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb25-184"><a href="#cb25-184" aria-hidden="true" tabindex="-1"></a>data_trt <span class="ot">&lt;-</span> mltools<span class="sc">::</span><span class="fu">one_hot</span>(data)</span>
<span id="cb25-185"><a href="#cb25-185" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb25-186"><a href="#cb25-186" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-187"><a href="#cb25-187" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-188"><a href="#cb25-188" aria-hidden="true" tabindex="-1"></a>We now have <span class="in">`ethn_black`</span>, <span class="in">`ethn_hispanic`</span>, and <span class="in">`ethn_other`</span> from <span class="in">`ethn`</span>. The model we are estimating is as follows:</span>
<span id="cb25-189"><a href="#cb25-189" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-190"><a href="#cb25-190" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb25-191"><a href="#cb25-191" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb25-192"><a href="#cb25-192" aria-hidden="true" tabindex="-1"></a>re78 &amp; = \theta(age, re74, educ, ethn<span class="sc">\_</span>hipanic, ethn<span class="sc">\_</span>black, married)\cdot treat + g(age, re74, educ, ethn<span class="sc">\_</span>hipanic, ethn<span class="sc">\_</span>black, married) + \varepsilon <span class="sc">\\</span></span>
<span id="cb25-193"><a href="#cb25-193" aria-hidden="true" tabindex="-1"></a>treat &amp; = f(age, re74, educ, ethn<span class="sc">\_</span>hipanic, ethn<span class="sc">\_</span>black, married) + \eta</span>
<span id="cb25-194"><a href="#cb25-194" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb25-195"><a href="#cb25-195" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb25-196"><a href="#cb25-196" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-197"><a href="#cb25-197" aria-hidden="true" tabindex="-1"></a>:::{.callout-warning}</span>
<span id="cb25-198"><a href="#cb25-198" aria-hidden="true" tabindex="-1"></a>Note that we are paying no attention to the potential endogeneity problem here. This is just a demonstration and the results are likely to be biased. We will look at instrumental forest later in @sec-cf-extension, which is consistent under confoundedness as long as you can find appropriate external instruments.</span>
<span id="cb25-199"><a href="#cb25-199" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb25-200"><a href="#cb25-200" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-201"><a href="#cb25-201" aria-hidden="true" tabindex="-1"></a>In running <span class="in">`causal_forest()`</span>, there are many hyper-parameters and options that we need to be aware of. </span>
<span id="cb25-202"><a href="#cb25-202" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-203"><a href="#cb25-203" aria-hidden="true" tabindex="-1"></a>Since CF is a GRF and GRF uses random forest algorithm on appropriate pseudo outcome, it is natural that some of the CF hyper-parameters are the same as the ones for RF (by <span class="in">`ranger()`</span>).</span>
<span id="cb25-204"><a href="#cb25-204" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-205"><a href="#cb25-205" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span><span class="in">`num.trees`</span>: number of trees</span>
<span id="cb25-206"><a href="#cb25-206" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span><span class="in">`mtry`</span>: number of variables tried in each split (default is $\sqrt{K}$)</span>
<span id="cb25-207"><a href="#cb25-207" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span><span class="in">`min.node.size`</span>: minimum number of observations in each leaf (default is 5)</span>
<span id="cb25-208"><a href="#cb25-208" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-209"><a href="#cb25-209" aria-hidden="true" tabindex="-1"></a>A hyper-parameter that certainly affects tree building process is <span class="in">`sample.fraction`</span>, which we saw earlier. </span>
<span id="cb25-210"><a href="#cb25-210" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-211"><a href="#cb25-211" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span><span class="in">`sample.fraction`</span>: fraction of the data used to build each tree (default is 0.5)</span>
<span id="cb25-212"><a href="#cb25-212" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-213"><a href="#cb25-213" aria-hidden="true" tabindex="-1"></a>A higher value of <span class="in">`sample.fraction`</span> means that the trees are more correlated as they share more of the same observations. </span>
<span id="cb25-214"><a href="#cb25-214" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-215"><a href="#cb25-215" aria-hidden="true" tabindex="-1"></a>There are three honesty-related options. </span>
<span id="cb25-216"><a href="#cb25-216" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-217"><a href="#cb25-217" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span><span class="in">`honesty`</span>: <span class="in">`TRUE`</span> if using honest-sampling, 0 otherwise (default is <span class="in">`TRUE`</span>)</span>
<span id="cb25-218"><a href="#cb25-218" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span><span class="in">`honesty.fraction`</span>: fraction of the data (after <span class="in">`sample.fraction`</span> is applied) that is grouped into $J_1$, which is used to determine splitting rules</span>
<span id="cb25-219"><a href="#cb25-219" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span><span class="in">`honesty.prune.leaves`</span>: <span class="in">`TRUE`</span> if the leaves with no samples are pruned (default is <span class="in">`TRUE`</span>)</span>
<span id="cb25-220"><a href="#cb25-220" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-221"><a href="#cb25-221" aria-hidden="true" tabindex="-1"></a>Let's now train CF using <span class="in">`data_trt`</span>. </span>
<span id="cb25-222"><a href="#cb25-222" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-225"><a href="#cb25-225" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-226"><a href="#cb25-226" aria-hidden="true" tabindex="-1"></a>cf_trained <span class="ot">&lt;-</span></span>
<span id="cb25-227"><a href="#cb25-227" aria-hidden="true" tabindex="-1"></a>  grf<span class="sc">::</span><span class="fu">causal_forest</span>(</span>
<span id="cb25-228"><a href="#cb25-228" aria-hidden="true" tabindex="-1"></a>    <span class="at">X =</span> data_trt[, .(age, re74, educ, ethn_hispanic, ethn_black, married)] <span class="sc">%&gt;%</span> <span class="fu">as.matrix</span>(),</span>
<span id="cb25-229"><a href="#cb25-229" aria-hidden="true" tabindex="-1"></a>    <span class="at">Y =</span> data_trt[, re78],</span>
<span id="cb25-230"><a href="#cb25-230" aria-hidden="true" tabindex="-1"></a>    <span class="at">W =</span> data_trt[, treat]</span>
<span id="cb25-231"><a href="#cb25-231" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb25-232"><a href="#cb25-232" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-233"><a href="#cb25-233" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-234"><a href="#cb25-234" aria-hidden="true" tabindex="-1"></a>Here is what <span class="in">`cf_trained`</span> has as its attributes.</span>
<span id="cb25-235"><a href="#cb25-235" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-238"><a href="#cb25-238" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-239"><a href="#cb25-239" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(cf_trained)</span>
<span id="cb25-240"><a href="#cb25-240" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-241"><a href="#cb25-241" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-242"><a href="#cb25-242" aria-hidden="true" tabindex="-1"></a>For example, you can get $\theta(X_i)$ by accessing the <span class="in">`predictions`</span> attribute.</span>
<span id="cb25-243"><a href="#cb25-243" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-246"><a href="#cb25-246" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-247"><a href="#cb25-247" aria-hidden="true" tabindex="-1"></a>cf_trained<span class="sc">$</span>predictions <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span>
<span id="cb25-248"><a href="#cb25-248" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-249"><a href="#cb25-249" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-250"><a href="#cb25-250" aria-hidden="true" tabindex="-1"></a><span class="fu">### Predict and interpret CATE</span></span>
<span id="cb25-251"><a href="#cb25-251" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-252"><a href="#cb25-252" aria-hidden="true" tabindex="-1"></a>Before looking at how to predict $\theta(X)$, let's look at which variables are used to split tree nodes. You can get such information using <span class="in">`split_frequencies()`</span> on a trained causal forest.</span>
<span id="cb25-253"><a href="#cb25-253" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-256"><a href="#cb25-256" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-257"><a href="#cb25-257" aria-hidden="true" tabindex="-1"></a><span class="fu">split_frequencies</span>(cf_trained)</span>
<span id="cb25-258"><a href="#cb25-258" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-259"><a href="#cb25-259" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-262"><a href="#cb25-262" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-263"><a href="#cb25-263" aria-hidden="true" tabindex="-1"></a><span class="co">#| include: false</span></span>
<span id="cb25-264"><a href="#cb25-264" aria-hidden="true" tabindex="-1"></a>split_data <span class="ot">&lt;-</span> <span class="fu">split_frequencies</span>(cf_trained)</span>
<span id="cb25-265"><a href="#cb25-265" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-266"><a href="#cb25-266" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-267"><a href="#cb25-267" aria-hidden="true" tabindex="-1"></a>In this table, rows represent the depth of the nodes and columns represent covariates. For example, the second variable (<span class="in">`educ`</span>) was used to do the first split <span class="in">`r split_data[1, 2]`</span> times and split a node at the second depth <span class="in">`r split_data[2, 2]`</span> times. A variable with higher numbers of splits at earlier stages is more influential in driving treatment effect heterogeneity. <span class="in">`variable_importance()`</span> returns a measure of how important each variable is in explaining treatment effect heterogeneity based on the split information.</span>
<span id="cb25-268"><a href="#cb25-268" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-271"><a href="#cb25-271" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-272"><a href="#cb25-272" aria-hidden="true" tabindex="-1"></a><span class="fu">variable_importance</span>(cf_trained)</span>
<span id="cb25-273"><a href="#cb25-273" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-274"><a href="#cb25-274" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-275"><a href="#cb25-275" aria-hidden="true" tabindex="-1"></a>So, according to this measure, the second variable (<span class="in">`educ`</span>) is the most important variable. While variable of importance measure is informative, it does not tell us <span class="kw">&lt;span</span> <span class="er">style</span><span class="ot">=</span><span class="st">"color:blue"</span><span class="kw">&gt;</span>how <span class="kw">&lt;/span&gt;</span> the variables are affecting treatment effects. For that, we need to look at $\theta(X)$ at different values of $X$.</span>
<span id="cb25-276"><a href="#cb25-276" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-277"><a href="#cb25-277" aria-hidden="true" tabindex="-1"></a>You can use <span class="in">`predict()`</span> to predict the treatment effect at $X$. For example, consider the following evaluation point.</span>
<span id="cb25-278"><a href="#cb25-278" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-281"><a href="#cb25-281" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-282"><a href="#cb25-282" aria-hidden="true" tabindex="-1"></a>X_0 <span class="ot">&lt;-</span> </span>
<span id="cb25-283"><a href="#cb25-283" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.table</span>(</span>
<span id="cb25-284"><a href="#cb25-284" aria-hidden="true" tabindex="-1"></a>    <span class="at">age =</span> <span class="dv">30</span>,</span>
<span id="cb25-285"><a href="#cb25-285" aria-hidden="true" tabindex="-1"></a>    <span class="at">re74 =</span> <span class="dv">40000</span>, </span>
<span id="cb25-286"><a href="#cb25-286" aria-hidden="true" tabindex="-1"></a>    <span class="at">educ =</span> <span class="dv">10</span>,</span>
<span id="cb25-287"><a href="#cb25-287" aria-hidden="true" tabindex="-1"></a>    <span class="at">ethn_hispanic =</span> <span class="dv">0</span>,</span>
<span id="cb25-288"><a href="#cb25-288" aria-hidden="true" tabindex="-1"></a>    <span class="at">ethn_black =</span> <span class="dv">1</span>,</span>
<span id="cb25-289"><a href="#cb25-289" aria-hidden="true" tabindex="-1"></a>    <span class="at">married =</span> <span class="cn">TRUE</span></span>
<span id="cb25-290"><a href="#cb25-290" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb25-291"><a href="#cb25-291" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-292"><a href="#cb25-292" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-293"><a href="#cb25-293" aria-hidden="true" tabindex="-1"></a>Note that the order of columns of the evaluation data must be the same as that of the <span class="in">`X`</span> in <span class="in">`causal_forest()`</span> when you trained a CF model. You can set <span class="in">`estimate.variance`</span> to <span class="in">`TRUE`</span> to get $var(\hat{\theta}(X))$ along with the point estimate.</span>
<span id="cb25-294"><a href="#cb25-294" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-297"><a href="#cb25-297" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-298"><a href="#cb25-298" aria-hidden="true" tabindex="-1"></a><span class="fu">predict</span>(</span>
<span id="cb25-299"><a href="#cb25-299" aria-hidden="true" tabindex="-1"></a>  cf_trained, </span>
<span id="cb25-300"><a href="#cb25-300" aria-hidden="true" tabindex="-1"></a>  <span class="at">newdata =</span> X_0, </span>
<span id="cb25-301"><a href="#cb25-301" aria-hidden="true" tabindex="-1"></a>  <span class="at">estimate.variance =</span> <span class="cn">TRUE</span></span>
<span id="cb25-302"><a href="#cb25-302" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb25-303"><a href="#cb25-303" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-304"><a href="#cb25-304" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-305"><a href="#cb25-305" aria-hidden="true" tabindex="-1"></a>Unlike linear-in-parameter model, there are no coefficients that can immediately tell us how influential each of $X$ is in driving the treatment effect heterogeneity. One way to see the impact of a variable is to change its value while the value of the rest of $X$ is fixed. For example, for the given observed value of $X$ except <span class="in">`educ`</span>, you can vary the value of <span class="in">`educ`</span> to see how <span class="in">`educ`</span> affects the treatment effect. We can do this for all the observations and then can get a good picture of how the treatment effect varies across individuals at different values of <span class="in">`educ`</span>. </span>
<span id="cb25-306"><a href="#cb25-306" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-307"><a href="#cb25-307" aria-hidden="true" tabindex="-1"></a>Let's first create a sequence of <span class="in">`educ`</span> values at which $\hat{\theta}$ is predicted. </span>
<span id="cb25-308"><a href="#cb25-308" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-311"><a href="#cb25-311" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-312"><a href="#cb25-312" aria-hidden="true" tabindex="-1"></a>age_seq <span class="ot">&lt;-</span></span>
<span id="cb25-313"><a href="#cb25-313" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.table</span>(</span>
<span id="cb25-314"><a href="#cb25-314" aria-hidden="true" tabindex="-1"></a>    <span class="at">age =</span> data_trt[, <span class="fu">seq</span>(<span class="fu">min</span>(age), <span class="fu">max</span>(age), <span class="at">length =</span> <span class="dv">30</span>)]</span>
<span id="cb25-315"><a href="#cb25-315" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb25-316"><a href="#cb25-316" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-317"><a href="#cb25-317" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-318"><a href="#cb25-318" aria-hidden="true" tabindex="-1"></a>We then create a dataset where every single individual (observation) in the original data <span class="in">`data_trt`</span> to have all the age values in <span class="in">`age_seq`</span> while the value of the rest of $X$ fixed at their own values.</span>
<span id="cb25-319"><a href="#cb25-319" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-320"><a href="#cb25-320" aria-hidden="true" tabindex="-1"></a>::: {.column-margin}</span>
<span id="cb25-321"><a href="#cb25-321" aria-hidden="true" tabindex="-1"></a>Confirm what <span class="in">`reshape::expand.grid.df`</span> does with this simple example.</span>
<span id="cb25-322"><a href="#cb25-322" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-325"><a href="#cb25-325" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-326"><a href="#cb25-326" aria-hidden="true" tabindex="-1"></a>reshape<span class="sc">::</span><span class="fu">expand.grid.df</span>(</span>
<span id="cb25-327"><a href="#cb25-327" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.table</span>(<span class="at">a =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>)), <span class="co"># first data set</span></span>
<span id="cb25-328"><a href="#cb25-328" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.table</span>(<span class="at">b =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">c =</span> <span class="fu">c</span>(<span class="st">"a"</span>, <span class="st">"b"</span>)) <span class="co"># second data set</span></span>
<span id="cb25-329"><a href="#cb25-329" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb25-330"><a href="#cb25-330" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-331"><a href="#cb25-331" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb25-332"><a href="#cb25-332" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-335"><a href="#cb25-335" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-336"><a href="#cb25-336" aria-hidden="true" tabindex="-1"></a>data_te <span class="ot">&lt;-</span> </span>
<span id="cb25-337"><a href="#cb25-337" aria-hidden="true" tabindex="-1"></a>  reshape<span class="sc">::</span><span class="fu">expand.grid.df</span>(</span>
<span id="cb25-338"><a href="#cb25-338" aria-hidden="true" tabindex="-1"></a>    age_seq, </span>
<span id="cb25-339"><a href="#cb25-339" aria-hidden="true" tabindex="-1"></a>    data_trt[, .(re74, educ, ethn_hispanic, ethn_black, married, id)]</span>
<span id="cb25-340"><a href="#cb25-340" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span> </span>
<span id="cb25-341"><a href="#cb25-341" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.table</span>()</span>
<span id="cb25-342"><a href="#cb25-342" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-343"><a href="#cb25-343" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-344"><a href="#cb25-344" aria-hidden="true" tabindex="-1"></a>Let's now predict $\hat{\theta}$ with their standard error estimates.</span>
<span id="cb25-345"><a href="#cb25-345" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-348"><a href="#cb25-348" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-349"><a href="#cb25-349" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb25-350"><a href="#cb25-350" aria-hidden="true" tabindex="-1"></a>theta_hat_with_se <span class="ot">&lt;-</span> </span>
<span id="cb25-351"><a href="#cb25-351" aria-hidden="true" tabindex="-1"></a>  <span class="fu">predict</span>(cf_trained, <span class="at">newdata =</span> dplyr<span class="sc">::</span><span class="fu">select</span>(data_te, <span class="sc">-</span>id), <span class="at">estimate.variance =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb25-352"><a href="#cb25-352" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.table</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb25-353"><a href="#cb25-353" aria-hidden="true" tabindex="-1"></a>  .[, <span class="at">se :=</span> <span class="fu">sqrt</span>(variance.estimates)] <span class="sc">%&gt;%</span> </span>
<span id="cb25-354"><a href="#cb25-354" aria-hidden="true" tabindex="-1"></a>  <span class="fu">setnames</span>(<span class="st">"predictions"</span>, <span class="st">"theta_hat"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb25-355"><a href="#cb25-355" aria-hidden="true" tabindex="-1"></a>  .[, .(theta_hat, se)]</span>
<span id="cb25-356"><a href="#cb25-356" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb25-357"><a href="#cb25-357" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-358"><a href="#cb25-358" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-359"><a href="#cb25-359" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-360"><a href="#cb25-360" aria-hidden="true" tabindex="-1"></a>@fig-ind-3 shows the impact of <span class="in">`age`</span> on treatment effect for the first three individuals of <span class="in">`data_trt`</span>. For example, if an individual that has the identical values for $X$ except <span class="in">`age`</span> and also this person is 40 years old, then the treatment effect of the training program would be about $1,000. Standard errors are fairly large and treatment effects are not statistically significantly different from 0 at any value of <span class="in">`age`</span> for all three individuals. The impact of <span class="in">`age`</span> seems to be very similar for all the individuals. However, you can see shifts in $\hat{\theta}$ among them. Those shifts are due to the differences in other covariates. </span>
<span id="cb25-361"><a href="#cb25-361" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-364"><a href="#cb25-364" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-365"><a href="#cb25-365" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: Estimated treatment effect for the first three individuals at different values of the age variable</span></span>
<span id="cb25-366"><a href="#cb25-366" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-ind-3</span></span>
<span id="cb25-367"><a href="#cb25-367" aria-hidden="true" tabindex="-1"></a>plot_data <span class="ot">&lt;-</span> <span class="fu">cbind</span>(data_te, theta_hat_with_se)</span>
<span id="cb25-368"><a href="#cb25-368" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-369"><a href="#cb25-369" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(plot_data[id <span class="sc">%in%</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, ]) <span class="sc">+</span></span>
<span id="cb25-370"><a href="#cb25-370" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">y =</span> theta_hat, <span class="at">x =</span> age)) <span class="sc">+</span></span>
<span id="cb25-371"><a href="#cb25-371" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_ribbon</span>(</span>
<span id="cb25-372"><a href="#cb25-372" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(</span>
<span id="cb25-373"><a href="#cb25-373" aria-hidden="true" tabindex="-1"></a>      <span class="at">ymin =</span> theta_hat <span class="sc">-</span> <span class="fl">1.96</span> <span class="sc">*</span> se, </span>
<span id="cb25-374"><a href="#cb25-374" aria-hidden="true" tabindex="-1"></a>      <span class="at">ymax =</span> theta_hat <span class="sc">+</span> <span class="fl">1.96</span> <span class="sc">*</span> se, </span>
<span id="cb25-375"><a href="#cb25-375" aria-hidden="true" tabindex="-1"></a>      <span class="at">x =</span> age</span>
<span id="cb25-376"><a href="#cb25-376" aria-hidden="true" tabindex="-1"></a>    ),</span>
<span id="cb25-377"><a href="#cb25-377" aria-hidden="true" tabindex="-1"></a>    <span class="at">fill =</span> <span class="st">"blue"</span>,</span>
<span id="cb25-378"><a href="#cb25-378" aria-hidden="true" tabindex="-1"></a>    <span class="at">alpha =</span> <span class="fl">0.4</span></span>
<span id="cb25-379"><a href="#cb25-379" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb25-380"><a href="#cb25-380" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(. <span class="sc">~</span> id) <span class="sc">+</span></span>
<span id="cb25-381"><a href="#cb25-381" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span>
<span id="cb25-382"><a href="#cb25-382" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-383"><a href="#cb25-383" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-384"><a href="#cb25-384" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-385"><a href="#cb25-385" aria-hidden="true" tabindex="-1"></a>@fig-all-box shows the box-plot of treatment effects for all the individuals. Note that variations observed at each <span class="in">`age`</span> value is due to heterogeneity in treatment effect driven by covariates other than <span class="in">`age`</span>. It looks like the three individuals looked at are exceptions. For the majority of individuals, the estimated treatment effects are negative at any value of <span class="in">`age`</span>.</span>
<span id="cb25-386"><a href="#cb25-386" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-389"><a href="#cb25-389" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb25-390"><a href="#cb25-390" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: Box-plot of the estimated treatment effects from all the individuals at different values of the age variable </span></span>
<span id="cb25-391"><a href="#cb25-391" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-all-box</span></span>
<span id="cb25-392"><a href="#cb25-392" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-393"><a href="#cb25-393" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(plot_data) <span class="sc">+</span></span>
<span id="cb25-394"><a href="#cb25-394" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="at">y =</span> theta_hat, <span class="at">x =</span> <span class="fu">factor</span>(<span class="fu">round</span>(age, <span class="at">digits =</span> <span class="dv">2</span>)))) <span class="sc">+</span></span>
<span id="cb25-395"><a href="#cb25-395" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb25-396"><a href="#cb25-396" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Age"</span>) <span class="sc">+</span></span>
<span id="cb25-397"><a href="#cb25-397" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Estimated treatment effect"</span>) <span class="sc">+</span></span>
<span id="cb25-398"><a href="#cb25-398" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb25-399"><a href="#cb25-399" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb25-400"><a href="#cb25-400" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>)</span>
<span id="cb25-401"><a href="#cb25-401" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb25-402"><a href="#cb25-402" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-403"><a href="#cb25-403" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb25-404"><a href="#cb25-404" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-405"><a href="#cb25-405" aria-hidden="true" tabindex="-1"></a>You can easily repeat this analysis for other covariates to see their impacts as well.</span>
<span id="cb25-406"><a href="#cb25-406" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-407"><a href="#cb25-407" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-408"><a href="#cb25-408" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-409"><a href="#cb25-409" aria-hidden="true" tabindex="-1"></a><span class="fu">## Orthogonal Random Forest</span></span>
<span id="cb25-410"><a href="#cb25-410" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-411"><a href="#cb25-411" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-412"><a href="#cb25-412" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-413"><a href="#cb25-413" aria-hidden="true" tabindex="-1"></a><span class="fu">## References {.unnumbered}</span></span>
<span id="cb25-414"><a href="#cb25-414" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-415"><a href="#cb25-415" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;div</span> <span class="er">id</span><span class="ot">=</span><span class="st">"refs"</span><span class="kw">&gt;&lt;/div&gt;</span></span>
<span id="cb25-416"><a href="#cb25-416" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-417"><a href="#cb25-417" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-418"><a href="#cb25-418" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-419"><a href="#cb25-419" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-420"><a href="#cb25-420" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-421"><a href="#cb25-421" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-422"><a href="#cb25-422" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-423"><a href="#cb25-423" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-424"><a href="#cb25-424" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-425"><a href="#cb25-425" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-426"><a href="#cb25-426" aria-hidden="true" tabindex="-1"></a></span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->



</body></html>