<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-0.9.629">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Introduction to Machine Learning for Economists (Under Construction) - 17&nbsp; Generalized Random Forest</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./PROG-R-01-mlr3.html" rel="next">
<link href="./E01-spatial-cv.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
<link href="style.css" rel="stylesheet" type="text/css">

<!-- Global site tag (gtag.js) - Google Analytics -->
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-45VKT2HZSL"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag() { dataLayer.push(arguments); }
  gtag('js', new Date());

  gtag('config', 'G-45VKT2HZSL');
</script>
<script src="site_libs/htmlwidgets-1.5.4/htmlwidgets.js"></script>
<script src="site_libs/viz-1.8.2/viz.js"></script>
<link href="site_libs/DiagrammeR-styles-0.2/styles.css" rel="stylesheet">
<script src="site_libs/grViz-binding-1.0.9/grViz.js"></script>

  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="nav-sidebar floating slimcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Generalized Random Forest</span></h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Introduction to Machine Learning for Economists (Under Construction)</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/tmieno2/ML-Economist" title="Source Code" class="sidebar-tool px-1"><i class="bi bi-github"></i></a>
    <a href="" title="Share" id="sidebar-tool-dropdown-0" class="sidebar-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi bi-share"></i></a>
    <ul class="dropdown-menu" aria-labelledby="sidebar-tool-dropdown-0">
        <li>
          <a class="dropdown-item sidebar-tools-main-item" href="https://twitter.com/intent/tweet?url=|url|">
            <i class="bi bi-bi-twitter pe-1"></i>
          Twitter
          </a>
        </li>
        <li>
          <a class="dropdown-item sidebar-tools-main-item" href="https://www.facebook.com/sharer/sharer.php?u=|url|">
            <i class="bi bi-bi-facebook pe-1"></i>
          Facebook
          </a>
        </li>
        <li>
          <a class="dropdown-item sidebar-tools-main-item" href="https://www.linkedin.com/sharing/share-offsite/?url=|url|">
            <i class="bi bi-bi-linkedin pe-1"></i>
          LinkedIn
          </a>
        </li>
    </ul>
</div>
    </div>
      </div>
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">Welcome</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./H00-preface.html" class="sidebar-item-text sidebar-link">Preface</a>
  </div>
</li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
    <div class="sidebar-item-container"> 
        <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">Basics</a>
      <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
        <i class="bi bi-chevron-right ms-2"></i>
      </a>
    </div>
    <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./B01-nonlinear.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Non-linear function estimation</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./B02-bias-variance-tradeoff.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Bias-variance Trade-off</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./B03-cross-validation.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Cross-validation</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./B04-regularization.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Regression Shrinkage Methods</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./B05-bootstrap.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Bootstrap</span></a>
  </div>
</li>
    </ul>
  </li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
    <div class="sidebar-item-container"> 
        <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">Prediction ML</a>
      <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
        <i class="bi bi-chevron-right ms-2"></i>
      </a>
    </div>
    <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./P01-random-forest.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Random Forest</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./P02-boosted-regression-forest.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Boosted Regression Forest</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./P03-xgb.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Extreme Gradient Boosting</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./P04-local-linear-forest.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Local Linear Forest</span></a>
  </div>
</li>
    </ul>
  </li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
    <div class="sidebar-item-container"> 
        <a href="./C0P-causal-ml.html" class="sidebar-item-text sidebar-link">Causal Machine Learning (CML) Methods</a>
      <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
        <i class="bi bi-chevron-right ms-2"></i>
      </a>
    </div>
    <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./C00-why-not-this.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Why can’t we just do this?</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./C01-dml.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Double Machine Learning</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./C02-xstr-learner.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">S-, X-, T-, and R-learner</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./C03-cf-orf.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Forest-based CATE Estimators</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./C04-cf-extension.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Extensions of Causal Forest</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./C05-causal-model-selection.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Causal Model Selection</span></a>
  </div>
</li>
    </ul>
  </li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
    <div class="sidebar-item-container"> 
        <a href="./E0P-extensions.html" class="sidebar-item-text sidebar-link">Extensions</a>
      <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
        <i class="bi bi-chevron-right ms-2"></i>
      </a>
    </div>
    <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./E01-spatial-cv.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Spatial Cross-validation</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./E02-grf.html" class="sidebar-item-text sidebar-link active"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Generalized Random Forest</span></a>
  </div>
</li>
    </ul>
  </li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
    <div class="sidebar-item-container"> 
        <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">Programming: R</a>
      <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">
        <i class="bi bi-chevron-right ms-2"></i>
      </a>
    </div>
    <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./PROG-R-01-mlr3.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Machine Learning with <code>mlr3</code></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./PROG-R-02-reticulate.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Running Python from R</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./PROG-R-03-model-selection-prediction.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Model Selection (Prediction)</span></a>
  </div>
</li>
    </ul>
  </li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
    <div class="sidebar-item-container"> 
        <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true">Programming: Python</a>
      <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true">
        <i class="bi bi-chevron-right ms-2"></i>
      </a>
    </div>
    <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./PROG-P-01-scikitlearn.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Prediction</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./PROG-P-02-CATE-econml.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">22</span>&nbsp; <span class="chapter-title">Treatment Effect Estimation</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./PROG-P-03-model-selection.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">23</span>&nbsp; <span class="chapter-title">Model selection</span></a>
  </div>
</li>
    </ul>
  </li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
    <div class="sidebar-item-container"> 
        <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" aria-expanded="true">Appendices</a>
      <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" aria-expanded="true">
        <i class="bi bi-chevron-right ms-2"></i>
      </a>
    </div>
    <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth1 show">  
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./A01-mc-simulation.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Monte Carlo (MC) Simulation</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./A02-method-of-moment.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">Primer on method of moment</span></a>
  </div>
</li>
    </ul>
  </li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#random-forest-as-a-local-constant-regression" id="toc-random-forest-as-a-local-constant-regression" class="nav-link active" data-scroll-target="#random-forest-as-a-local-constant-regression"> <span class="header-section-number">17.1</span> Random forest as a local constant regression</a></li>
  <li><a href="#grf" id="toc-grf" class="nav-link" data-scroll-target="#grf"> <span class="header-section-number">17.2</span> GRF</a></li>
  <li><a href="#examples-of-grf" id="toc-examples-of-grf" class="nav-link" data-scroll-target="#examples-of-grf"> <span class="header-section-number">17.3</span> Examples of GRF</a>
  <ul class="collapse">
  <li><a href="#random-forest-as-a-grf" id="toc-random-forest-as-a-grf" class="nav-link" data-scroll-target="#random-forest-as-a-grf"> <span class="header-section-number">17.3.1</span> Random forest as a GRF</a></li>
  <li><a href="#causal-forest-as-a-grf" id="toc-causal-forest-as-a-grf" class="nav-link" data-scroll-target="#causal-forest-as-a-grf"> <span class="header-section-number">17.3.2</span> Causal forest as a GRF</a></li>
  </ul></li>
  <li><a href="#sec-grf-honest" id="toc-sec-grf-honest" class="nav-link" data-scroll-target="#sec-grf-honest"> <span class="header-section-number">17.4</span> Honesty</a></li>
  <li><a href="#sec-understand-grf-example" id="toc-sec-understand-grf-example" class="nav-link" data-scroll-target="#sec-understand-grf-example"> <span class="header-section-number">17.5</span> Understanding GRF better by example</a>
  <ul class="collapse">
  <li><a href="#orthogonalization-of-y-and-t" id="toc-orthogonalization-of-y-and-t" class="nav-link" data-scroll-target="#orthogonalization-of-y-and-t"> <span class="header-section-number">17.5.1</span> Orthogonalization of <span class="math inline">\(Y\)</span> and <span class="math inline">\(T\)</span></a></li>
  <li><a href="#sec-build-trees" id="toc-sec-build-trees" class="nav-link" data-scroll-target="#sec-build-trees"> <span class="header-section-number">17.5.2</span> Building trees</a></li>
  <li><a href="#prediction" id="toc-prediction" class="nav-link" data-scroll-target="#prediction"> <span class="header-section-number">17.5.3</span> Prediction</a></li>
  </ul></li>
  <li><a href="#clustering" id="toc-clustering" class="nav-link" data-scroll-target="#clustering"> <span class="header-section-number">17.6</span> Clustering</a></li>
  <li><a href="#hyper-parameter-tuning" id="toc-hyper-parameter-tuning" class="nav-link" data-scroll-target="#hyper-parameter-tuning"> <span class="header-section-number">17.7</span> Hyper-parameter tuning</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/tmieno2/ML-Economist/edit/master/E02-grf.qmd" class="toc-action">Edit this page</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title"><span id="sec-grf" class="quarto-section-identifier d-none d-lg-block"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Generalized Random Forest</span></span></h1><button type="button" class="btn code-tools-button dropdown-toggle" id="quarto-code-tools-menu" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi"></i> Code</button><ul class="dropdown-menu dropdown-menu-end" aria-labelelledby="quarto-code-tools-menu"><li><a id="quarto-show-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Show All Code</a></li><li><a id="quarto-hide-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Hide All Code</a></li><li><hr class="dropdown-divider"></li><li><a id="quarto-view-source" class="dropdown-item" href="javascript:void(0)" role="button">View Source</a></li></ul></div></div>
</div>



<div class="quarto-title-meta">

    
    
  </div>
  

</header>

<div class="callout-important callout callout-style-default no-icon callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
What you will learn
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>What GRF is</li>
<li>How GRF was motivated</li>
<li>Random forest is a special case of GRF</li>
<li>Honesty rule</li>
</ul>
</div>
</div>
<div class="callout-tip callout callout-style-default no-icon callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Background knowledge
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>(necessary) how a regression tree is built (<a href="P01-random-forest.html#sec-rt"><span>Section&nbsp;6.1</span></a>)</li>
<li>(necessary) random forest (<a href="P01-random-forest.html#sec-rf"><span>Section&nbsp;6.2</span></a>)</li>
<li>(necessary) local regression (<a href="B01-nonlinear.html#sec-local"><span>Section&nbsp;1.3</span></a>)</li>
</ul>
</div>
</div>
<div class="callout-note callout callout-style-default no-icon callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Packages to load for replication
</div>
</div>
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ranger)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(parallel)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(grf)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rpart)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rattle)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(wooldridge)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<section id="random-forest-as-a-local-constant-regression" class="level2" data-number="17.1">
<h2 data-number="17.1" class="anchored" data-anchor-id="random-forest-as-a-local-constant-regression"><span class="header-section-number">17.1</span> Random forest as a local constant regression</h2>
<p>Suppose you are interested in estimating <span class="math inline">\(E[y|X]\)</span> using a dataset and you have trained a random forest model with <span class="math inline">\(T\)</span> tress. Now, let <span class="math inline">\(\eta_{i,t}(X)\)</span> takes <span class="math inline">\(1\)</span> if observation <span class="math inline">\(i\)</span> belongs to the same leaf as <span class="math inline">\(X\)</span> in tree <span class="math inline">\(t\)</span>, where <span class="math inline">\(X\)</span> is a vector of covariates (<span class="math inline">\(K\)</span> variables). Then, the RF’s predicted value of <span class="math inline">\(y\)</span> conditional on a particular value of <span class="math inline">\(X\)</span> (say, <span class="math inline">\(X_0\)</span>) can be written as follows:</p>
<p><span class="math display">\[
\begin{aligned}
\hat{y}(X_0) = \frac{1}{T} \cdot\sum_{t=1}^T\sum_{i=1}^N \frac{\eta_{i,t}(X_0)}{\sum_{i=1}^N\eta_{i,t}(X_0)}\cdot y_i
\end{aligned}
\]</span></p>
<p>Note that <span class="math inline">\(\sum_{i=1}^N\eta_{i,t}(X_0)\)</span> represents the number of observations in the same leaf as <span class="math inline">\(X_0\)</span>. Therefore, <span class="math inline">\(\sum_{i=1}^N \frac{\eta_{i,t}(X_0)}{\sum_{i=1}^N\eta_{i,t}(X_0)}\cdot y_i\)</span> is the average value of <span class="math inline">\(y\)</span> of the leaf that <span class="math inline">\(X_0\)</span> belongs to. So, while looking slightly complicated, it is the average value of <span class="math inline">\(y\)</span> of the tree <span class="math inline">\(X_0\)</span> belongs to averaged across the trees, which we know is how RF predicts <span class="math inline">\(y\)</span> at <span class="math inline">\(X_0\)</span>.</p>
<p>We can switch the summations like this,</p>
<p><span class="math display">\[
\begin{aligned}
\hat{y}(X_0) = \sum_{i=1}^N \cdot\frac{1}{T}\sum_{t=1}^T\frac{\eta_{i,t}(X_0)}{\sum_{i=1}^N\eta_{i,t}(X_0)}\cdot y_i
\end{aligned}
\]</span></p>
<p>Let <span class="math inline">\(\alpha(X_i, X_0)\)</span> denote <span class="math inline">\(\frac{1}{T}\sum_{t=1}^T\frac{\eta_{i,t}(X_0)}{\sum_{i=1}^N\eta_{i,t}(X_0)}\)</span>. Then, we can rewrite the above equation as</p>
<p><span class="math display">\[
\begin{aligned}
\hat{y}(X_0) = \sum_{i=1}^N \alpha(X_i,X_0) \cdot y_i
\end{aligned}
\]</span></p>
<p>Now, it is easy to show that <span class="math inline">\(\hat{y}(X_0)\)</span> is a solution to the following minimization problem.</p>
<p><span id="eq-ll-constant"><span class="math display">\[
\begin{aligned}
Min_{\theta} \sum_{i=1}^N \alpha(X_i,X_0)\cdot[y_i -\theta]^2
\end{aligned}
\tag{17.1}\]</span></span></p>
<p>In this formulation of the problem, <span class="math inline">\(\alpha(X_i,X_0)\)</span> can be considered the weight given to observation <span class="math inline">\(i\)</span>.</p>
<p>By definition,</p>
<ul>
<li><span class="math inline">\(0 \leq \alpha(X_i,X_0) \leq 1\)</span></li>
<li><span class="math inline">\(\sum_{i=1}^N \alpha(X_i,X_0) = 1\)</span></li>
</ul>
<p>You may notice that <a href="#eq-ll-constant">Equation&nbsp;<span>17.1</span></a> is actually a special case of <span style="color:blue">local constant regression</span> where the individual weights are <span class="math inline">\(\alpha(X_i, X_0)\)</span>. Roughly speaking, <span class="math inline">\(\alpha(X_i, X_0)\)</span> measures how often observation <span class="math inline">\(i\)</span> share the same leaves as the evaluation point (<span class="math inline">\(X_0\)</span>) across <span class="math inline">\(T\)</span> trees. So, it measures how similar <span class="math inline">\(X_i\)</span> is to <span class="math inline">\(X_0\)</span> in the RF way, but not based on euclidean distance (which is subject to curse of dimensionality). So, RF is actually a local constant regression with a special way of distributing weights to the individual observations.</p>
</section>
<section id="grf" class="level2 page-columns page-full" data-number="17.2">
<h2 data-number="17.2" class="anchored" data-anchor-id="grf"><span class="header-section-number">17.2</span> GRF</h2>
<p>Interpretation of RF as a local regression led <span class="citation" data-cites="athey2019generalized">Athey, Tibshirani, and Wager (<a href="#ref-athey2019generalized" role="doc-biblioref">2019</a>)</span> to conceive GRF, under which various statistics (e.g., conditional average treatment effect, conditional quantile) can be estimated under the unified framework.</p>
<p>You can consider prediction of <span class="math inline">\(y\)</span> at <span class="math inline">\(X = X_0\)</span> using RF as a three-step process.</p>
<div class="callout-note callout callout-style-default no-icon callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Predicting <span class="math inline">\(y\)</span> at <span class="math inline">\(X=X_0\)</span> using RF
</div>
</div>
<div class="callout-body-container callout-body">
<ol type="1">
<li>Grow trees (find splitting rules for each tree)</li>
<li>For <span class="math inline">\(X = X_0\)</span>, find the weights for individual observations based on the trees grown<br>
</li>
<li>Solve <a href="#eq-ll-constant">Equation&nbsp;<span>17.1</span></a> based on the weights</li>
</ol>
</div>
</div>
<p>Step 1 is done only once. Every time you make a prediction at a different value of <span class="math inline">\(X\)</span>, you go over steps 2 and 3.</p>
<p>GRF follows exactly the same steps except that it <span style="color:red">adjusts the way trees are grown </span>in step 1 and <span style="color:red">adjusts the way the local minimization problem</span> is solved at step 3 (they are actually interrelated) depending on what you would like to estimate.</p>
<p>Here, we follow the notations used in <span class="citation" data-cites="athey2019generalized">Athey, Tibshirani, and Wager (<a href="#ref-athey2019generalized" role="doc-biblioref">2019</a>)</span> as much as possible. Here are the list of notations:</p>
<ul>
<li><span class="math inline">\(O_i\)</span>: data for observation <span class="math inline">\(i\)</span></li>
<li><span class="math inline">\(\theta\)</span>: the statistics of interest (e.g., CATE for causal forest, conditional quantile for for quantile forest)</li>
<li><span class="math inline">\(\nu\)</span>: nuisance statistics (you are not interested in estimating this).</li>
<li><span class="math inline">\(\Psi_{\theta, \nu}(O)\)</span>: score function</li>
<li><span class="math inline">\(\alpha_i(x)\)</span>: weight given to observation <span class="math inline">\(i\)</span> when predicting at <span class="math inline">\(X=x\)</span>.</li>
</ul>
<p>In general, GRF solves the following problem to find the estimate of <span class="math inline">\(\theta\)</span> conditional on <span class="math inline">\(X_i= x\)</span>:</p>
<p><span id="eq-opt"><span class="math display">\[
\begin{aligned}
\sum_{i=1}^n \alpha_i(x)\Psi_{\theta, \nu}(O_i) = 0
\end{aligned}
\tag{17.2}\]</span></span></p>
<p>As we saw earlier, GRF is RF when <span class="math inline">\(\Psi_{\theta, \nu}(O_i) = Y_i-\theta\)</span>. There is no nuisance statistics, <span class="math inline">\(\nu(x)\)</span>, in the RF case. By changing how the score function (<span class="math inline">\(\Psi_{\theta, \nu}(O_i)\)</span>) is defined, you can estimate <span style="color:blue"> a wide range of statistics using different approaches </span>under the <span style="color:blue"> same</span> estimation framework (this is why it is called <span style="color:red">generalized </span>random forest). Here are some of the statistics and estimation approaches that are under the GRF framework.</p>
<ul>
<li>Conditional expectation (<span class="math inline">\(E[Y|X]\)</span>)
<ul>
<li>Regression forest (Random forest for regression)</li>
<li>Local linear forest</li>
<li>Boosted regression forest</li>
</ul></li>
<li>Conditional average treatment effect (CATE)
<ul>
<li>Causal forest</li>
<li>Instrumental forest</li>
</ul></li>
<li>Conditional quantile
<ul>
<li>Quantile forest</li>
</ul></li>
</ul>
<div class="callout-note callout callout-style-default no-icon callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Score function examples
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li><p>Random forest: <span class="math inline">\(\Psi_{\theta, \nu}(O_i) = Y_i - \theta\)</span></p></li>
<li><p>Causal forest: <span class="math inline">\(\Psi_{\theta, \nu}(O_i) = (\tilde{Y}_i- \theta\tilde{T}_i)\cdot\tilde{T}_i\)</span>, where <span class="math inline">\(\tilde{v_i} = v_i - E[v_i|X_i]\)</span></p></li>
<li><p>Quantile forest: <span class="math inline">\(\Psi_{\theta, \nu}(O_i) = qI\{Y_i &gt; \theta\} - (1-q)I\{Y_i \leq \theta\}\)</span></p></li>
</ul>
</div>
</div>

<div class="no-row-height column-margin column-container"><div class="">
<p><span class="math inline">\(I\{\}\)</span> is an indicator function that takes 1 if the condition inside the curly brackets and 0 otherwise.</p>
</div></div><hr>
<p>So far, we have only talked about score functions so far, but not the weights. Do all the approaches listed above use the weights derived from the trees grown by the traditional RF in solving <a href="#eq-opt">Equation&nbsp;<span>17.2</span></a>? You could (you are free to use any weights), but that would not be wise. As mentioned earlier, GRF adjusts the way trees are grown (thus weights) as well according to the score function so that weights are optimized for your objective. This makes sense. Right <span style="color:blue"> neighbors </span>should be different based on what you are interesting in estimating.</p>
<p>Specifically, GRF uses the random forest <span style="color:blue">algorithm</span> to grow trees based on <span style="color:blue"> pseudo outcome (<span class="math inline">\(\rho_i\)</span>) </span> derived from the score function that is specific to the type of regression you are running, but not on <span class="math inline">\(Y\)</span>. Basically, you are using exactly the same algorithm as the traditional RF we saw in <a href="P01-random-forest.html#sec-rf"><span>Section&nbsp;6.2</span></a> except that <span class="math inline">\(Y\)</span> is swapped with the pseudo outcome.</p>

<div class="no-row-height column-margin column-container"><div class="">
<p>See <span class="citation" data-cites="athey2019generalized">Athey, Tibshirani, and Wager (<a href="#ref-athey2019generalized" role="doc-biblioref">2019</a>)</span> for how the pseudo outcome is derived from a score function in general.</p>
</div></div><div class="callout-note callout callout-style-default no-icon callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Example pseudo outcomes
</div>
</div>
<div class="callout-body-container callout-body">
<p><span style="color:blue">Random forest</span>:</p>
<p><span class="math display">\[
\begin{aligned}
\rho_i = Y_i - \hat{\theta}_P
\end{aligned}
\]</span></p>
<p><span style="color:blue"> Causal forest</span>:</p>
<p><span class="math display">\[
\begin{aligned}
\rho_i = (\tilde{Y}_i- \hat{\theta}_P \tilde{T}_i)\cdot\tilde{T}_i
\end{aligned}
\]</span></p>
<p><span style="color:blue"> Quantile forest</span>:</p>
<p><span class="math display">\[
\begin{aligned}
\rho_i = I\{Y_i &gt; \hat{\theta}_P\}
\end{aligned}
\]</span></p>
</div>
</div>
<p>Note that <span class="math inline">\(\hat{\theta}_P\)</span> in the pseudo outcomes presented above is the solution to <a href="#eq-opt">Equation&nbsp;<span>17.2</span></a> with their respective score functions using the data in the parent node. For example, <span class="math inline">\(\hat{\theta}_P = \bar{Y}_p\)</span> for RF, which is the average value of <span class="math inline">\(Y\)</span> in the parent node. In quantile forest, <span class="math inline">\(\hat{\theta}_P\)</span> is the <span class="math inline">\(q\)</span>th quantile of the parent node if you are estimating the <span class="math inline">\(q\)</span>the conditional quantile.</p>
<hr>
<p>Finally, here are the conceptual steps of GRF:</p>
<div class="callout-note callout callout-style-default no-icon callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
GRF: conceptual steps
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Training (Forest Growing) Step:
<ol type="1">
<li>Determine what statistics you are interested in (<span class="math inline">\(\theta(X)\)</span>, e.g., CATE, conditional quantile)</li>
<li>Define the appropriate score function according to the statistics of interest</li>
<li>Derive the pseudo outcome based on the score function</li>
<li>Grow trees using the RF algorithm based on the pseudo outcome</li>
</ol></li>
<li>Prediction Step:
<ol type="1">
<li><p>Determine what value of <span class="math inline">\(X\)</span> you would like to predict <span class="math inline">\(\theta(X)\)</span> (call it <span class="math inline">\(X_0\)</span>)</p></li>
<li><p>Find the individual weights <span class="math inline">\(\alpha(X_i, X_0)\)</span> according to <span class="math display">\[
\begin{aligned}
\alpha(X_i, X_0) = \frac{1}{T}\sum_{t=1}^T\frac{\eta_{i,t}(X_0)}{\sum_{i=1}^N\eta_{i,t}(X_0)}
\end{aligned}
\]</span> , which measures how often observation <span class="math inline">\(i\)</span> belongs to the same node as the evaluation point <span class="math inline">\(X_0\)</span>.</p></li>
<li><p>Solve <a href="#eq-opt">Equation&nbsp;<span>17.2</span></a> with the weights obtained above and the score function specified earlier</p></li>
</ol></li>
</ul>
</div>
</div>
<p>The training step is done only once (trees are built only once). But, whenever you predict <span class="math inline">\(\theta(X)\)</span> at different values of <span class="math inline">\(X\)</span>, you go through the prediction step.</p>

<div class="no-row-height column-margin column-container"><div class="">
<p>Orthogonal random forest (a forest-based heterogeneous treatment effect estimator like causal forest), on the other hand, build trees every time when predicting treatment effect <span class="math inline">\(\theta(X)\)</span> at particular values of <span class="math inline">\(X\)</span>, which is why orthogonal random forest takes a very long time especially when there are many evaluation points. Orthogonal random forest is covered in <a href="C03-cf-orf.html"><span>Chapter&nbsp;13</span></a>.</p>
</div></div></section>
<section id="examples-of-grf" class="level2" data-number="17.3">
<h2 data-number="17.3" class="anchored" data-anchor-id="examples-of-grf"><span class="header-section-number">17.3</span> Examples of GRF</h2>
<section id="random-forest-as-a-grf" class="level3" data-number="17.3.1">
<h3 data-number="17.3.1" class="anchored" data-anchor-id="random-forest-as-a-grf"><span class="header-section-number">17.3.1</span> Random forest as a GRF</h3>
<p>Here, we take a look at RF as a GRF as an illustration to understand the general GRF procedure better. When <span class="math inline">\(\Psi_{\theta, \nu}(Y_i)\)</span> is set to <span class="math inline">\(Y_i - \theta\)</span>, then GRF is equivalent to the traditional RF. By plugging <span class="math inline">\(Y_i - \theta\)</span> into <a href="#eq-opt">Equation&nbsp;<span>17.2</span></a>, the estimate of <span class="math inline">\(E[Y|X=X_0]\)</span>, <span class="math inline">\(\theta(X_0)\)</span>, is identified by solving</p>
<p><span id="eq-rf"><span class="math display">\[
\begin{aligned}
\sum_{i=1}^n \alpha_i(X_0)(Y_i - \theta) = 0
\end{aligned}
\tag{17.3}\]</span></span></p>
<p>The weights <span class="math inline">\(\alpha_i(X_i, X_0)\)</span> are defined as follows:</p>
<p><span class="math display">\[
\begin{aligned}
\alpha_i(X_i, X_0) = \frac{1}{T}\cdot\sum_{t=1}^T\frac{\eta_{i,t}(X_0)}{\sum_{i=1}^N\eta_{i,t}(X_0)}
\end{aligned}
\]</span></p>
<p>, where <span class="math inline">\(\eta_{i,t}(X_0)\)</span> is 1 if observation <span class="math inline">\(i\)</span> which has feature values <span class="math inline">\(X_i\)</span> belongs to the same leaf as <span class="math inline">\(X_0\)</span>.</p>
<hr>
<p><span style="color:blue"> Step 1: Grow trees</span>:</p>
<p>Now, let’s consider growing trees to find <span class="math inline">\(\alpha_i(X_0)\)</span> in <a href="#eq-rf">Equation&nbsp;<span>17.3</span></a>. For a given sample and set of variables randomly selected, GRF starts with solving the unweighted version of <a href="#eq-rf">Equation&nbsp;<span>17.3</span></a>.</p>
<p><span id="eq-rf-initial"><span class="math display">\[
\begin{aligned}
\sum_{i=1}^n Y_i - \theta = 0
\end{aligned}
\tag{17.4}\]</span></span></p>
<p>The solution to this problem is simply the mean of <span class="math inline">\(Y\)</span>, which will be denoted as <span class="math inline">\(\bar{Y}_P\)</span>, where <span class="math inline">\(P\)</span> represents the parent node. Here, the parent node include all the data points as this is the first split.</p>
<p>Then the pseudo outcome (<span class="math inline">\(\rho_i\)</span>) that is used in splitting is</p>
<p><span class="math display">\[
\begin{aligned}
\rho_i = Y_i - \bar{Y}_P
\end{aligned}
\]</span></p>
<p>Now, a standard CART regression split is applied on the pseudo outcomes. That is, the variable-threshold combination that maximizes the following criteria is found:</p>
<p><span id="eq-criteria"><span class="math display">\[
\begin{aligned}
\tilde{\Delta}(C_1, C_2) = \frac{(\sum_{i \in C_1} \rho_i)^2}{N_{C_1}} + \frac{(\sum_{i \in C_2} \rho_i)^2}{N_{C_2}}
\end{aligned}
\tag{17.5}\]</span></span></p>
<p>where <span class="math inline">\(C_1\)</span> and <span class="math inline">\(C_2\)</span> represent two child node candidates for a given split. Since <span class="math inline">\(\bar{Y}_P\)</span> is just a constant, it is equivalent to splitting on <span class="math inline">\(Y_i\)</span>. So, this is exactly the same as how the traditional RF builds trees (see <a href="P01-random-forest.html#sec-split-sim"><span>Section&nbsp;6.1.2</span></a> for the rationale behind maximizing the criteria presented in <a href="#eq-criteria">Equation&nbsp;<span>17.5</span></a>).</p>
<p>Once the best split is identified, each of the new child nodes is split following the exactly the same procedure. Splitting continues until one of the user-specified condition prevent a further splitting. Many trees from bootstrapped samples are created (just like the regular random forest) and they form a forest. This shows that the GRF with <span class="math inline">\(\Psi_{\theta, \nu}(Y_i, X_i) = Y_i - \theta\)</span> grows trees in the same manner as the traditional RF. RF in GRF is implemented by <code>regression_forest()</code>. But, note that running <code>ranger()</code> and <code>regression_forest()</code> will not result in the same forest because of the randomness involved in resampling and random selection of variables. Only their algorithms are equivalent</p>
<hr>
<p><span style="color:blue"> Step 2: Predict</span>:</p>
<p>To predict <span class="math inline">\(E[Y|X=X_0]\)</span>, <a href="#eq-rf">Equation&nbsp;<span>17.3</span></a> is solved</p>
<p><span class="math display">\[
\begin{aligned}
\sum_{i=1}^N \alpha_i(X_i, X_0)(Y_i-\theta) = 0
\end{aligned}
\]</span></p>
<p>So,</p>
<p><span class="math display">\[
\begin{aligned}
\theta(X_0) &amp; = \frac{\sum_{i=1}^N \alpha_i(X_0)Y_i}{\sum_{i=1}^N \alpha_i(X_0)}\\
&amp; = \sum_{i=1}^N \alpha_i(X_0)Y_i \;\; \mbox{(since } \sum_{i=1}^N \alpha_i(X_0) = 1\mbox{)} \\
&amp; = \sum_{i=1}^N \huge[\normalsize \frac{1}{T}\cdot\sum_{t=1}^T\frac{\eta_{i,t}(X_0)}{\sum_{i=1}^N\eta_{i,t}(X_0)}\cdot y_i\huge]\\
&amp; = \frac{1}{T}  \cdot\sum_{t=1}^T\sum_{i=1}^N \frac{\eta_{i,t}(X_0)}{\sum_{i=1}^N\eta_{i,t}(X_0)}\cdot y_i \;\; \mbox{(changing the order of the summations)} \\
&amp; = \frac{1}{T} \cdot\sum_{t=1}^T \bar{Y}_t
\end{aligned}
\]</span></p>
<p>So, <span class="math inline">\(\theta(X_0)\)</span> from GRF is the average of tree-specific predictions for <span class="math inline">\(X_0\)</span>, which is exactly how RF predicts <span class="math inline">\(E[Y|X=X_0]\)</span> as well.</p>
<p>So, it has been shown that GRF with <span class="math inline">\(\Psi_{\theta, \nu}(Y_i)=Y_i - \theta\)</span> grows trees in the same manner as the traditional RF and also that GRF predicts <span class="math inline">\(E[Y|X=X_0]\)</span> in the same manner as the traditional RF. So, RF is a special case of GRF, where <span class="math inline">\(\Psi_{\theta, \nu}(Y_i)=Y_i - \theta\)</span>.</p>
</section>
<section id="causal-forest-as-a-grf" class="level3" data-number="17.3.2">
<h3 data-number="17.3.2" class="anchored" data-anchor-id="causal-forest-as-a-grf"><span class="header-section-number">17.3.2</span> Causal forest as a GRF</h3>
<p>Causal forest (with a single treatment variable) as an R-learner is a GRF with</p>
<p><span class="math display">\[
\begin{aligned}
\Psi_{\theta, \nu}(O_i) = [(Y_i - E[Y|X_i])- \theta(X_i)(T_i - E[T|X_i])](T_i - E[T|X_i])
\end{aligned}
\]</span></p>
<p>In practice <span class="math inline">\(E[Y|X_i]\)</span> and <span class="math inline">\(E[T|X_i]\)</span> are first estimated using appropriate machine learning methods (e.g., lasso, random forest, gradient boosted forest) in a cross-fitting manner and then the estimation of <span class="math inline">\(Y_i - E[Y|X_i]\)</span> and <span class="math inline">\(T_i - E[T|X_i]\)</span> are constructed. Let’s denote them by <span class="math inline">\(\tilde{Y}_i\)</span> and <span class="math inline">\(\tilde{T}_i\)</span>. Then the score function is written as</p>
<p><span id="eq-cf-score"><span class="math display">\[
\begin{aligned}
\Psi_{\theta} = (\tilde{Y}_i- \theta\tilde{T}_i)\tilde{T}_i
\end{aligned}
\tag{17.6}\]</span></span></p>
<p>So, the conditional treatment effect at <span class="math inline">\(X=X_0\)</span>, <span class="math inline">\(\theta(X_0)\)</span>, is estimated by solving the following problem:</p>
<p><span id="eq-cf-solve"><span class="math display">\[
\begin{aligned}
\hat{\theta}(X_0) = \sum_{i=1}^N \alpha_i(x)(\tilde{Y_i} - \theta\cdot \tilde{T_i})\tilde{T_i} = 0
\end{aligned}
\tag{17.7}\]</span></span></p>
<p>where <span class="math inline">\(\alpha_i(x)\)</span> is the weight obtained from the trees built using random forest on the pseudo outcomes that are derived from the score function (<a href="#eq-cf-score">Equation&nbsp;<span>17.6</span></a>).</p>
<p>For a given parent node <span class="math inline">\(p\)</span>, the pseudo outcomes are defined as</p>
<p><span class="math display">\[
\begin{aligned}
\rho_i = (\tilde{Y}_i- \hat{\theta}_P\tilde{T}_i)\tilde{T}_i
\end{aligned}
\]</span></p>
<p>where <span class="math inline">\(\hat{\theta}_P\)</span> is the solution of the following problem using the observations in the parent node:</p>
<p><span class="math display">\[
\begin{aligned}
\sum_{i=1}^N \tilde{T_i}(\tilde{Y_i}-\theta \tilde{T_i}) = 0
\end{aligned}
\]</span></p>
<p>, which is the unweighted version of <a href="#eq-cf-solve">Equation&nbsp;<span>17.7</span></a>.</p>
<p>The standard CART splitting algorithm is applied on the pseudo outcomes to build trees.</p>
</section>
</section>
<section id="sec-grf-honest" class="level2" data-number="17.4">
<h2 data-number="17.4" class="anchored" data-anchor-id="sec-grf-honest"><span class="header-section-number">17.4</span> Honesty</h2>
<p>GRF applies <span style="color:blue"> honesty </span>when it trains forests. Specifically, when building a tree, the subsample for the tree randomly selected without replacement is first split into two groups: subsample for <span style="color:blue"> splitting</span> and <span style="color:blue">prediction</span>. Then, the a tree is trained on the subsample for splitting and then find the splitting rules using the standard CART algorithm. However, when predicting (say <span class="math inline">\(E[Y|X]\)</span> at <span class="math inline">\(X=x\)</span>), the value of <span class="math inline">\(Y\)</span> from the subsample for splitting are NOT used. Rather, only the splitting rules are taken from the trained tree and then they are applied to the subsample for prediction (<a href="#fig-honest">Figure&nbsp;<span>17.1</span></a> illustrates this process).</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>DiagrammeR<span class="sc">::</span><span class="fu">grViz</span>(</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="st">"</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="st">digraph {</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="st">  graph [ranksep = 0.2, fontsize = 4]</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="st">  node [shape = box]</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="st">    SS [label = 'Subsample for splitting']</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="st">    SP [label = 'Subsample for predicting']</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="st">    BD [label = 'Sample for the tree']</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="st">    TT [label = 'Trained tree']</span></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="st">    PV [label = 'Predicted value']</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a><span class="st">  edge [minlen = 2]</span></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="st">    BD-&gt;SP</span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a><span class="st">    BD-&gt;SS</span></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a><span class="st">    SS-&gt;TT</span></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a><span class="st">    SP-&gt;PV</span></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a><span class="st">    TT-&gt;SP [label='apply the splitting rules']</span></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a><span class="st">  { rank = same; SS; SP}</span></span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a><span class="st">  { rank = same; TT}</span></span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a><span class="st">  { rank = same; PV}</span></span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a><span class="st">}</span></span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a><span class="st">"</span></span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div id="fig-honest" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<div id="htmlwidget-4ba79d189d84acc867b8" style="width:100%;height:325px;" class="grViz html-widget"></div>
<script type="application/json" data-for="htmlwidget-4ba79d189d84acc867b8">{"x":{"diagram":"\ndigraph {\n  graph [ranksep = 0.2, fontsize = 4]\n  node [shape = box]\n    SS [label = \"Subsample for splitting\"]\n    SP [label = \"Subsample for predicting\"]\n    BD [label = \"Sample for the tree\"]\n    TT [label = \"Trained tree\"]\n    PV [label = \"Predicted value\"]\n  edge [minlen = 2]\n    BD->SP\n    BD->SS\n    SS->TT\n    SP->PV\n    TT->SP [label=\"apply the splitting rules\"]\n  { rank = same; SS; SP}\n  { rank = same; TT}\n  { rank = same; PV}\n}\n","config":{"engine":"dot","options":null}},"evals":[],"jsHooks":[]}</script>
<p></p><figcaption aria-hidden="true" class="figure-caption">Figure 17.1: Illustration of an honest tree</figcaption><p></p>
</figure>
</div>
</div>
<p>Let’s demonstrate this using a very simple regression tree with two terminal nodes using the <code>mlb</code> data from the <code>wooldridge</code> package.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(mlb1, <span class="at">package =</span> <span class="st">"wooldridge"</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>mlb1_dt <span class="ot">&lt;-</span> <span class="fu">data.table</span>(mlb1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We would like to train a RF using this data where the dependent variable is logged salary (<code>lsalary</code>). We will illustrate the honesty rule by working on building a single tree within the process of building a forest.</p>
<p>We first sample the data without replacement. Here, we draw half of the entire sample randomly without replacement following the default value of <code>0.5</code> for the <code>sample.fraction</code> parameter.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">358349</span>)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>num_obs <span class="ot">&lt;-</span> <span class="fu">nrow</span>(mlb1_dt)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>row_indices <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="fu">seq_len</span>(num_obs), num_obs <span class="sc">/</span> <span class="dv">2</span>, <span class="at">replace =</span> <span class="cn">FALSE</span>)</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>boot_mlb1_dt <span class="ot">&lt;-</span> mlb1_dt[row_indices, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>boot_mlb1_dt</code> is the data that are involved in this tree. The rest of the samples in <code>mlb1_dt</code> will have nothing to do with this tree whatsoever.</p>
<p>We now split the sampled data into two groups: for splitting and prediction.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>num_obs_sub <span class="ot">&lt;-</span> <span class="fu">nrow</span>(boot_mlb1_dt) </span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>rows_split <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="fu">seq_len</span>(num_obs_sub), num_obs_sub <span class="sc">/</span> <span class="dv">2</span>, <span class="at">replace =</span> <span class="cn">FALSE</span>)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="co">#=== data for splitting ===#</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>split_data <span class="ot">&lt;-</span> boot_mlb1_dt[rows_split, ]</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="co">#=== data for prediction ===#</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>eval_data <span class="ot">&lt;-</span> boot_mlb1_dt[<span class="sc">-</span>rows_split, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We then train a tree using the data for splitting (<code>split_data</code>):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co">#=== build a simple tree ===#</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>tree_trained <span class="ot">&lt;-</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rpart</span>(</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>    lsalary <span class="sc">~</span> hits <span class="sc">+</span> runsyr, </span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> split_data, </span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">control =</span> <span class="fu">rpart.control</span>(<span class="at">minsplit =</span> <span class="dv">60</span>)</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a><span class="fu">fancyRpartPlot</span>(tree_trained, <span class="at">digits =</span> <span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-tree-sub" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="E02-grf_files/figure-html/fig-tree-sub-1.png" class="img-fluid figure-img" width="672"></p>
<p></p><figcaption aria-hidden="true" class="figure-caption">Figure 17.2: A simple regression tree using the subsamples for splitting</figcaption><p></p>
</figure>
</div>
</div>
</div>
<p>So the splitting rule is <code>hits &lt; 338</code> as shown in <a href="#fig-tree-sub">Figure&nbsp;<span>17.2</span></a>. At the terminal nodes, you see the prediction of <code>lsalary</code>: <span class="math inline">\(12.41\)</span> for the left and <span class="math inline">\(14.36\)</span> for the right. These predictions are NOT honest. They are obtained from the observed values of <code>lsalary</code> within the node using the splitting data (the data the tree is trained for). Instead of using these prediction values, an honest prediction applied the splitting rules (<code>hits &lt; 338</code>) to the data reserved for prediction.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>honest_pred <span class="ot">&lt;-</span> eval_data[, <span class="fu">mean</span>(lsalary), <span class="at">by =</span> hits <span class="sc">&lt;</span> <span class="dv">338</span>]</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    hits       V1
1:  TRUE 12.51945
2: FALSE 14.09740</code></pre>
</div>
</div>
<p>So, instead of <span class="math inline">\(12.41\)</span> and <span class="math inline">\(14.36\)</span>, the predicted values from the honest tree are <span class="math inline">\(12.52\)</span> and <span class="math inline">\(14.1\)</span> for the left and right nodes, respectively. Trees are built in this manner many times to form a forest.</p>
<p>More generally, in GRF, honesty is applied by using the evaluation data to solve <a href="#eq-opt">Equation&nbsp;<span>17.2</span></a> based on the weight <span class="math inline">\(\alpha_i(x)\)</span> derived from the trained forest using the splitting data. Honesty is required for the GRF estimator to be consistent and asymptotically normal <span class="citation" data-cites="athey2019generalized">(<a href="#ref-athey2019generalized" role="doc-biblioref">Athey, Tibshirani, and Wager 2019</a>)</span>. However, the application of honesty can do more damage than help when the sample size is small.</p>
</section>
<section id="sec-understand-grf-example" class="level2 page-columns page-full" data-number="17.5">
<h2 data-number="17.5" class="anchored" data-anchor-id="sec-understand-grf-example"><span class="header-section-number">17.5</span> Understanding GRF better by example</h2>
<p>This section goes through CF model training and CATE prediction step by step with codes to enhance our understanding of how GRF (and CF) works. In the process, we also learn the role of many of the hyper-parameters that are common across models that fall under GRF. We also learn the role of hyper-parameter that are specific to CF as well.</p>
<p>While the process explained here is for CF, the vast majority of the process is exactly the same for other models under GRF. Here are some differences:</p>
<ul>
<li>Orthogonalization of the data at the beginning for CF</li>
<li>The definition of the pseudo outcome (but, the way they are used in the algorithm is identical once they are defined)</li>
<li>The role of <code>min.node.size</code>, <code>alpha</code>, and <code>imbalance.penalty</code> used for safeguarding against extreme splits</li>
</ul>

<div class="no-row-height column-margin column-container"><div class="">
<p>Obviously, the codes here are only for illustration and should not be used in practice.</p>
</div></div><p>We will use a synthetic data generated from the following DGP:</p>
<p><span class="math display">\[
\begin{aligned}
y = \theta(X)\cdot T + g(X) + \mu
\end{aligned}
\]</span></p>
<p>, where</p>
<p><span class="math display">\[
\begin{aligned}
\theta(X) &amp; = 2 + 3 \times x_1^2 - \sqrt{3 \times  x_2^3 + 1}\\
g(X) &amp; = 10 \cdot [log(x_2+1) + 2\cdot x_3\times x_2]
\end{aligned}
\]</span></p>
<p>We have 10 feature variables <span class="math inline">\(x_1, \dots, x_{10}\)</span> and only <span class="math inline">\(x_1\)</span> through <span class="math inline">\(x_3\)</span> play a role.</p>
<ul>
<li><span class="math inline">\(x_1, x_2 \sim U[0, 3]^2\)</span></li>
<li><span class="math inline">\(x_3 \sim N(1, 1)\)</span></li>
<li><span class="math inline">\(x_4, \dots, x_{10} \sim U[0, 1]^{7}\)</span></li>
</ul>
<p>The error term and the treatment variable is defined as follows.</p>
<ul>
<li><span class="math inline">\(\mu \sim N(0, 1)\)</span></li>
<li><span class="math inline">\(T \sim Ber(0.5)\)</span></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">73843</span>)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>N <span class="ot">&lt;-</span> <span class="dv">1000</span> <span class="co"># number of observations</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>x4_x10 <span class="ot">&lt;-</span> </span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">matrix</span>(<span class="fu">runif</span>(N <span class="sc">*</span> <span class="dv">7</span>), <span class="at">nrow =</span> N) <span class="sc">%&gt;%</span> </span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.table</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">setnames</span>(<span class="fu">names</span>(.), <span class="fu">paste0</span>(<span class="st">"x"</span>, <span class="dv">4</span><span class="sc">:</span><span class="dv">10</span>))</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> </span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.table</span>(</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">x1 =</span> <span class="dv">3</span> <span class="sc">*</span> <span class="fu">runif</span>(N),</span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">x2 =</span> <span class="dv">3</span> <span class="sc">*</span> <span class="fu">runif</span>(N),</span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">x3 =</span> <span class="fu">rnorm</span>(N, <span class="at">mean =</span> <span class="dv">1</span>, <span class="at">sd =</span> <span class="dv">1</span>),</span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">mu =</span> <span class="fu">rnorm</span>(N),</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">T =</span> <span class="fu">runif</span>(N) <span class="sc">&gt;</span> <span class="fl">0.5</span></span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span> </span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a>  <span class="co">#=== theta(X) ===#</span></span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a>  .[, <span class="at">theta_x :=</span> <span class="dv">2</span> <span class="sc">+</span> <span class="dv">3</span> <span class="sc">*</span> x1<span class="sc">^</span><span class="dv">2</span> <span class="sc">-</span> <span class="fu">sqrt</span>(<span class="dv">3</span><span class="sc">*</span> x2<span class="sc">^</span><span class="dv">3</span> <span class="sc">+</span> <span class="dv">2</span>)] <span class="sc">%&gt;%</span></span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a>  <span class="co">#=== g(X) ===#</span></span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a>  .[, <span class="at">g_x :=</span> <span class="dv">10</span> <span class="sc">*</span> (<span class="fu">log</span>(x2 <span class="sc">+</span> <span class="dv">1</span>) <span class="sc">+</span> <span class="dv">2</span> <span class="sc">*</span> x3 <span class="sc">*</span> x2)] <span class="sc">%&gt;%</span> </span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a>  .[, <span class="at">y :=</span> theta_x <span class="sc">*</span> T <span class="sc">+</span> g_x <span class="sc">+</span> mu] <span class="sc">%&gt;%</span> </span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a>  .[, .(y, T, x1, x2, x3)] <span class="sc">%&gt;%</span> </span>
<span id="cb9-24"><a href="#cb9-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cbind</span>(., x4_x10)</span>
<span id="cb9-25"><a href="#cb9-25" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               y     T        x1        x2          x3        x4        x5
   1:  31.239440 FALSE 2.6951250 1.1965331  0.96678638 0.7210003 0.6207226
   2:  10.144462 FALSE 0.7919764 0.2722983  1.34510173 0.6267043 0.8869069
   3:  48.939026  TRUE 2.7143095 0.9040185  1.19934789 0.5620508 0.5455276
   4:  36.405483  TRUE 1.0628966 1.0659064  1.20667390 0.3638505 0.3201202
   5:  11.893620 FALSE 2.8824810 1.8198348  0.03356061 0.2570614 0.1447951
  ---                                                                     
 996:  74.029985  TRUE 1.6485987 2.8588233  1.04216169 0.6835214 0.7266654
 997:  21.176748  TRUE 2.0208231 0.4248353  0.49803263 0.2851495 0.8076055
 998:   3.388729 FALSE 1.9203044 0.5051511 -0.02137629 0.9961128 0.8446825
 999:   9.459730 FALSE 1.7274390 0.3468632  0.85507892 0.9669933 0.4943592
1000: 109.282950 FALSE 1.7030476 2.1927368  2.21194227 0.1442201 0.6783044
              x6         x7         x8         x9       x10
   1: 0.78698773 0.88919853 0.25543604 0.05640156 0.3560356
   2: 0.44413365 0.27657781 0.08810472 0.20397022 0.7982468
   3: 0.45238337 0.41956478 0.69912161 0.05602666 0.5905471
   4: 0.71060284 0.11400965 0.43547480 0.57807247 0.5707400
   5: 0.46912302 0.31966840 0.90711440 0.15150039 0.9426457
  ---                                                      
 996: 0.05036928 0.18803515 0.40981100 0.04716956 0.6443493
 997: 0.14171969 0.08375323 0.78676094 0.27132948 0.6570423
 998: 0.36225572 0.47555166 0.40332714 0.28135063 0.9657502
 999: 0.43818445 0.59426511 0.53555370 0.91860942 0.3339725
1000: 0.23674030 0.25005706 0.05512990 0.24352242 0.7227397</code></pre>
</div>
</div>
<section id="orthogonalization-of-y-and-t" class="level3" data-number="17.5.1">
<h3 data-number="17.5.1" class="anchored" data-anchor-id="orthogonalization-of-y-and-t"><span class="header-section-number">17.5.1</span> Orthogonalization of <span class="math inline">\(Y\)</span> and <span class="math inline">\(T\)</span></h3>
<p>Causal forest (and instrumental forest) first orthogonalizes <span class="math inline">\(Y\)</span> and <span class="math inline">\(T\)</span>. This is not the case for the other GRF methods. By default, <code>causal_forest()</code> uses random forest to estimate <span class="math inline">\(E[Y|X]\)</span> and <span class="math inline">\(E[T|X]\)</span>, and then use out-of-bag predictions. Since the treatment assignment is randomized, we could use <span class="math inline">\(0.5\)</span> for as the estimate for <span class="math inline">\(E[T|X]\)</span> for all the observations. But, we will estimate both in this example.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co">#--------------------------</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="co"># E[Y|X]</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="co">#--------------------------</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="co">#=== train an RF on y ===#</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>y_rf_trained <span class="ot">&lt;-</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">regression_forest</span>(</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">X =</span> data[, .(x1, x2, x3)],</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">Y =</span> data[, y]</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a><span class="co">#=== OOB estimates of E[Y|X] ===#</span></span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>ey_x_hat <span class="ot">&lt;-</span> y_rf_trained<span class="sc">$</span>predictions</span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a><span class="co">#--------------------------</span></span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a><span class="co"># E[T|X]</span></span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a><span class="co">#--------------------------</span></span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a><span class="co">#=== train an RF on T ===#</span></span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true" tabindex="-1"></a>t_rf_trained <span class="ot">&lt;-</span></span>
<span id="cb11-19"><a href="#cb11-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">probability_forest</span>(</span>
<span id="cb11-20"><a href="#cb11-20" aria-hidden="true" tabindex="-1"></a>    <span class="at">X =</span> data[, .(x1, x2, x3)],</span>
<span id="cb11-21"><a href="#cb11-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">Y =</span> data[, <span class="fu">factor</span>(T)]</span>
<span id="cb11-22"><a href="#cb11-22" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb11-23"><a href="#cb11-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-24"><a href="#cb11-24" aria-hidden="true" tabindex="-1"></a><span class="co">#=== OOB estimates of E[T|X] ===#</span></span>
<span id="cb11-25"><a href="#cb11-25" aria-hidden="true" tabindex="-1"></a>et_x_hat <span class="ot">&lt;-</span> t_rf_trained<span class="sc">$</span>predictions[, <span class="dv">2</span>] <span class="co"># get the probability of T = 1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s now orthogonalize <span class="math inline">\(Y\)</span> and <span class="math inline">\(T\)</span>,</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>data[, <span class="st">`</span><span class="at">:=</span><span class="st">`</span>(<span class="at">y_tilde =</span> y <span class="sc">-</span> ey_x_hat, <span class="at">t_tilde =</span> T <span class="sc">-</span> et_x_hat)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here is what the data looks like:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          y     T        x1        x2         x3        x4        x5        x6
1: 31.23944 FALSE 2.6951250 1.1965331 0.96678638 0.7210003 0.6207226 0.7869877
2: 10.14446 FALSE 0.7919764 0.2722983 1.34510173 0.6267043 0.8869069 0.4441337
3: 48.93903  TRUE 2.7143095 0.9040185 1.19934789 0.5620508 0.5455276 0.4523834
4: 36.40548  TRUE 1.0628966 1.0659064 1.20667390 0.3638505 0.3201202 0.7106028
5: 11.89362 FALSE 2.8824810 1.8198348 0.03356061 0.2570614 0.1447951 0.4691230
6: 78.17796  TRUE 1.3823141 1.6858092 1.91198906 0.6044051 0.8637295 0.8509355
          x7         x8         x9       x10    y_tilde    t_tilde
1: 0.8891985 0.25543604 0.05640156 0.3560356 -5.6636071 -0.5296217
2: 0.2765778 0.08810472 0.20397022 0.7982468 -4.3210587 -0.4788832
3: 0.4195648 0.69912161 0.05602666 0.5905471 11.9768866  0.4703248
4: 0.1140096 0.43547480 0.57807247 0.5707400  0.2253922  0.5534958
5: 0.3196684 0.90711440 0.15150039 0.9426457 -8.4757974 -0.5011004
6: 0.7302472 0.91788817 0.76401947 0.9238519 -0.6532169  0.5113809</code></pre>
</div>
</div>
</section>
<section id="sec-build-trees" class="level3" data-number="17.5.2">
<h3 data-number="17.5.2" class="anchored" data-anchor-id="sec-build-trees"><span class="header-section-number">17.5.2</span> Building trees</h3>
<hr>
<p><span style="color:blue"> Preparing Data</span>:</p>
<p>When building a tree, GRF use sampling without replacement instead of sampling with replacement as a default for RF implemented by <code>ranger()</code>. <code>sample.fraction</code> parameter determines the fraction of the entire sample drawn for each tree. Let’s use the default value, which is <span class="math inline">\(0.5\)</span>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>data_for_the_first_tree <span class="ot">&lt;-</span> data[<span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span>N, <span class="fl">0.5</span> <span class="sc">*</span> N, <span class="at">replace =</span> <span class="cn">FALSE</span>), ] </span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>              y     T         x1        x2         x3         x4         x5
  1: 24.9025560  TRUE 2.33286846 0.2477951  0.8980258 0.50396757 0.08485642
  2: 34.4617113 FALSE 0.36784858 0.9011077  1.5373583 0.83508147 0.10087835
  3:  4.8131199 FALSE 0.35464284 0.1992028  0.9367359 0.84134117 0.24929184
  4:  9.7958485 FALSE 2.54230941 0.7135930  0.3197887 0.05861231 0.34530906
  5: -0.2399451 FALSE 0.88094406 0.1036436 -0.4209629 0.63817225 0.24831807
 ---                                                                       
496: 81.0907489 FALSE 0.03481615 2.9859855  1.1341612 0.93868895 0.10845090
497: 45.7373362  TRUE 1.11687411 2.1342024  0.8219353 0.18140031 0.52911682
498: 81.7731653  TRUE 0.50482529 1.3572611  2.7416683 0.18128877 0.12207408
499: -9.1361410 FALSE 2.68976646 2.2431148 -0.4614975 0.27532629 0.87256993
500: 21.7882929  TRUE 0.41004493 2.6174698  0.3060092 0.43608964 0.80218867
            x6        x7         x8         x9       x10    y_tilde    t_tilde
  1: 0.5330777 0.0207677 0.75254547 0.28037254 0.5548497  10.348401  0.5165125
  2: 0.6637159 0.8061779 0.34381605 0.74622506 0.3308037  -3.174286 -0.4424109
  3: 0.8667869 0.2351434 0.26094758 0.48312612 0.3273708  -4.899194 -0.4703968
  4: 0.7823649 0.8942680 0.73156533 0.55466979 0.6760430 -13.150179 -0.5710382
  5: 0.8369564 0.9980671 0.02740824 0.36841730 0.6603217  -2.098577 -0.5943349
 ---                                                                          
496: 0.6255471 0.4409065 0.73793156 0.16597397 0.2782975   6.647750 -0.4746825
497: 0.7703280 0.8028466 0.79829377 0.72936803 0.3953351  -1.998624  0.4751614
498: 0.1248504 0.7362142 0.28138516 0.45800679 0.1410164  -3.481113  0.5902025
499: 0.9890033 0.6309494 0.11897912 0.05518469 0.2093569  -7.215089 -0.5353184
500: 0.7045508 0.3854207 0.34308977 0.88430430 0.5601304  -5.074021  0.4222428</code></pre>
</div>
</div>
<p>By default, honest splitting is implemented (<code>honesty = TRUE</code>). The <code>honesty.fraction</code> parameter determines the fraction of the randomly sampled data (<code>data_for_the_first_tree</code>), which will be used for determining splitting decisions. We will use the default value of <code>honesty.fraction</code>, which is <span class="math inline">\(0.5\)</span>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co">#=== number of observations ===#</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>N_d <span class="ot">&lt;-</span> <span class="fu">nrow</span>(data_for_the_first_tree)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="co">#=== indices ===#</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>J1_index <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span>N_d, <span class="fl">0.5</span> <span class="sc">*</span> N_d, <span class="at">replace =</span> <span class="cn">FALSE</span>)</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a><span class="co">#=== data for determining splitting rules ===#</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>data_J1 <span class="ot">&lt;-</span> data_for_the_first_tree[J1_index, ]</span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             y     T        x1        x2          x3        x4         x5
  1: 32.168156  TRUE 1.3202547 1.1420641  0.91158029 0.1686512 0.79225674
  2: 63.700906 FALSE 0.5488392 1.5795742  1.65239047 0.3851233 0.64512684
  3: 41.849736  TRUE 2.7792941 0.8210709  0.79316984 0.9278186 0.49904726
  4: 30.063201 FALSE 0.9203382 1.2250174  0.87263820 0.6032529 0.54814557
  5: 74.092869  TRUE 1.3308917 1.7572783  1.75875073 0.1445248 0.19666780
 ---                                                                     
246: 59.044593 FALSE 2.9685008 1.6885064  1.49969895 0.5313985 0.15660748
247: 79.666353  TRUE 2.3577697 1.1953114  2.31067728 0.4455486 0.01321071
248: 37.203768 FALSE 2.1998783 1.6630233  0.78424732 0.1420179 0.95015816
249: 14.141558  TRUE 1.1786313 1.1804383  0.05553601 0.8050410 0.10626896
250:  5.156938 FALSE 0.5296012 1.1397422 -0.09301202 0.5284173 0.92249024
             x6        x7        x8          x9       x10     y_tilde
  1: 0.70056216 0.3013820 0.7349177 0.434629694 0.3931467   1.1037999
  2: 0.94440025 0.7271771 0.4302308 0.009344572 0.2357119  -3.6279591
  3: 0.02594909 0.2632990 0.7471883 0.913540913 0.4693372  12.4304959
  4: 0.79733568 0.1367518 0.9528761 0.445233749 0.8557955  -2.0274153
  5: 0.50274839 0.7840356 0.2542098 0.023763730 0.9715826  -3.1022041
 ---                                                                 
246: 0.52848998 0.7282229 0.5323203 0.261717018 0.2347180 -13.0882951
247: 0.64246528 0.4421114 0.8446926 0.637843209 0.5676528   4.2071069
248: 0.20126400 0.1563281 0.2404751 0.415536469 0.7778129  -3.7512625
249: 0.16361912 0.7145285 0.7101117 0.600831013 0.5476061  -0.2123989
250: 0.61208680 0.1649016 0.9978272 0.497071444 0.5043290  -1.0637976
        t_tilde
  1:  0.4440064
  2: -0.4874906
  3:  0.4410276
  4: -0.5482126
  5:  0.5162645
 ---           
246: -0.5135525
247:  0.5212281
248: -0.3979431
249:  0.4469103
250: -0.5387802</code></pre>
</div>
</div>
<p>As you can see, we are just using a quarter of the original dataset (<code>sample.fraction</code> <span class="math inline">\(\times\)</span> <code>honesty.fraction</code> = <span class="math inline">\(0.5 \times 0.5\)</span>). Here is the data for repopulating the tree once the tree is built.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co">#=== data for repopulate the tree ===#</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>data_J2 <span class="ot">&lt;-</span> data_for_the_first_tree[<span class="sc">-</span>J1_index, ]</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               y     T        x1        x2          x3         x4         x5
  1:  34.4617113 FALSE 0.3678486 0.9011077  1.53735835 0.83508147 0.10087835
  2:   9.7958485 FALSE 2.5423094 0.7135930  0.31978875 0.05861231 0.34530906
  3:  -0.2399451 FALSE 0.8809441 0.1036436 -0.42096291 0.63817225 0.24831807
  4:  33.8904668  TRUE 2.7083542 0.7223860  0.28685672 0.47576373 0.60371551
  5:  42.0187752 FALSE 1.6180228 1.3468851  1.20854200 0.98942551 0.32931981
 ---                                                                        
246: 112.6993157 FALSE 1.8652859 2.4059719  2.08180730 0.34569836 0.76473310
247:  -1.1378434  TRUE 0.4855630 1.9038798 -0.28249533 0.67252894 0.21756645
248:  48.9883794  TRUE 2.2455929 1.4029347  1.00599231 0.30995650 0.47503141
249:  21.5579803  TRUE 1.9840075 2.8772854  0.05766394 0.95584669 0.02832178
250:  45.7373362  TRUE 1.1168741 2.1342024  0.82193525 0.18140031 0.52911682
            x6        x7         x8        x9       x10    y_tilde    t_tilde
  1: 0.6637159 0.8061779 0.34381605 0.7462251 0.3308037  -3.174286 -0.4424109
  2: 0.7823649 0.8942680 0.73156533 0.5546698 0.6760430 -13.150179 -0.5710382
  3: 0.8369564 0.9980671 0.02740824 0.3684173 0.6603217  -2.098577 -0.5943349
  4: 0.4801798 0.4673316 0.63830962 0.5097753 0.5330969  10.822273  0.4342292
  5: 0.5914005 0.7734454 0.18005650 0.1522792 0.8920292  -5.334219 -0.5456630
 ---                                                                         
246: 0.3229140 0.7352488 0.41670762 0.7729425 0.9237995  -3.439394 -0.4548755
247: 0.8127880 0.2702545 0.58958613 0.2344333 0.7674877  -2.027607  0.4568964
248: 0.3251172 0.1759780 0.42293907 0.3705066 0.1681080   9.109288  0.5642331
249: 0.5679956 0.6444142 0.53908269 0.2941420 0.9354850   1.957145  0.5125393
250: 0.7703280 0.8028466 0.79829377 0.7293680 0.3953351  -1.998624  0.4751614</code></pre>
</div>
</div>
<hr>
<p><span style="color:blue"> Determining splitting rules</span>:</p>
<p>Now let’s find the splitting rules using <code>data_J1</code> (Remember, <code>data_J2</code> is not used).</p>
<p>We are at the root node to which all the observations belong. Let’s first calculate the pseudo outcome. Finding pseudo outcomes starts from solving the unweighted version of <a href="#eq-opt">Equation&nbsp;<span>17.2</span></a> using the samples in the parent node (root node here) in general. For CF, it is</p>
<p><span id="eq-theta-p"><span class="math display">\[
\begin{aligned}
\sum_{i=1}^n (\tilde{Y}_i - \theta\tilde{T}_i)\cdot \tilde{T}_i = 0
\end{aligned}
\tag{17.8}\]</span></span></p>
<p>because the score function for CF is <span class="math inline">\((\tilde{Y}_i - \theta\tilde{T}_i)\cdot \tilde{T}_i\)</span>. So,</p>
<p><span class="math display">\[
\begin{aligned}
\theta_P = \frac{\sum_{i=1}^n \tilde{Y}_i\cdot \tilde{T}_i}{\sum_{i=1}^n\tilde{T}_i\cdot \tilde{T}_i}
\end{aligned}
\]</span></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>theta_p <span class="ot">&lt;-</span> data_J1[, <span class="fu">sum</span>(y_tilde <span class="sc">*</span> t_tilde)<span class="sc">/</span><span class="fu">sum</span>(t_tilde <span class="sc">*</span> t_tilde)]</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 7.212476</code></pre>
</div>
</div>
<p>The pseudo outcome for CF is <span class="math inline">\((\tilde{Y}_i - \theta_P\tilde{T}_i)\cdot \tilde{T}_i\)</span>. Let’s define that in <code>data_J1</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>data_J1[, rho <span class="sc">:</span><span class="er">=</span> (y_tilde <span class="sc">-</span> theta_p <span class="sc">*</span> t_tilde) <span class="sc">*</span> t_tilde]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, the number of variables to use for splitting is determined by the <code>mtry</code> parameter. Its default is <code>min(ceiling(sqrt(ncol(X))+20), ncol(X))</code>, where <code>X</code> is the feature matrix. Here <code>ncol(X)</code> is 10. So, <code>mtry</code> is set to 10. Following this, we will use all three variables for splitting here. Let <span class="math inline">\(C_j\)</span> (<span class="math inline">\(j = 1, 2\)</span>) and <span class="math inline">\(N_j\)</span> denote the child node <span class="math inline">\(j\)</span> and the number of observations in child node <span class="math inline">\(j\)</span>, respectively. We will write a function that works on a single feature variable to find all the threshold values that would result in unique splits, and then return the following information:</p>
<ul>
<li><span class="math inline">\((\sum_{i\in C_j} \rho_i)^2\)</span></li>
<li><span class="math inline">\(N_j\)</span></li>
<li><span class="math inline">\(\sum_{i=1}^{N_j} (T_i - \bar{T})^2\)</span> (we will call this <code>info_size</code>)</li>
<li>the number of treated and control units</li>
</ul>
<p>The output allows us to calculate the heterogeneity score of the split defined as</p>
<p><span class="math display">\[
\begin{aligned}
\sum_{j=1}^2 \frac{(\sum_{i\in C_j} \rho_i)^2}{N_j}.
\end{aligned}
\]</span></p>
<p>It also allows us to eliminate some of the splits based on <code>mtry</code>, <code>alpha</code>, and <code>imbalance.penalty</code> parameters.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>get_ss_by_var <span class="ot">&lt;-</span> <span class="cf">function</span>(feature_var, outcome_var, parent_data)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>{</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>  temp_data <span class="ot">&lt;-</span> </span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">copy</span>(parent_data) <span class="sc">%&gt;%</span> </span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">setnames</span>(feature_var, <span class="st">"temp_var"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">setnames</span>(outcome_var, <span class="st">"outcome"</span>)  </span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>  <span class="co">#=== define a sequence of values of hruns ===#</span></span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>  thr_seq <span class="ot">&lt;-</span></span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>    temp_data[<span class="fu">order</span>(temp_var), <span class="fu">unique</span>(temp_var)] <span class="sc">%&gt;%</span></span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a>    <span class="co">#=== get the rolling mean ===#</span></span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">frollmean</span>(<span class="dv">2</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a>    .[<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a>  <span class="co">#=== get RSS ===#</span></span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a>  ss_value <span class="ot">&lt;-</span></span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">lapply</span>(</span>
<span id="cb24-18"><a href="#cb24-18" aria-hidden="true" tabindex="-1"></a>      thr_seq,</span>
<span id="cb24-19"><a href="#cb24-19" aria-hidden="true" tabindex="-1"></a>      <span class="cf">function</span>(x){</span>
<span id="cb24-20"><a href="#cb24-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-21"><a href="#cb24-21" aria-hidden="true" tabindex="-1"></a>        return_data <span class="ot">&lt;-</span> </span>
<span id="cb24-22"><a href="#cb24-22" aria-hidden="true" tabindex="-1"></a>          temp_data[, .(</span>
<span id="cb24-23"><a href="#cb24-23" aria-hidden="true" tabindex="-1"></a>            <span class="at">het_score =</span> <span class="fu">sum</span>(outcome)<span class="sc">^</span><span class="dv">2</span><span class="sc">/</span>.N, </span>
<span id="cb24-24"><a href="#cb24-24" aria-hidden="true" tabindex="-1"></a>            <span class="at">nobs =</span> .N,</span>
<span id="cb24-25"><a href="#cb24-25" aria-hidden="true" tabindex="-1"></a>            <span class="at">info_size =</span> <span class="fu">sum</span>((T <span class="sc">-</span> <span class="fu">mean</span>(T))<span class="sc">^</span><span class="dv">2</span>),</span>
<span id="cb24-26"><a href="#cb24-26" aria-hidden="true" tabindex="-1"></a>            <span class="at">num_treated =</span> <span class="fu">sum</span>(T),</span>
<span id="cb24-27"><a href="#cb24-27" aria-hidden="true" tabindex="-1"></a>            <span class="at">num_ctrl =</span> <span class="fu">sum</span>(<span class="sc">!</span>T)</span>
<span id="cb24-28"><a href="#cb24-28" aria-hidden="true" tabindex="-1"></a>          ), by <span class="ot">=</span> temp_var <span class="sc">&lt;</span> x] <span class="sc">%&gt;%</span> </span>
<span id="cb24-29"><a href="#cb24-29" aria-hidden="true" tabindex="-1"></a>          <span class="fu">setnames</span>(<span class="st">"temp_var"</span>, <span class="st">"child"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb24-30"><a href="#cb24-30" aria-hidden="true" tabindex="-1"></a>          .[, child <span class="sc">:</span><span class="er">=</span> <span class="fu">ifelse</span>(child <span class="sc">==</span> <span class="cn">TRUE</span>, <span class="st">"Child 1"</span>, <span class="st">"Child 2"</span>)] <span class="sc">%&gt;%</span> </span>
<span id="cb24-31"><a href="#cb24-31" aria-hidden="true" tabindex="-1"></a>          .[, thr <span class="sc">:</span><span class="er">=</span> x]</span>
<span id="cb24-32"><a href="#cb24-32" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb24-33"><a href="#cb24-33" aria-hidden="true" tabindex="-1"></a>        <span class="fu">return</span>(return_data)</span>
<span id="cb24-34"><a href="#cb24-34" aria-hidden="true" tabindex="-1"></a>      } </span>
<span id="cb24-35"><a href="#cb24-35" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">%&gt;%</span> </span>
<span id="cb24-36"><a href="#cb24-36" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rbindlist</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb24-37"><a href="#cb24-37" aria-hidden="true" tabindex="-1"></a>    .[, var <span class="sc">:</span><span class="er">=</span> feature_var] <span class="sc">%&gt;%</span> </span>
<span id="cb24-38"><a href="#cb24-38" aria-hidden="true" tabindex="-1"></a>    <span class="fu">relocate</span>(var, thr, child)</span>
<span id="cb24-39"><a href="#cb24-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-40"><a href="#cb24-40" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(ss_value)</span>
<span id="cb24-41"><a href="#cb24-41" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For example, here is the output for <code>x1</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">get_ss_by_var</span>(<span class="st">"x1"</span>, <span class="st">"rho"</span>, data_J1)[]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     var        thr   child   het_score nobs  info_size num_treated num_ctrl
  1:  x1 0.00332332 Child 2  0.04640943  249 62.1285141         130      119
  2:  x1 0.00332332 Child 1 11.55594717    1  0.0000000           1        0
  3:  x1 0.02031884 Child 2  0.03481370  248 61.8548387         130      118
  4:  x1 0.02031884 Child 1  4.31689906    2  0.5000000           1        1
  5:  x1 0.03951186 Child 2  0.24122929  247 61.5789474         130      117
 ---                                                                        
494:  x1 2.92655569 Child 2 76.66320951    3  0.6666667           1        2
495:  x1 2.95306339 Child 1  0.30401476  248 61.8548387         130      118
496:  x1 2.95306339 Child 2 37.69783018    2  0.5000000           1        1
497:  x1 2.96828632 Child 1  0.09327709  249 62.0803213         131      118
498:  x1 2.96828632 Child 2 23.22599563    1  0.0000000           0        1</code></pre>
</div>
</div>
<p>Repeating this for all the feature variables,</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>thr_score_data <span class="ot">&lt;-</span></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">paste0</span>(<span class="st">"x"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>),</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">function</span>(x) <span class="fu">get_ss_by_var</span>(x, <span class="st">"rho"</span>, data_J1)</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span> </span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rbindlist</span>()</span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      var        thr   child    het_score nobs  info_size num_treated num_ctrl
   1:  x1 0.00332332 Child 2 4.640943e-02  249 62.1285141         130      119
   2:  x1 0.00332332 Child 1 1.155595e+01    1  0.0000000           1        0
   3:  x1 0.02031884 Child 2 3.481370e-02  248 61.8548387         130      118
   4:  x1 0.02031884 Child 1 4.316899e+00    2  0.5000000           1        1
   5:  x1 0.03951186 Child 2 2.412293e-01  247 61.5789474         130      117
  ---                                                                         
4976: x10 0.98350685 Child 2 4.706587e-01    3  0.6666667           2        1
4977: x10 0.98856960 Child 1 7.289755e-04  248 61.8991935         129      119
4978: x10 0.98856960 Child 2 9.039296e-02    2  0.0000000           2        0
4979: x10 0.99656474 Child 1 3.553642e-02  249 62.1285141         130      119
4980: x10 0.99656474 Child 2 8.848568e+00    1  0.0000000           1        0</code></pre>
</div>
</div>
<p>Now, we ensure that we have at least as many <code>min.node.size</code> numbers of treated and control units in both child nodes. This provides a safeguard against having a very inaccurate treatment effect estimation. The default value of <code>min.node.size</code> is 5.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>min.node.size <span class="ot">&lt;-</span> <span class="dv">5</span> <span class="co"># referred to as mns</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a><span class="co">#=== check if both child nodes have at least mns control and treatment units  ===#</span></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>thr_score_data[, mns_met_grup <span class="sc">:</span><span class="er">=</span> <span class="fu">all</span>(num_ctrl <span class="sc">&gt;=</span> min.node.size <span class="sc">&amp;</span> num_treated <span class="sc">&gt;=</span> min.node.size), by <span class="ot">=</span> .(var, thr)]</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>msn_met_data <span class="ot">&lt;-</span> thr_score_data[mns_met_grup <span class="sc">==</span> <span class="cn">TRUE</span>, ]</span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      var       thr   child   het_score nobs info_size num_treated num_ctrl
   1:  x1 0.2647367 Child 2   4.5435708  238 59.294118         126      112
   2:  x1 0.2647367 Child 1  90.1141537   12  2.916667           5        7
   3:  x1 0.2957597 Child 2   5.5810225  237 59.071730         125      112
   4:  x1 0.2957597 Child 1 101.7463325   13  3.230769           6        7
   5:  x1 0.3148280 Child 2   5.0451340  236 58.792373         125      111
  ---                                                                      
4510: x10 0.9262760 Child 2   0.4888869   18  3.611111          13        5
4511: x10 0.9331452 Child 1   0.1739244  233 58.223176         119      114
4512: x10 0.9331452 Child 2   2.3837878   17  3.529412          12        5
4513: x10 0.9403202 Child 1   0.1370827  234 58.461538         120      114
4514: x10 0.9403202 Child 2   2.0048350   16  3.437500          11        5
      mns_met_grup
   1:         TRUE
   2:         TRUE
   3:         TRUE
   4:         TRUE
   5:         TRUE
  ---             
4510:         TRUE
4511:         TRUE
4512:         TRUE
4513:         TRUE
4514:         TRUE</code></pre>
</div>
</div>
<p>Now, we consider how <code>alpha</code> and <code>imbalance.penalty</code> affect the potential pool of feature-threshold combinations. Each child node needs to have at least as large <code>info_size</code> as <code>alpha</code> <span class="math inline">\(\times\)</span> <code>info_size</code> of the parent node. The <code>info_size</code> of the parent node is</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>info_size_p <span class="ot">&lt;-</span> data_J1[, <span class="fu">sum</span>((T <span class="sc">-</span> <span class="fu">mean</span>(T))<span class="sc">^</span><span class="dv">2</span>)]</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 62.356</code></pre>
</div>
</div>
<p>By default, <code>alpha</code> is set to 0.05. We use this number.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>alpha <span class="ot">&lt;-</span> <span class="fl">0.05</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>msn_met_data[, info_size_met <span class="sc">:</span><span class="er">=</span> <span class="fu">all</span>(info_size <span class="sc">&gt;</span> (alpha <span class="sc">*</span> info_size_p)), by <span class="ot">=</span> .(var, thr)]</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>msn_ifs_met <span class="ot">&lt;-</span> msn_met_data[info_size_met <span class="sc">==</span> <span class="cn">TRUE</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>imbalance.penalty</code> is used to further punish splits that introduce imbalance. For example, for feature <code>x1</code> and <code>thr</code> of 0.2957597,</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>   var       thr   child het_score nobs info_size num_treated num_ctrl
1:  x1 0.2647367 Child 2  4.543571  238 59.294118         126      112
2:  x1 0.2647367 Child 1 90.114154   12  2.916667           5        7
   mns_met_grup info_size_met
1:         TRUE         FALSE
2:         TRUE         FALSE</code></pre>
</div>
</div>
<p>the unpunished heterogeneity score is 94.6577245. With a non-zero value of <code>imbalance.penalty</code>, the following penalty will be subtracted from the unpunished heterogeneity score: <code>imbalance.penalty * (1/info_size_1 + 1/info_size_2)</code>. By default, <code>imbalance.penalty</code> is 0. But, let’s use <code>imbalance.penalty</code> <span class="math inline">\(= 1\)</span> here.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>imbalance.penalty <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>msn_ifs_met[, imb_penalty <span class="sc">:</span><span class="er">=</span> imbalance.penalty <span class="sc">*</span> (<span class="dv">1</span><span class="sc">/</span>info_size)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, we calculate the heterogeneity score for each feature-threshold with imbalance penalty.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>het_score_data <span class="ot">&lt;-</span> msn_ifs_met[, .(<span class="at">het_with_imbp =</span> <span class="fu">sum</span>(het_score <span class="sc">-</span> imb_penalty)), <span class="at">by =</span> .(var, thr)]</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      var       thr het_with_imbp
   1:  x1 0.2957597   107.0009025
   2:  x1 0.3148280    89.7830030
   3:  x1 0.3405788    90.7423400
   4:  x1 0.3557459    82.0711971
   5:  x1 0.3599312    88.7968930
  ---                            
2226: x10 0.9199764     0.9424508
2227: x10 0.9231653     0.1361964
2228: x10 0.9262760     0.2326482
2229: x10 0.9331452     2.2572036
2230: x10 0.9403202     1.8339034</code></pre>
</div>
</div>
<p>We now find the feature-threshold combination that maximizes the score,</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>best_split <span class="ot">&lt;-</span> het_score_data[<span class="fu">which.max</span>(het_with_imbp), ]</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   var      thr het_with_imbp
1:  x1 1.930929      953.3717</code></pre>
</div>
</div>
<p>So, we are splitting the root node using <code>x1</code> with threshold of 1.9309. Here is a visualization of the split,</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(data_J1) <span class="sc">+</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">y =</span> rho, <span class="at">x =</span> x1, <span class="at">color =</span> (x1 <span class="sc">&lt;</span> best_split[, thr]))) <span class="sc">+</span></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_discrete</span>(<span class="st">"Less than the threshold"</span>) <span class="sc">+</span></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"bottom"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="E02-grf_files/figure-html/viz-first-split-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Now, let’s look at the pseudo outcome and elaborate more on what we are seeing here.</p>
<p><span class="math display">\[
\begin{aligned}
\rho_i = (\tilde{Y}_i- \hat{\theta}_P\tilde{T}_i)\tilde{T}_i
\end{aligned}
\]</span></p>
<p>Note that <span class="math inline">\(\hat{\theta}_P\)</span> is just a constant and represents the average treatment effect for the parent node. So, <span class="math inline">\(\rho_i\)</span> tends to be higher for the observations whose CATE is higher (above-average treatment effect). So, splitting on <span class="math inline">\(\rho\)</span> (instead of <span class="math inline">\(Y\)</span>) made it possible to pick the right feature variable <code>x1</code>. This is because <code>x1</code> is influential in determining CATE and higher values of <code>x1</code> lead to more positive CATE.</p>
<p>Note that the checks implemented with <code>mtry</code>, <code>alpha</code>, and <code>imbalance.penalty</code> is more important further down a tree. But, the examples above should have illustrated how they are used.</p>
<p>Okay, now let’s split the root node into two child nodes according to the best split on <span class="math inline">\(\rho\)</span> we found earlier.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>depth_2_node_1 <span class="ot">&lt;-</span> data_J1[x1 <span class="sc">&lt;</span> best_split[, thr], ]</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>depth_2_node_2 <span class="ot">&lt;-</span> data_J1[x1 <span class="sc">&gt;=</span> best_split[, thr], ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s consider splitting the second node, <code>depth_2_node_1</code>. We basically follow exactly the same process the we just did. First, find <span class="math inline">\(\hat{\theta}_P\)</span> for this (parent) node using <a href="#eq-theta-p">Equation&nbsp;<span>17.8</span></a> and define the pseudo outcomes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>theta_p <span class="ot">&lt;-</span> depth_2_node_1[, <span class="fu">sum</span>(y_tilde <span class="sc">*</span> t_tilde)<span class="sc">/</span><span class="fu">sum</span>(t_tilde <span class="sc">*</span> t_tilde)]</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.8311055</code></pre>
</div>
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>depth_2_node_1[, rho <span class="sc">:</span><span class="er">=</span> (y_tilde <span class="sc">-</span> theta_p <span class="sc">*</span> t_tilde) <span class="sc">*</span> t_tilde]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, we find the best split (here we ignore various checks),</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>best_split_d2_n1 <span class="ot">&lt;-</span></span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">paste0</span>(<span class="st">"x"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>),</span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">function</span>(x) <span class="fu">get_ss_by_var</span>(x, <span class="st">"rho"</span>, depth_2_node_1)</span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span> </span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rbindlist</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a>  .[, .(<span class="at">het_score =</span> <span class="fu">sum</span>(het_score)), <span class="at">by =</span> .(var, thr)] <span class="sc">%&gt;%</span> </span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a>  .[<span class="fu">which.max</span>(het_score), ]</span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   var      thr het_score
1:  x3 3.827172  610.0837</code></pre>
</div>
</div>
<p>We keep splitting nodes until no splits is possible any more based on the value of <code>min.node.size</code>, <code>alpha</code>, and <code>imbalance.penalty</code>.</p>
<hr>
<p>To see the role of <code>honesty.prune.leaves</code>, let’s suppose we stopped splitting after the first split based on <code>x</code> with threshold of 1.9309. So, we just two terminal nodes. As stated in <a href="#sec-grf-honest"><span>Section&nbsp;17.4</span></a>, when we do prediction, <code>data_J1</code> (which is used for determining the splitting rule) is not used. Rather, <code>data_J2</code> (the data that was set aside) is used. When <code>honesty.prune.leaves = TRUE</code> (default), the tree is pruned so that no leaves are empty. Yes, we made sure that we have certain number of observations in each node with <code>min.node.size</code>, however, that is only for <code>data_J1</code>, not <code>data_J2</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>data_J2[, .N, by <span class="ot">=</span> x1 <span class="sc">&lt;</span> best_split[, thr]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      x1   N
1:  TRUE 172
2: FALSE  78</code></pre>
</div>
</div>
<p>Okay, so, each child has at least one observation. So, we are not going to prune the tree. If either of the child nodes was empty, then we would have removed the leaves and had the root node as the tree.</p>
<p>As you can see, while <code>data_J1</code> plays a central role in the tree building process, the other half, <code>data_J2</code>, plays a very small role. However, <code>data_J2</code> plays the central role in prediction, which will be seen in the next section.</p>
</section>
<section id="prediction" class="level3 page-columns page-full" data-number="17.5.3">
<h3 data-number="17.5.3" class="anchored" data-anchor-id="prediction"><span class="header-section-number">17.5.3</span> Prediction</h3>
<p>Suppose you have built <span class="math inline">\(T\)</span> trees already. In general for GRF, statistics of interest, <span class="math inline">\(\theta(X)\)</span>, at <span class="math inline">\(X = X_0\)</span> is found by solving the following equation:</p>
<p><span class="math display">\[
\begin{aligned}
\sum_{i=1}^N \alpha_i(X_i, X_0)\psi_{\theta,\nu}(O_i) = 0
\end{aligned}
\]</span></p>
<p>where <span class="math inline">\(\alpha_i(X_i, X_0)\)</span> is the weight given to each <span class="math inline">\(i\)</span> calculated based on the trees that have been built. Let <span class="math inline">\(\eta_{i,t}(X_i, X_0)\)</span> be 1 if observation <span class="math inline">\(i\)</span> belongs to the same leaf as <span class="math inline">\(X_0\)</span> in tree <span class="math inline">\(t\)</span>. Then,</p>
<p><span id="eq-weight-cf"><span class="math display">\[
\begin{aligned}
\alpha_i(X_i, X_0) = \frac{1}{T}\sum_{t=1}^T\frac{\eta_{i,t}(X_i, X_0)}{\sum_{i=1}^N\eta_{i,t}(X_i, X_0)}
\end{aligned}
\tag{17.9}\]</span></span></p>
<p>So, the weight given to observation <span class="math inline">\(i\)</span> is higher if observation <span class="math inline">\(i\)</span> belongs to the same leaf as the evaluation point <span class="math inline">\(X_0\)</span> in more trees.</p>
<p>For CF, since the score function is defined as <span class="math inline">\(\psi_{\theta,\nu}(O_i) = \tilde{Y}_i - \theta\tilde{T}_i)\cdot \tilde{T}_i\)</span>, the equation can be written as</p>
<p><span class="math display">\[
\begin{aligned}
\sum_{i=1}^N \alpha_i(X_i, X_0)[\tilde{Y_i} - \theta\cdot \tilde{T_i}]\hat{\tilde{T_i}} = 0
\end{aligned}
\]</span></p>
<p>So,</p>
<p><span class="math display">\[
\begin{aligned}
\theta_P = \frac{\sum_{i=1}^n \alpha_i(X_i, X_0)\tilde{Y}_i\cdot \tilde{T}_i}{\sum_{i=1}^n\alpha_i(X_i, X_0)\tilde{T}_i\cdot \tilde{T}_i}
\end{aligned}
\]</span></p>
<hr>
<p>We now illustrate how predictions are done once trees have been built. Let’s first build trees using <code>causal_forest()</code>. <code>min.node.size</code> is set deliberately high so we can work with very simple trees.</p>

<div class="no-row-height column-margin column-container"><div class="">
<p>You do not need to know how <code>causal_forest()</code> works. You only need to know that <code>causal_forest()</code> build trees for CATE estimation.</p>
</div></div><div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">443784</span>)</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>cf_trained <span class="ot">&lt;-</span></span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">causal_forest</span>(</span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">X =</span> <span class="fu">select</span>(data, <span class="fu">starts_with</span>(<span class="st">"x"</span>)),</span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">Y =</span> data[, y],</span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">W =</span> data[, T],</span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">min.node.size =</span> <span class="dv">50</span></span>
<span id="cb49-9"><a href="#cb49-9" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>After training a causal forest model, we have trees like the one shown in <a href="#fig-cf-tree">Figure&nbsp;<span>17.3</span></a>, which is the first of the <span class="math inline">\(2000\)</span> trees.</p>
<div class="cell">
<div id="fig-cf-tree" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<div id="htmlwidget-493f9da442108ae486f6" style="width:100%;height:371px;" class="grViz html-widget"></div>
<script type="application/json" data-for="htmlwidget-493f9da442108ae486f6">{"x":{"diagram":"digraph nodes { \n node [shape=box] ;\n0 [label=\" x1 <= 1.44 \"] ;\n0 -> 1 [labeldistance=2.5, labelangle=45, headlabel=\"True\", arrowhead=normal];\n0 -> 2 [labeldistance=2.5, labelangle=-45, headlabel=\"False\", arrowhead=normal];\n1  [shape=box,style=filled,color=\".7 .3 1.0\" , label=\"size =  121 \n avg_Y = 38.6\navg_W = 0.55 \"];\n2  [shape=box,style=filled,color=\".7 .3 1.0\" , label=\"size =  129 \n avg_Y = 42.1\navg_W = 0.5 \"];\n}","config":{"engine":"dot","options":null}},"evals":[],"jsHooks":[]}</script>
<p></p><figcaption aria-hidden="true" class="figure-caption">Figure 17.3: Example trees built in causal forest estimation</figcaption><p></p>
</figure>
</div>
</div>
<p>You probably noticed that the total number of samples in the leaves is only <span class="math inline">\(250\)</span> instead of <span class="math inline">\(1000\)</span>, which is the total number of observations in <code>data</code>. When causal forest was trained on this dataset, only half of the entire sample are randomly selected for building each tree (due to the default setting of <code>sample.fraction = 0.5</code>). The halved sample is further split into two groups, each containing <span class="math inline">\(250\)</span> observations (due to the default setting of <code>honesty = TRUE</code> and <code>honest.fraction = 0.5</code>). Let’s call them <span class="math inline">\(J_1\)</span> and <span class="math inline">\(J_2\)</span>. Then, <span class="math inline">\(J_1\)</span> is used to train a tree to find the splitting rules (e.g., <span class="math inline">\(x_1 \leq 1.44\)</span> for the first tree). See <a href="#sec-build-trees"><span>Section&nbsp;17.5.2</span></a> for how only a subset of the drawn samples was used to determine the splitting rules. Once the splitting rules are determined (tree building process is complete), then <span class="math inline">\(J_1\)</span> “vacates” the tree. Then, <span class="math inline">\(J_2\)</span> repopulates the tree. In other words, the splitting rule associated with the tree is applied to each of the samples in <span class="math inline">\(J_2\)</span> and place it in the node they belong to. So, the size value in each of the terminal nodes represents how many samples in <span class="math inline">\(J_2\)</span> are in each node. The statistics for each terminal node are from the samples in <span class="math inline">\(J_2\)</span> within the node. This will become clearer when we talk about finding individual weights.</p>
<p>Let’s take a look at a tree to see what happened. We can use <code>get_tree()</code> to access individual trees from <code>cf_trained</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co">#=== get the first tree ===#</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>a_tree <span class="ot">&lt;-</span> <span class="fu">get_tree</span>(cf_trained, <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>drawn_samples</code> attribute of the tree contains row indices that are selected randomly for this tree.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(a_tree<span class="sc">$</span>drawn_samples)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 788 968 693  47 673 584</code></pre>
</div>
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(a_tree<span class="sc">$</span>drawn_samples)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 500</code></pre>
</div>
</div>
<p>As you can see, there are 500 samples (due to <code>sample.fraction = 0.5</code>). The rest of the observations were not used for this tree at all. Accessing <code>nodes</code> attribute will give you the splitting rules for the tree built and which samples are in what node.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>nodes <span class="ot">&lt;-</span> a_tree<span class="sc">$</span>nodes</span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[[1]]
[[1]]$is_leaf
[1] FALSE

[[1]]$split_variable
[1] 1

[[1]]$split_value
[1] 1.438102

[[1]]$send_missing_left
[1] TRUE

[[1]]$left_child
[1] 2

[[1]]$right_child
[1] 3


[[2]]
[[2]]$is_leaf
[1] TRUE

[[2]]$samples
  [1] 539  75 927 812 489 264 344 479 567 116 750 744 295 363 580 754 210 601
 [19] 246  68 237 528 751 782 422 842 801 517 916 319 308   7 164 449 485  37
 [37] 576 235 431 154  22  14 867 354 940 911 373 123 104 798 133 587 506 219
 [55] 531 577 562 693 298 654 895 991 974 566 176 475  21 184 380 706  59 522
 [73] 288 936 792 117 141  89 783 252 663 513 417 756 605 427 126 401 242 818
 [91] 708 983 271 512 737 817 753 811 658 494 142 932 969  69  70 370 837 124
[109] 265 128 470 860 550 611 200 950 721 584 293  81 262

[[2]]$leaf_stats
avg_Y avg_W 
38.60  0.55 


[[3]]
[[3]]$is_leaf
[1] TRUE

[[3]]$samples
  [1] 411 478  38 307  39 997 869 268 276 480 425 633   8 645 999 544 984 579
 [19] 855 690 592 673 776 253   3 909 939 423 490 409 607 337 103 274 734 174
 [37] 342 339 526 878 634 503 407 429 161 788 486 699 937 330 361 406 435 809
 [55] 948 163 747 724 416 665 329  52 714 864 483 957 946 970 204 458   5 543
 [73] 624 333 698 358  34 903 791 378 581  41 146 107 735 109 741 858 666 495
 [91] 955 438 106 125 606 840  44 151  28 334 891  49 498 583 879 518 291 908
[109] 275 460 310 767 968 959 722  10 873 487 619 178 393 474 352 364 914 136
[127] 770 757 976

[[3]]$leaf_stats
avg_Y avg_W 
 42.1   0.5 </code></pre>
</div>
</div>
<p><code>nodes</code> is a list of three elements (one root node and two terminal nodes here). The <code>samples</code> attribute gives you row indices of the samples that belong to the terminal nodes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="co">#=== left node ===#  </span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>nodes[[<span class="dv">2</span>]]<span class="sc">$</span>samples</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  [1] 539  75 927 812 489 264 344 479 567 116 750 744 295 363 580 754 210 601
 [19] 246  68 237 528 751 782 422 842 801 517 916 319 308   7 164 449 485  37
 [37] 576 235 431 154  22  14 867 354 940 911 373 123 104 798 133 587 506 219
 [55] 531 577 562 693 298 654 895 991 974 566 176 475  21 184 380 706  59 522
 [73] 288 936 792 117 141  89 783 252 663 513 417 756 605 427 126 401 242 818
 [91] 708 983 271 512 737 817 753 811 658 494 142 932 969  69  70 370 837 124
[109] 265 128 470 860 550 611 200 950 721 584 293  81 262</code></pre>
</div>
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="co">#=== right node ===#  </span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>nodes[[<span class="dv">3</span>]]<span class="sc">$</span>samples</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  [1] 411 478  38 307  39 997 869 268 276 480 425 633   8 645 999 544 984 579
 [19] 855 690 592 673 776 253   3 909 939 423 490 409 607 337 103 274 734 174
 [37] 342 339 526 878 634 503 407 429 161 788 486 699 937 330 361 406 435 809
 [55] 948 163 747 724 416 665 329  52 714 864 483 957 946 970 204 458   5 543
 [73] 624 333 698 358  34 903 791 378 581  41 146 107 735 109 741 858 666 495
 [91] 955 438 106 125 606 840  44 151  28 334 891  49 498 583 879 518 291 908
[109] 275 460 310 767 968 959 722  10 873 487 619 178 393 474 352 364 914 136
[127] 770 757 976</code></pre>
</div>
</div>
<p>It is important to keep in mind that these observations with these row indices belong to <span class="math inline">\(J_2\)</span>. These observations were <span style="color:red">NOT</span> used in determining the splitting rule of <span class="math inline">\(x_1 \leq 1.44\)</span>. They were populating the terminal nodes by simply following the splitting rule, which is determined using the data in <span class="math inline">\(J_1\)</span> following the process described in <a href="#sec-build-trees"><span>Section&nbsp;17.5.2</span></a>. The difference in <code>a_tree$drawn_samples</code> and the combination of <code>nodes[[2]]$samples</code> and <code>nodes[[3]]$samples</code> is <span class="math inline">\(J_1\)</span>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>J2_rows <span class="ot">&lt;-</span> <span class="fu">c</span>(nodes[[<span class="dv">2</span>]]<span class="sc">$</span>samples, nodes[[<span class="dv">3</span>]]<span class="sc">$</span>samples)</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>J1_J2_rows <span class="ot">&lt;-</span> a_tree<span class="sc">$</span>drawn_samples</span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a>J1_rows <span class="ot">&lt;-</span> J1_J2_rows[J1_J2_rows <span class="sc">%in%</span> J2_rows]</span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  [1] 788 968 693 673 584  10 706 378 879 506 576 344 809 690 867 818 393 480
 [19] 665 107 908 486   7 352 513 109 339 744 298   8 747 983 767 174 291  22
 [37]  49 754 125 478 539 531 724 310 204 801 253 409 624 550 361 423 106  70
 [55] 406 543 526 293 237 490 663 528 991 136 869 911 518 811 946 722 916  39
 [73] 487 370 333 425 658 184 579  21 337 698 937 219 708 566 782 936 737 178
 [91]  89 334 721  28 592 128 970 435 798 271 753 151  59 116 460 103 948  38
[109] 154 431 817 751 133 909 489 633 878 308 276 999 950 997 619 939  14 873
[127] 562 252 605 458   3 163 714 246 858 164 319  68  81 984 959 123 581 438
[145] 666 416 645 757 380 756 634 407   5 587 770 265 483 611 927 401 274 268
[163] 974  34 567 842 429  75 577 475 812 734 914 517 449 494 141 606 161 932
[181] 474 699 330 792 601 275 583 895 860 295 783 544 503 117 235 855 654 891
[199] 126 864 969 411 176 485 840 142 735 976 512 373 262 417 364 498 422 903
[217] 342 607 522 580 307 358 791  52 288 210 354 146 200 124 427  44 837 104
[235]  37 495 940 329 470 479 955 264 242 776 741 363 957  69 750  41</code></pre>
</div>
</div>
<p>As you can see, there are 250 samples in <span class="math inline">\(J_1\)</span> as well. They were used to find the splitting rule.</p>
<p>For all the trees, the information we went through here is stored and can be used for predicting <span class="math inline">\(\theta(X)\)</span>.</p>
<hr>
<p>Suppose you are interested in predicting <span class="math inline">\(\hat{\theta}\)</span> (CATE, here) at <span class="math inline">\(X_0 = \{x_1 = 2, x_2,\dots, x_{10} = 1\}\)</span>. We can simply use <code>predict()</code> function in practice. But, we look into what is happening in <code>prediction()</code> behind the scene here.</p>
<p>For a given tree, we give 1 to the observations in <span style="color:red"><span class="math inline">\(J_2\)</span></span> that belong to the same leaf as <span class="math inline">\(X_0\)</span>. For example, for the first tree, <span class="math inline">\(X_0\)</span> belongs to the right leaf because <span class="math inline">\(x1 = 2 &gt; 1.44\)</span> for <span class="math inline">\(X_0\)</span>. We can tell which node <span class="math inline">\(X_0\)</span> belongs to by supplying <span class="math inline">\(X_0\)</span> to <code>get_leaf_node()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>X_0 <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="dv">2</span>, <span class="fu">rep</span>(<span class="dv">1</span>, <span class="dv">9</span>)), <span class="at">nrow =</span> <span class="dv">1</span>)</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a>which_tree_is_X0_in <span class="ot">&lt;-</span> <span class="fu">get_leaf_node</span>(a_tree, X_0)</span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 3</code></pre>
</div>
</div>
<p>So, we give <span class="math inline">\(1/N_t(X_0)\)</span> to all those in the right leaf (the third node in <code>nodes</code>) and 0 to those in the left leaf, where <span class="math inline">\(N_t(X_0)\)</span> is the number of observations that belong to the same leaf as <span class="math inline">\(X_0\)</span>. All the other observations are assigned 0.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="co">#=== which row numbers in the same leaf as X_0? ===#</span></span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a>rows_1 <span class="ot">&lt;-</span> nodes[[which_tree_is_X0_in]]<span class="sc">$</span>samples</span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a><span class="co">#=== define eta for tree 1  ===#</span></span>
<span id="cb65-5"><a href="#cb65-5" aria-hidden="true" tabindex="-1"></a>data[, eta_t1 <span class="sc">:</span><span class="er">=</span> <span class="dv">0</span>] <span class="co"># first set eta to 0 for all the observations</span></span>
<span id="cb65-6"><a href="#cb65-6" aria-hidden="true" tabindex="-1"></a>data[rows_1, eta_t1 <span class="sc">:</span><span class="er">=</span> <span class="dv">1</span> <span class="sc">/</span> <span class="fu">length</span>(rows_1)] <span class="co"># replace eta with 1/N_t(X_0) if in the right node</span></span>
<span id="cb65-7"><a href="#cb65-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-8"><a href="#cb65-8" aria-hidden="true" tabindex="-1"></a><span class="co">#=== see the data ===#</span></span>
<span id="cb65-9"><a href="#cb65-9" aria-hidden="true" tabindex="-1"></a>data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               y     T        x1        x2          x3        x4        x5
   1:  31.239440 FALSE 2.6951250 1.1965331  0.96678638 0.7210003 0.6207226
   2:  10.144462 FALSE 0.7919764 0.2722983  1.34510173 0.6267043 0.8869069
   3:  48.939026  TRUE 2.7143095 0.9040185  1.19934789 0.5620508 0.5455276
   4:  36.405483  TRUE 1.0628966 1.0659064  1.20667390 0.3638505 0.3201202
   5:  11.893620 FALSE 2.8824810 1.8198348  0.03356061 0.2570614 0.1447951
  ---                                                                     
 996:  74.029985  TRUE 1.6485987 2.8588233  1.04216169 0.6835214 0.7266654
 997:  21.176748  TRUE 2.0208231 0.4248353  0.49803263 0.2851495 0.8076055
 998:   3.388729 FALSE 1.9203044 0.5051511 -0.02137629 0.9961128 0.8446825
 999:   9.459730 FALSE 1.7274390 0.3468632  0.85507892 0.9669933 0.4943592
1000: 109.282950 FALSE 1.7030476 2.1927368  2.21194227 0.1442201 0.6783044
              x6         x7         x8         x9       x10    y_tilde
   1: 0.78698773 0.88919853 0.25543604 0.05640156 0.3560356 -5.6636071
   2: 0.44413365 0.27657781 0.08810472 0.20397022 0.7982468 -4.3210587
   3: 0.45238337 0.41956478 0.69912161 0.05602666 0.5905471 11.9768866
   4: 0.71060284 0.11400965 0.43547480 0.57807247 0.5707400  0.2253922
   5: 0.46912302 0.31966840 0.90711440 0.15150039 0.9426457 -8.4757974
  ---                                                                 
 996: 0.05036928 0.18803515 0.40981100 0.04716956 0.6443493  6.1784463
 997: 0.14171969 0.08375323 0.78676094 0.27132948 0.6570423  6.4855740
 998: 0.36225572 0.47555166 0.40332714 0.28135063 0.9657502 -8.0385906
 999: 0.43818445 0.59426511 0.53555370 0.91860942 0.3339725 -3.2516710
1000: 0.23674030 0.25005706 0.05512990 0.24352242 0.7227397 -1.5570047
         t_tilde      eta_t1
   1: -0.5296217 0.000000000
   2: -0.4788832 0.000000000
   3:  0.4703248 0.007751938
   4:  0.5534958 0.000000000
   5: -0.5011004 0.007751938
  ---                       
 996:  0.5504383 0.000000000
 997:  0.5684978 0.007751938
 998: -0.5107851 0.000000000
 999: -0.5187684 0.007751938
1000: -0.3917276 0.000000000</code></pre>
</div>
</div>
<p>We repeat this for all the trees and use <a href="#eq-weight-cf">Equation&nbsp;<span>17.9</span></a> to calculate the weights for the individual observations. The following function gets <span class="math inline">\(\eta_{i,t}(X_i, X_0)\)</span> for a given tree for all the observations.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>get_eta <span class="ot">&lt;-</span> <span class="cf">function</span>(t, X_0) {</span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>  w_tree <span class="ot">&lt;-</span> <span class="fu">get_tree</span>(cf_trained, t)</span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a>  which_tree_is_X0_in <span class="ot">&lt;-</span> <span class="fu">get_leaf_node</span>(w_tree, X_0)</span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a>  rows <span class="ot">&lt;-</span> w_tree<span class="sc">$</span>nodes[[which_tree_is_X0_in]]<span class="sc">$</span>samples</span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a>  eta_data <span class="ot">&lt;-</span> </span>
<span id="cb67-7"><a href="#cb67-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.table</span>(</span>
<span id="cb67-8"><a href="#cb67-8" aria-hidden="true" tabindex="-1"></a>      <span class="at">row_id =</span> <span class="fu">seq_len</span>(<span class="fu">nrow</span>(data)),</span>
<span id="cb67-9"><a href="#cb67-9" aria-hidden="true" tabindex="-1"></a>      <span class="at">eta =</span> <span class="fu">rep</span>(<span class="dv">0</span>, <span class="fu">nrow</span>(data))</span>
<span id="cb67-10"><a href="#cb67-10" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">%&gt;%</span> </span>
<span id="cb67-11"><a href="#cb67-11" aria-hidden="true" tabindex="-1"></a>    .[rows, eta <span class="sc">:</span><span class="er">=</span> <span class="dv">1</span> <span class="sc">/</span> <span class="fu">length</span>(rows)]</span>
<span id="cb67-12"><a href="#cb67-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-13"><a href="#cb67-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(eta_data)</span>
<span id="cb67-14"><a href="#cb67-14" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We apply <code>get_eta()</code> to each of the 2000 trees.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>eta_all <span class="ot">&lt;-</span></span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(</span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>    <span class="dv">1</span><span class="sc">:</span><span class="dv">2000</span>,</span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">function</span>(x) <span class="fu">get_eta</span>(x, X_0)</span>
<span id="cb68-6"><a href="#cb68-6" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span> </span>
<span id="cb68-7"><a href="#cb68-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rbindlist</span>(<span class="at">idcol =</span> <span class="st">"t"</span>)</span>
<span id="cb68-8"><a href="#cb68-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            t row_id         eta
      1:    1      1 0.000000000
      2:    1      2 0.000000000
      3:    1      3 0.007751938
      4:    1      4 0.000000000
      5:    1      5 0.007751938
     ---                        
1999996: 2000    996 0.006622517
1999997: 2000    997 0.000000000
1999998: 2000    998 0.000000000
1999999: 2000    999 0.000000000
2000000: 2000   1000 0.000000000</code></pre>
</div>
</div>
<p>Calculate the mean of <span class="math inline">\(\eta_{i,t}\)</span> by <code>row_id</code> (observation) to calculate <span class="math inline">\(\alpha(X_i, X_0)\)</span>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a>weights <span class="ot">&lt;-</span> </span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>  eta_all <span class="sc">%&gt;%</span> </span>
<span id="cb70-4"><a href="#cb70-4" aria-hidden="true" tabindex="-1"></a>  .[, .(<span class="at">weight =</span> <span class="fu">mean</span>(eta)), <span class="at">by =</span> row_id]</span>
<span id="cb70-5"><a href="#cb70-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      row_id       weight
   1:      1 0.0020590623
   2:      2 0.0002230001
   3:      3 0.0021290241
   4:      4 0.0002044371
   5:      5 0.0018758049
  ---                    
 996:    996 0.0007930756
 997:    997 0.0020828039
 998:    998 0.0022451720
 999:    999 0.0011955923
1000:   1000 0.0009944451</code></pre>
</div>
</div>
<p>Here is the observations that was given the highest and lowest weights.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>data_with_wights <span class="ot">&lt;-</span> <span class="fu">cbind</span>(data, weights)</span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a><span class="co">#=== highest (1st) and lowest (2nd) ===#</span></span>
<span id="cb72-4"><a href="#cb72-4" aria-hidden="true" tabindex="-1"></a>data_with_wights[weight <span class="sc">%in%</span> <span class="fu">c</span>(<span class="fu">max</span>(weight), <span class="fu">min</span>(weight)), ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            y     T        x1       x2        x3         x4        x5
1:   9.795849 FALSE 2.5423094 0.713593 0.3197887 0.05861231 0.3453091
2: 140.585346  TRUE 0.8604428 2.631900 2.4980926 0.62398495 0.2051564
           x6        x7        x8        x9       x10    y_tilde    t_tilde
1: 0.78236485 0.8942680 0.7315653 0.5546698 0.6760430 -13.150179 -0.5710382
2: 0.09216643 0.4551321 0.2967835 0.3810359 0.3444893   1.000322  0.4291793
   eta_t1 row_id       weight
1:      0    563 2.400879e-03
2:      0    987 1.020478e-05</code></pre>
</div>
</div>
<p>Then, we can use <a href="C03-cf-orf.html#eq-theta-solution">Equation&nbsp;<span>13.3</span></a> to calculate <span class="math inline">\(\hat{\theta}(X_0)\)</span>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a>theta_X0 <span class="ot">&lt;-</span> <span class="fu">sum</span>(data_with_wights[, weight <span class="sc">*</span> (T<span class="sc">-</span>cf_trained<span class="sc">$</span>W.hat) <span class="sc">*</span> (y<span class="sc">-</span>cf_trained<span class="sc">$</span>Y.hat)]) <span class="sc">/</span> <span class="fu">sum</span>(data_with_wights[, weight <span class="sc">*</span> (T<span class="sc">-</span>cf_trained<span class="sc">$</span>W.hat)<span class="sc">^</span><span class="dv">2</span>])</span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 13.44869</code></pre>
</div>
</div>
<p>Note that <span class="math inline">\(Y.hat\)</span> and <span class="math inline">\(W.hat\)</span> attributes of <code>cf_trained</code> are the estimates of <span class="math inline">\(E[Y|X]\)</span> and <span class="math inline">\(E[T|X]\)</span>, respectively. By subtracting them from <span class="math inline">\(Y\)</span> and <span class="math inline">\(T\)</span>, <span class="math inline">\(\tilde{Y}\)</span> and <span class="math inline">\(\tilde{T}\)</span> are calculated in the above code.</p>
<p>To get the weights, we could have just used <code>get_forest_weights()</code> like below:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a>weights <span class="ot">&lt;-</span> <span class="fu">get_forest_weights</span>(cf_trained, X_0)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s compare the weights and see if we did it right.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(<span class="fu">abs</span>(<span class="fu">as.matrix</span>(weights) <span class="sc">-</span> data_with_wights<span class="sc">$</span>weight))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1.478887e-18</code></pre>
</div>
</div>
<p>Great, virtually the same.</p>
</section>
</section>
<section id="clustering" class="level2 page-columns page-full" data-number="17.6">
<h2 data-number="17.6" class="anchored" data-anchor-id="clustering"><span class="header-section-number">17.6</span> Clustering</h2>
<p>In this section, we will conduct MC simulations to see how cluserting affects CF estimation of CATE (<span class="math inline">\(\theta(X)\)</span>) and the standard error of <span class="math inline">\(\hat{\theta}(X)\)</span>. We use the folloinwg data generating process:</p>
<div class="callout-note callout callout-style-default no-icon callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
DGP 1
</div>
</div>
<div class="callout-body-container callout-body">
<p><span class="math display">\[
\begin{aligned}
Y_i &amp; = (3 x_{i,1} + 0.5(x_{i,2} x_{i,3}))\cdot T_i + x_{i,3} + 2 x_{i,3} x_{i,4} + \mu_i \\
T_i|X_i &amp; = Bernouli((1+x_{i,1})/2) \\
\mu_i|X_i &amp; = N(0,1)
\end{aligned}
\]</span></p>
<ul>
<li><span class="math inline">\(x_1 \sim U[0,1]\)</span></li>
<li><span class="math inline">\(x_2\sim U[0,2]\)</span></li>
<li><span class="math inline">\(x_3\sim U[0,5]\)</span></li>
<li><span class="math inline">\(x_4\dots, x_10\sim N(0,1)^7\)</span></li>
<li><span class="math inline">\(\mu\sim N(0, 9)\)</span></li>
</ul>
</div>
</div>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a>gen_cl_var <span class="ot">&lt;-</span>  <span class="cf">function</span>(G, Ng){</span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a>  var <span class="ot">&lt;-</span></span>
<span id="cb79-3"><a href="#cb79-3" aria-hidden="true" tabindex="-1"></a>    MASS<span class="sc">::</span><span class="fu">mvrnorm</span>(</span>
<span id="cb79-4"><a href="#cb79-4" aria-hidden="true" tabindex="-1"></a>      G, <span class="at">mu =</span> <span class="fu">rep</span>(<span class="dv">0</span>, Ng), </span>
<span id="cb79-5"><a href="#cb79-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">Sigma =</span> <span class="fu">matrix</span>(<span class="fl">0.75</span>, <span class="at">nrow =</span> Ng, <span class="at">ncol =</span> Ng) <span class="sc">+</span> <span class="fl">0.25</span> <span class="sc">*</span> <span class="fu">diag</span>(Ng)</span>
<span id="cb79-6"><a href="#cb79-6" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">%&gt;%</span> <span class="fu">t</span>() <span class="sc">%&gt;%</span> <span class="fu">c</span>()</span>
<span id="cb79-7"><a href="#cb79-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-8"><a href="#cb79-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(var)</span>
<span id="cb79-9"><a href="#cb79-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb79-10"><a href="#cb79-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-11"><a href="#cb79-11" aria-hidden="true" tabindex="-1"></a>gen_data <span class="ot">&lt;-</span> <span class="cf">function</span>(<span class="at">N =</span> <span class="dv">1000</span>, <span class="at">G =</span> <span class="dv">40</span>, <span class="at">K =</span> <span class="dv">10</span>){</span>
<span id="cb79-12"><a href="#cb79-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-13"><a href="#cb79-13" aria-hidden="true" tabindex="-1"></a>  Ng <span class="ot">&lt;-</span> N<span class="sc">/</span>G <span class="co"># number of observations per group</span></span>
<span id="cb79-14"><a href="#cb79-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-15"><a href="#cb79-15" aria-hidden="true" tabindex="-1"></a>  <span class="co">#=== error term ===#    </span></span>
<span id="cb79-16"><a href="#cb79-16" aria-hidden="true" tabindex="-1"></a>  mu <span class="ot">&lt;-</span> <span class="fu">gen_cl_var</span>(G, Ng) <span class="sc">*</span> <span class="dv">3</span></span>
<span id="cb79-17"><a href="#cb79-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-18"><a href="#cb79-18" aria-hidden="true" tabindex="-1"></a>  <span class="co">#=== X ===#</span></span>
<span id="cb79-19"><a href="#cb79-19" aria-hidden="true" tabindex="-1"></a>  X <span class="ot">&lt;-</span> </span>
<span id="cb79-20"><a href="#cb79-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">lapply</span>(</span>
<span id="cb79-21"><a href="#cb79-21" aria-hidden="true" tabindex="-1"></a>      <span class="dv">1</span><span class="sc">:</span>K,</span>
<span id="cb79-22"><a href="#cb79-22" aria-hidden="true" tabindex="-1"></a>      <span class="cf">function</span>(x) <span class="fu">gen_cl_var</span>(G, Ng)</span>
<span id="cb79-23"><a href="#cb79-23" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb79-24"><a href="#cb79-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">do.call</span>(<span class="st">"cbind"</span>, .) <span class="sc">%&gt;%</span></span>
<span id="cb79-25"><a href="#cb79-25" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.table</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb79-26"><a href="#cb79-26" aria-hidden="true" tabindex="-1"></a>    <span class="fu">setnames</span>(<span class="fu">names</span>(.), <span class="fu">paste0</span>(<span class="st">"x"</span>, <span class="fu">seq_len</span>(K)))</span>
<span id="cb79-27"><a href="#cb79-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-28"><a href="#cb79-28" aria-hidden="true" tabindex="-1"></a>  group <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">1</span><span class="sc">:</span>G, <span class="at">each =</span> Ng)</span>
<span id="cb79-29"><a href="#cb79-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-30"><a href="#cb79-30" aria-hidden="true" tabindex="-1"></a>  data <span class="ot">&lt;-</span></span>
<span id="cb79-31"><a href="#cb79-31" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cbind</span>(X, mu, group) <span class="sc">%&gt;%</span> </span>
<span id="cb79-32"><a href="#cb79-32" aria-hidden="true" tabindex="-1"></a>    <span class="co">#=== convert x1 and x3 to unif ===#</span></span>
<span id="cb79-33"><a href="#cb79-33" aria-hidden="true" tabindex="-1"></a>    .[, <span class="st">`</span><span class="at">:=</span><span class="st">`</span>(</span>
<span id="cb79-34"><a href="#cb79-34" aria-hidden="true" tabindex="-1"></a>      <span class="at">x1 =</span> <span class="fu">pnorm</span>(x1, <span class="at">sd =</span> <span class="fu">sqrt</span>(<span class="dv">2</span>)),</span>
<span id="cb79-35"><a href="#cb79-35" aria-hidden="true" tabindex="-1"></a>      <span class="at">x2 =</span> <span class="fu">pnorm</span>(x2, <span class="at">sd =</span> <span class="fu">sqrt</span>(<span class="dv">2</span>)) <span class="sc">*</span> <span class="dv">2</span>,</span>
<span id="cb79-36"><a href="#cb79-36" aria-hidden="true" tabindex="-1"></a>      <span class="at">x3 =</span> <span class="fu">pnorm</span>(x3, <span class="at">sd =</span> <span class="fu">sqrt</span>(<span class="dv">2</span>)) <span class="sc">*</span> <span class="dv">5</span></span>
<span id="cb79-37"><a href="#cb79-37" aria-hidden="true" tabindex="-1"></a>    )] <span class="sc">%&gt;%</span> </span>
<span id="cb79-38"><a href="#cb79-38" aria-hidden="true" tabindex="-1"></a>    .[, T <span class="sc">:</span><span class="er">=</span> <span class="fu">runif</span>(N) <span class="sc">&lt;</span> ((<span class="fl">0.5</span><span class="sc">+</span>x1)<span class="sc">/</span><span class="dv">2</span>)] <span class="sc">%&gt;%</span> </span>
<span id="cb79-39"><a href="#cb79-39" aria-hidden="true" tabindex="-1"></a>    .[, theta_x <span class="sc">:</span><span class="er">=</span> <span class="dv">3</span> <span class="sc">*</span> x1 <span class="sc">+</span> x2 <span class="sc">*</span> x3 <span class="sc">/</span> <span class="dv">2</span>] <span class="sc">%&gt;%</span></span>
<span id="cb79-40"><a href="#cb79-40" aria-hidden="true" tabindex="-1"></a>    .[, g_x <span class="sc">:</span><span class="er">=</span> x3 <span class="sc">+</span> <span class="dv">2</span> <span class="sc">*</span> x1 <span class="sc">*</span> x4] <span class="sc">%&gt;%</span></span>
<span id="cb79-41"><a href="#cb79-41" aria-hidden="true" tabindex="-1"></a>    .[, det_Y <span class="sc">:</span><span class="er">=</span> theta_x <span class="sc">*</span> T <span class="sc">+</span> g_x] <span class="sc">%&gt;%</span></span>
<span id="cb79-42"><a href="#cb79-42" aria-hidden="true" tabindex="-1"></a>    .[, Y <span class="sc">:</span><span class="er">=</span> det_Y <span class="sc">+</span> mu] <span class="sc">%&gt;%</span> </span>
<span id="cb79-43"><a href="#cb79-43" aria-hidden="true" tabindex="-1"></a>    .[, id <span class="sc">:</span><span class="er">=</span> <span class="dv">1</span><span class="sc">:</span>.N]</span>
<span id="cb79-44"><a href="#cb79-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-45"><a href="#cb79-45" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(data)</span>
<span id="cb79-46"><a href="#cb79-46" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Here is a dataset that is generated according to the DGP.</p>

<div class="no-row-height column-margin column-container"><div class="">
<p>Unfold the Code chunk above to see how <code>gen_data()</code> is defined.</p>
</div></div><div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">428943</span>)</span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">gen_data</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><a href="#fig-viz-cluster-data">Figure&nbsp;<span>17.4</span></a> presents the values of individual outcomes (<span class="math inline">\(Y\)</span>), error (<span class="math inline">\(\mu\)</span>), and outcomes less errors (<span class="math inline">\(\theta(X)\cdot T + g(X)\)</span>) and their respective means by cluster (<code>group</code>). You can observe that data are clustered by <code>group</code>.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a>plot_data <span class="ot">&lt;-</span></span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>  data[, .(id, Y, det_Y, mu, group)] <span class="sc">%&gt;%</span> </span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">melt</span>(<span class="at">id.var =</span> <span class="fu">c</span>(<span class="st">"id"</span>, <span class="st">"group"</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb81-4"><a href="#cb81-4" aria-hidden="true" tabindex="-1"></a>  .[, type <span class="sc">:</span><span class="er">=</span> <span class="fu">fcase</span>(</span>
<span id="cb81-5"><a href="#cb81-5" aria-hidden="true" tabindex="-1"></a>    variable <span class="sc">==</span> <span class="st">"Y"</span>, <span class="st">"Y"</span>,</span>
<span id="cb81-6"><a href="#cb81-6" aria-hidden="true" tabindex="-1"></a>    variable <span class="sc">==</span> <span class="st">"det_Y"</span>, <span class="st">"Y - Error"</span>,</span>
<span id="cb81-7"><a href="#cb81-7" aria-hidden="true" tabindex="-1"></a>    variable <span class="sc">==</span> <span class="st">"mu"</span>, <span class="st">"Error"</span></span>
<span id="cb81-8"><a href="#cb81-8" aria-hidden="true" tabindex="-1"></a>  )] <span class="sc">%&gt;%</span> </span>
<span id="cb81-9"><a href="#cb81-9" aria-hidden="true" tabindex="-1"></a>  .[, type <span class="sc">:</span><span class="er">=</span> <span class="fu">factor</span>(type, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Y"</span>, <span class="st">"Error"</span>, <span class="st">"Y - Error"</span>))]</span>
<span id="cb81-10"><a href="#cb81-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-11"><a href="#cb81-11" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> plot_data) <span class="sc">+</span></span>
<span id="cb81-12"><a href="#cb81-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">y =</span> value, <span class="at">x =</span> <span class="fu">factor</span>(group)), <span class="at">size =</span> <span class="fl">0.3</span>) <span class="sc">+</span></span>
<span id="cb81-13"><a href="#cb81-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(</span>
<span id="cb81-14"><a href="#cb81-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> plot_data[, .(<span class="at">value =</span> <span class="fu">mean</span>(value)), <span class="at">by =</span> .(group, type)],</span>
<span id="cb81-15"><a href="#cb81-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">y =</span> value, <span class="at">x =</span> <span class="fu">factor</span>(group)),</span>
<span id="cb81-16"><a href="#cb81-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">color =</span> <span class="st">"red"</span></span>
<span id="cb81-17"><a href="#cb81-17" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb81-18"><a href="#cb81-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"group"</span>) <span class="sc">+</span></span>
<span id="cb81-19"><a href="#cb81-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(type <span class="sc">~</span> .) <span class="sc">+</span></span>
<span id="cb81-20"><a href="#cb81-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb81-21"><a href="#cb81-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">6</span>, <span class="at">angle =</span> <span class="dv">90</span>)</span>
<span id="cb81-22"><a href="#cb81-22" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb81-23"><a href="#cb81-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-viz-cluster-data" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="E02-grf_files/figure-html/fig-viz-cluster-data-1.png" class="img-fluid figure-img" width="672"></p>
<p></p><figcaption aria-hidden="true" class="figure-caption">Figure 17.4: Visualization of clustered data</figcaption><p></p>
</figure>
</div>
</div>
</div>
<p>We now run MC simulations. We use the same test data for all the iterations so that we can find the standard error estimates averaged over iterations. The following function trains CFs with and without clustering by <code>group</code>, estimate <span class="math inline">\(\theta(X)\)</span> at the test data along with the estimate of the variance of <span class="math inline">\(\hat{\theta}(X)\)</span>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="co">#=== use fixed test data ===#</span></span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>test_data <span class="ot">&lt;-</span> <span class="fu">gen_data</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a>  .[, test_id <span class="sc">:</span><span class="er">=</span> <span class="dv">1</span><span class="sc">:</span>.N]</span>
<span id="cb82-4"><a href="#cb82-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-5"><a href="#cb82-5" aria-hidden="true" tabindex="-1"></a>run_MC_sim <span class="ot">&lt;-</span> <span class="cf">function</span>(i){</span>
<span id="cb82-6"><a href="#cb82-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-7"><a href="#cb82-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(i)</span>
<span id="cb82-8"><a href="#cb82-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-9"><a href="#cb82-9" aria-hidden="true" tabindex="-1"></a>  train_data <span class="ot">&lt;-</span> <span class="fu">gen_data</span>()</span>
<span id="cb82-10"><a href="#cb82-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-11"><a href="#cb82-11" aria-hidden="true" tabindex="-1"></a>  cf_trained_no_cluster <span class="ot">&lt;-</span></span>
<span id="cb82-12"><a href="#cb82-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">causal_forest</span>(</span>
<span id="cb82-13"><a href="#cb82-13" aria-hidden="true" tabindex="-1"></a>      <span class="at">X =</span> <span class="fu">select</span>(train_data, <span class="fu">starts_with</span>(<span class="st">"x"</span>)),</span>
<span id="cb82-14"><a href="#cb82-14" aria-hidden="true" tabindex="-1"></a>      <span class="at">Y =</span> train_data[, Y],</span>
<span id="cb82-15"><a href="#cb82-15" aria-hidden="true" tabindex="-1"></a>      <span class="at">W =</span> train_data[, T]</span>
<span id="cb82-16"><a href="#cb82-16" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb82-17"><a href="#cb82-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-18"><a href="#cb82-18" aria-hidden="true" tabindex="-1"></a>  thta_hat_no_cl <span class="ot">&lt;-</span> </span>
<span id="cb82-19"><a href="#cb82-19" aria-hidden="true" tabindex="-1"></a>    <span class="fu">predict</span>(</span>
<span id="cb82-20"><a href="#cb82-20" aria-hidden="true" tabindex="-1"></a>      cf_trained_no_cluster, </span>
<span id="cb82-21"><a href="#cb82-21" aria-hidden="true" tabindex="-1"></a>      <span class="fu">select</span>(test_data, <span class="fu">starts_with</span>(<span class="st">"x"</span>)),</span>
<span id="cb82-22"><a href="#cb82-22" aria-hidden="true" tabindex="-1"></a>      <span class="at">estimate.variance =</span> <span class="cn">TRUE</span></span>
<span id="cb82-23"><a href="#cb82-23" aria-hidden="true" tabindex="-1"></a>    )  <span class="sc">%&gt;%</span> </span>
<span id="cb82-24"><a href="#cb82-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.table</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb82-25"><a href="#cb82-25" aria-hidden="true" tabindex="-1"></a>    .[, theta_true <span class="sc">:</span><span class="er">=</span> test_data<span class="sc">$</span>theta_x] <span class="sc">%&gt;%</span> </span>
<span id="cb82-26"><a href="#cb82-26" aria-hidden="true" tabindex="-1"></a>    .[, id <span class="sc">:</span><span class="er">=</span> test_data<span class="sc">$</span>id] <span class="sc">%&gt;%</span> </span>
<span id="cb82-27"><a href="#cb82-27" aria-hidden="true" tabindex="-1"></a>    .[, type <span class="sc">:</span><span class="er">=</span> <span class="st">"No clustering"</span>]</span>
<span id="cb82-28"><a href="#cb82-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-29"><a href="#cb82-29" aria-hidden="true" tabindex="-1"></a>  cf_trained_with_cluster <span class="ot">&lt;-</span></span>
<span id="cb82-30"><a href="#cb82-30" aria-hidden="true" tabindex="-1"></a>    <span class="fu">causal_forest</span>(</span>
<span id="cb82-31"><a href="#cb82-31" aria-hidden="true" tabindex="-1"></a>      <span class="at">X =</span> <span class="fu">select</span>(train_data, <span class="fu">starts_with</span>(<span class="st">"x"</span>)),</span>
<span id="cb82-32"><a href="#cb82-32" aria-hidden="true" tabindex="-1"></a>      <span class="at">Y =</span> train_data[, Y],</span>
<span id="cb82-33"><a href="#cb82-33" aria-hidden="true" tabindex="-1"></a>      <span class="at">W =</span> train_data[, T],</span>
<span id="cb82-34"><a href="#cb82-34" aria-hidden="true" tabindex="-1"></a>      <span class="at">cluster =</span> train_data[, group]</span>
<span id="cb82-35"><a href="#cb82-35" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb82-36"><a href="#cb82-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-37"><a href="#cb82-37" aria-hidden="true" tabindex="-1"></a>  thta_hat_with_cl <span class="ot">&lt;-</span> </span>
<span id="cb82-38"><a href="#cb82-38" aria-hidden="true" tabindex="-1"></a>    <span class="fu">predict</span>(</span>
<span id="cb82-39"><a href="#cb82-39" aria-hidden="true" tabindex="-1"></a>      cf_trained_with_cluster, </span>
<span id="cb82-40"><a href="#cb82-40" aria-hidden="true" tabindex="-1"></a>      <span class="fu">select</span>(test_data, <span class="fu">starts_with</span>(<span class="st">"x"</span>)),</span>
<span id="cb82-41"><a href="#cb82-41" aria-hidden="true" tabindex="-1"></a>      <span class="at">estimate.variance =</span> <span class="cn">TRUE</span></span>
<span id="cb82-42"><a href="#cb82-42" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">%&gt;%</span> </span>
<span id="cb82-43"><a href="#cb82-43" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.table</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb82-44"><a href="#cb82-44" aria-hidden="true" tabindex="-1"></a>    .[, theta_true <span class="sc">:</span><span class="er">=</span> test_data<span class="sc">$</span>theta_x] <span class="sc">%&gt;%</span> </span>
<span id="cb82-45"><a href="#cb82-45" aria-hidden="true" tabindex="-1"></a>    .[, id <span class="sc">:</span><span class="er">=</span> test_data<span class="sc">$</span>id] <span class="sc">%&gt;%</span> </span>
<span id="cb82-46"><a href="#cb82-46" aria-hidden="true" tabindex="-1"></a>    .[, type <span class="sc">:</span><span class="er">=</span> <span class="st">"With clustering"</span>]</span>
<span id="cb82-47"><a href="#cb82-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-48"><a href="#cb82-48" aria-hidden="true" tabindex="-1"></a>  return_data <span class="ot">&lt;-</span> </span>
<span id="cb82-49"><a href="#cb82-49" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rbind</span>(thta_hat_no_cl, thta_hat_with_cl) <span class="sc">%&gt;%</span> </span>
<span id="cb82-50"><a href="#cb82-50" aria-hidden="true" tabindex="-1"></a>    .[, sim <span class="sc">:</span><span class="er">=</span> i]</span>
<span id="cb82-51"><a href="#cb82-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-52"><a href="#cb82-52" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(return_data)</span>
<span id="cb82-53"><a href="#cb82-53" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s repeat the experiment 200 times.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">428943</span>)</span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-3"><a href="#cb83-3" aria-hidden="true" tabindex="-1"></a>mc_results <span class="ot">&lt;-</span></span>
<span id="cb83-4"><a href="#cb83-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(</span>
<span id="cb83-5"><a href="#cb83-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">seq_len</span>(<span class="dv">200</span>),</span>
<span id="cb83-6"><a href="#cb83-6" aria-hidden="true" tabindex="-1"></a>    run_MC_sim</span>
<span id="cb83-7"><a href="#cb83-7" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">

</div>
<div class="cell">

</div>
<p><a href="#fig-cate-group-mse">Figure&nbsp;<span>17.5</span></a> shows the distribution of MSE of CATE estimation by CF with and without clustering by <code>group</code>. As you can see, clustering is making things worse in this instance.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a>mc_results <span class="sc">%&gt;%</span> </span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a>  .[, .(<span class="at">mse =</span> <span class="fu">mean</span>((theta_true <span class="sc">-</span> predictions)<span class="sc">^</span><span class="dv">2</span>)), by <span class="ot">=</span> .(sim, type)] <span class="sc">%&gt;%</span> </span>
<span id="cb84-3"><a href="#cb84-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="at">data =</span> .) <span class="sc">+</span> </span>
<span id="cb84-4"><a href="#cb84-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_density</span>(<span class="fu">aes</span>(<span class="at">x =</span> mse, <span class="at">fill =</span> type), <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb84-5"><a href="#cb84-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-cate-group-mse" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="E02-grf_files/figure-html/fig-cate-group-mse-1.png" class="img-fluid figure-img" width="672"></p>
<p></p><figcaption aria-hidden="true" class="figure-caption">Figure 17.5: MSE of CATE estimation by CF with and without clustering by group</figcaption><p></p>
</figure>
</div>
</div>
</div>
<p><a href="#fig-sd-test">Figure&nbsp;<span>17.6</span></a> shows the distribution of the ratio of the standard deviation of <span class="math inline">\(\hat{\theta}(X)\)</span> to the mean of the estimate of the standard deviation of <span class="math inline">\(\hat{\theta}(X)\)</span>. Ratio less than 1 indicates that the estimate of the standard deviation of <span class="math inline">\(\hat{\theta}(X)\)</span> is under-estimating the true standard deviation of <span class="math inline">\(\hat{\theta}(X)\)</span>.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a>mc_results <span class="sc">%&gt;%</span> </span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a>  .[, .(<span class="at">sd =</span> <span class="fu">sd</span>(predictions), <span class="at">sd_est =</span> <span class="fu">mean</span>(<span class="fu">sqrt</span>(variance.estimates))), by <span class="ot">=</span> .(id, type)] <span class="sc">%&gt;%</span> </span>
<span id="cb85-3"><a href="#cb85-3" aria-hidden="true" tabindex="-1"></a>  .[, ratio <span class="sc">:</span><span class="er">=</span> sd_est<span class="sc">/</span>sd] <span class="sc">%&gt;%</span> </span>
<span id="cb85-4"><a href="#cb85-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="at">data =</span> .) <span class="sc">+</span> </span>
<span id="cb85-5"><a href="#cb85-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_density</span>(<span class="fu">aes</span>(<span class="at">x =</span> ratio, <span class="at">fill =</span> type), <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb85-6"><a href="#cb85-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-sd-test" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="E02-grf_files/figure-html/fig-sd-test-1.png" class="img-fluid figure-img" width="672"></p>
<p></p><figcaption aria-hidden="true" class="figure-caption">Figure 17.6: Ratio of the estimated standard error of CATE to the standard deviation of CATE</figcaption><p></p>
</figure>
</div>
</div>
</div>
</section>
<section id="hyper-parameter-tuning" class="level2" data-number="17.7">
<h2 data-number="17.7" class="anchored" data-anchor-id="hyper-parameter-tuning"><span class="header-section-number">17.7</span> Hyper-parameter tuning</h2>
<p>You can write your own codes to conduct cross-validation just like any other ML methods as we saw in <a href="B03-cross-validation.html"><span>Chapter&nbsp;3</span></a> (cross-validation is a general approach that can be applied to any ML methods) to tune hyper-parameters. However, the <code>grf</code> package offers internal tuning options based on out-of-bag error (see <a href="P01-random-forest.html#sec-rf-implementation"><span>Section&nbsp;6.3</span></a> for what out-of-bag prediction is).</p>
<p>You can define which hyper-parameters to tune using <code>tune.parameters</code> option. For example, <code>tune.parameters = c("min.node.size", "mtry")</code> would tune <code>min.node.size</code> and <code>mtry</code>.</p>
<p>The cross-validation procedure takes the following steps:</p>
<div class="callout-note callout callout-style-default no-icon callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Cross-validation steps
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Draw <code>tune.num.reps</code> numbers of random points in the space of possible values of parameters specified by <code>tune.parameters</code></li>
<li>For each of the <code>tune.num.reps</code> sets of parameter values, <code>tune.num.trees</code> of trees are built and out-of-bag errors are calculated</li>
<li>Regress the error estimates on parameter values using smoothing functions</li>
<li>Redraw <code>tune.num.draws</code> numbers of random points in the space of possible values of parameters</li>
<li>Pick the set of parameter values that minimizes out-of-bag error.</li>
</ul>
</div>
</div>
</section>
<section id="references" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="references">References</h2>
<div id="refs" class="references csl-bib-body hanging-indent" role="doc-bibliography">
<div id="ref-athey2019generalized" class="csl-entry" role="doc-biblioentry">
Athey, Susan, Julie Tibshirani, and Stefan Wager. 2019. <span>“Generalized Random Forests.”</span> <em>The Annals of Statistics</em> 47 (2): 1148–78.
</div>
</div>


<!-- -->

</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      let href = ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./E01-spatial-cv.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Spatial Cross-validation</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./PROG-R-01-mlr3.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Machine Learning with <code>mlr3</code></span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb86" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Generalized Random Forest {#sec-grf}</span></span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-3"><a href="#cb86-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-4"><a href="#cb86-4" aria-hidden="true" tabindex="-1"></a>:::{.callout-important}</span>
<span id="cb86-5"><a href="#cb86-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-6"><a href="#cb86-6" aria-hidden="true" tabindex="-1"></a><span class="fu">## What you will learn</span></span>
<span id="cb86-7"><a href="#cb86-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-8"><a href="#cb86-8" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>What GRF is</span>
<span id="cb86-9"><a href="#cb86-9" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>How GRF was motivated</span>
<span id="cb86-10"><a href="#cb86-10" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>Random forest is a special case of GRF</span>
<span id="cb86-11"><a href="#cb86-11" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>Honesty rule</span>
<span id="cb86-12"><a href="#cb86-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-13"><a href="#cb86-13" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb86-14"><a href="#cb86-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-15"><a href="#cb86-15" aria-hidden="true" tabindex="-1"></a>:::{.callout-tip}</span>
<span id="cb86-16"><a href="#cb86-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-17"><a href="#cb86-17" aria-hidden="true" tabindex="-1"></a><span class="fu">## Background knowledge</span></span>
<span id="cb86-18"><a href="#cb86-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-19"><a href="#cb86-19" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>(necessary) how a regression tree is built (@sec-rt)</span>
<span id="cb86-20"><a href="#cb86-20" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>(necessary) random forest (@sec-rf)</span>
<span id="cb86-21"><a href="#cb86-21" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>(necessary) local regression (@sec-local)</span>
<span id="cb86-22"><a href="#cb86-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-23"><a href="#cb86-23" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb86-24"><a href="#cb86-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-25"><a href="#cb86-25" aria-hidden="true" tabindex="-1"></a>::: {.callout-note}</span>
<span id="cb86-26"><a href="#cb86-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-27"><a href="#cb86-27" aria-hidden="true" tabindex="-1"></a><span class="fu">## Packages to load for replication</span></span>
<span id="cb86-28"><a href="#cb86-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-31"><a href="#cb86-31" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-32"><a href="#cb86-32" aria-hidden="true" tabindex="-1"></a><span class="co">#| include: false</span></span>
<span id="cb86-33"><a href="#cb86-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-34"><a href="#cb86-34" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb86-35"><a href="#cb86-35" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb86-36"><a href="#cb86-36" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ranger)</span>
<span id="cb86-37"><a href="#cb86-37" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(parallel)</span>
<span id="cb86-38"><a href="#cb86-38" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(grf)</span>
<span id="cb86-39"><a href="#cb86-39" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rpart)</span>
<span id="cb86-40"><a href="#cb86-40" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rattle)</span>
<span id="cb86-41"><a href="#cb86-41" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(wooldridge)</span>
<span id="cb86-42"><a href="#cb86-42" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-43"><a href="#cb86-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-46"><a href="#cb86-46" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-47"><a href="#cb86-47" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb86-48"><a href="#cb86-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-49"><a href="#cb86-49" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb86-50"><a href="#cb86-50" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb86-51"><a href="#cb86-51" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ranger)</span>
<span id="cb86-52"><a href="#cb86-52" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(parallel)</span>
<span id="cb86-53"><a href="#cb86-53" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(grf)</span>
<span id="cb86-54"><a href="#cb86-54" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rpart)</span>
<span id="cb86-55"><a href="#cb86-55" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rattle)</span>
<span id="cb86-56"><a href="#cb86-56" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(wooldridge)</span>
<span id="cb86-57"><a href="#cb86-57" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-58"><a href="#cb86-58" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb86-59"><a href="#cb86-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-60"><a href="#cb86-60" aria-hidden="true" tabindex="-1"></a><span class="fu">## Random forest as a local constant regression</span></span>
<span id="cb86-61"><a href="#cb86-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-62"><a href="#cb86-62" aria-hidden="true" tabindex="-1"></a>Suppose you are interested in estimating $E<span class="co">[</span><span class="ot">y|X</span><span class="co">]</span>$ using a dataset and you have trained a random forest model with $T$ tress. Now, let $\eta_{i,t}(X)$ takes $1$ if observation $i$ belongs to the same leaf as $X$ in tree $t$, where $X$ is a vector of covariates ($K$ variables). Then, the RF's predicted value of $y$ conditional on a particular value of $X$ (say, $X_0$) can be written as follows:</span>
<span id="cb86-63"><a href="#cb86-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-64"><a href="#cb86-64" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-65"><a href="#cb86-65" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb86-66"><a href="#cb86-66" aria-hidden="true" tabindex="-1"></a>\hat{y}(X_0) = \frac{1}{T} \cdot\sum_{t=1}^T\sum_{i=1}^N \frac{\eta_{i,t}(X_0)}{\sum_{i=1}^N\eta_{i,t}(X_0)}\cdot y_i</span>
<span id="cb86-67"><a href="#cb86-67" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb86-68"><a href="#cb86-68" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-69"><a href="#cb86-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-70"><a href="#cb86-70" aria-hidden="true" tabindex="-1"></a>Note that $\sum_{i=1}^N\eta_{i,t}(X_0)$ represents the number of observations in the same leaf as $X_0$. Therefore, $\sum_{i=1}^N \frac{\eta_{i,t}(X_0)}{\sum_{i=1}^N\eta_{i,t}(X_0)}\cdot y_i$ is the average value of $y$ of the leaf that $X_0$ belongs to. So, while looking slightly complicated, it is the average value of $y$ of the tree $X_0$ belongs to averaged across the trees, which we know is how RF predicts $y$ at $X_0$.</span>
<span id="cb86-71"><a href="#cb86-71" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-72"><a href="#cb86-72" aria-hidden="true" tabindex="-1"></a>We can switch the summations like this,</span>
<span id="cb86-73"><a href="#cb86-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-74"><a href="#cb86-74" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-75"><a href="#cb86-75" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb86-76"><a href="#cb86-76" aria-hidden="true" tabindex="-1"></a>\hat{y}(X_0) = \sum_{i=1}^N \cdot\frac{1}{T}\sum_{t=1}^T\frac{\eta_{i,t}(X_0)}{\sum_{i=1}^N\eta_{i,t}(X_0)}\cdot y_i</span>
<span id="cb86-77"><a href="#cb86-77" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb86-78"><a href="#cb86-78" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-79"><a href="#cb86-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-80"><a href="#cb86-80" aria-hidden="true" tabindex="-1"></a>Let $\alpha(X_i, X_0)$ denote $\frac{1}{T}\sum_{t=1}^T\frac{\eta_{i,t}(X_0)}{\sum_{i=1}^N\eta_{i,t}(X_0)}$. Then, we can rewrite the above equation as</span>
<span id="cb86-81"><a href="#cb86-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-82"><a href="#cb86-82" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-83"><a href="#cb86-83" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb86-84"><a href="#cb86-84" aria-hidden="true" tabindex="-1"></a>\hat{y}(X_0) = \sum_{i=1}^N \alpha(X_i,X_0) \cdot y_i</span>
<span id="cb86-85"><a href="#cb86-85" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb86-86"><a href="#cb86-86" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-87"><a href="#cb86-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-88"><a href="#cb86-88" aria-hidden="true" tabindex="-1"></a>Now, it is easy to show that $\hat{y}(X_0)$ is a solution to the following minimization problem.</span>
<span id="cb86-89"><a href="#cb86-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-90"><a href="#cb86-90" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-91"><a href="#cb86-91" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb86-92"><a href="#cb86-92" aria-hidden="true" tabindex="-1"></a>Min_{\theta} \sum_{i=1}^N \alpha(X_i,X_0)\cdot<span class="co">[</span><span class="ot">y_i -\theta</span><span class="co">]</span>^2</span>
<span id="cb86-93"><a href="#cb86-93" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb86-94"><a href="#cb86-94" aria-hidden="true" tabindex="-1"></a>$$ {#eq-ll-constant}</span>
<span id="cb86-95"><a href="#cb86-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-96"><a href="#cb86-96" aria-hidden="true" tabindex="-1"></a>In this formulation of the problem, $\alpha(X_i,X_0)$ can be considered the weight given to observation $i$. </span>
<span id="cb86-97"><a href="#cb86-97" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-98"><a href="#cb86-98" aria-hidden="true" tabindex="-1"></a>By definition,</span>
<span id="cb86-99"><a href="#cb86-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-100"><a href="#cb86-100" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>$0 \leq \alpha(X_i,X_0) \leq 1$</span>
<span id="cb86-101"><a href="#cb86-101" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>$\sum_{i=1}^N \alpha(X_i,X_0) = 1$</span>
<span id="cb86-102"><a href="#cb86-102" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-103"><a href="#cb86-103" aria-hidden="true" tabindex="-1"></a>You may notice that @eq-ll-constant is actually a special case of <span class="kw">&lt;span</span> <span class="er">style</span><span class="ot">=</span><span class="st">"color:blue"</span><span class="kw">&gt;</span>local constant regression<span class="kw">&lt;/span&gt;</span> where the individual weights are $\alpha(X_i, X_0)$. Roughly speaking, $\alpha(X_i, X_0)$ measures how often observation $i$ share the same leaves as the evaluation point ($X_0$) across $T$ trees. So, it measures how similar $X_i$ is to $X_0$ in the RF way, but not based on euclidean distance (which is subject to curse of dimensionality). So, RF is actually a local constant regression with a special way of distributing weights to the individual observations. </span>
<span id="cb86-104"><a href="#cb86-104" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-105"><a href="#cb86-105" aria-hidden="true" tabindex="-1"></a><span class="fu">## GRF</span></span>
<span id="cb86-106"><a href="#cb86-106" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-107"><a href="#cb86-107" aria-hidden="true" tabindex="-1"></a>Interpretation of RF as a local regression led @athey2019generalized to conceive GRF, under which various statistics (e.g., conditional average treatment effect, conditional quantile) can be estimated under the unified framework.</span>
<span id="cb86-108"><a href="#cb86-108" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-109"><a href="#cb86-109" aria-hidden="true" tabindex="-1"></a>You can consider prediction of $y$ at $X = X_0$ using RF as a three-step process.</span>
<span id="cb86-110"><a href="#cb86-110" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-111"><a href="#cb86-111" aria-hidden="true" tabindex="-1"></a>:::{.callout-note}</span>
<span id="cb86-112"><a href="#cb86-112" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-113"><a href="#cb86-113" aria-hidden="true" tabindex="-1"></a><span class="fu">## Predicting $y$ at $X=X_0$ using RF</span></span>
<span id="cb86-114"><a href="#cb86-114" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-115"><a href="#cb86-115" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>Grow trees (find splitting rules for each tree)</span>
<span id="cb86-116"><a href="#cb86-116" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>For $X = X_0$, find the weights for individual observations based on the trees grown  </span>
<span id="cb86-117"><a href="#cb86-117" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>Solve @eq-ll-constant based on the weights</span>
<span id="cb86-118"><a href="#cb86-118" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb86-119"><a href="#cb86-119" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-120"><a href="#cb86-120" aria-hidden="true" tabindex="-1"></a>Step 1 is done only once. Every time you make a prediction at a different value of $X$, you go over steps 2 and 3. </span>
<span id="cb86-121"><a href="#cb86-121" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-122"><a href="#cb86-122" aria-hidden="true" tabindex="-1"></a>GRF follows exactly the same steps except that it <span class="kw">&lt;span</span> <span class="er">style</span><span class="ot">=</span><span class="st">"color:red"</span><span class="kw">&gt;</span>adjusts the way trees are grown <span class="kw">&lt;/span&gt;</span>in step 1 and <span class="kw">&lt;span</span> <span class="er">style</span><span class="ot">=</span><span class="st">"color:red"</span><span class="kw">&gt;</span>adjusts the way the local minimization problem<span class="kw">&lt;/span&gt;</span> is solved at step 3 (they are actually interrelated) depending on what you would like to estimate.</span>
<span id="cb86-123"><a href="#cb86-123" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-124"><a href="#cb86-124" aria-hidden="true" tabindex="-1"></a>Here, we follow the notations used in @athey2019generalized as much as possible. Here are the list of notations:</span>
<span id="cb86-125"><a href="#cb86-125" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-126"><a href="#cb86-126" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>$O_i$: data for observation $i$</span>
<span id="cb86-127"><a href="#cb86-127" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>$\theta$: the statistics of interest (e.g., CATE for causal forest, conditional quantile for for quantile forest)</span>
<span id="cb86-128"><a href="#cb86-128" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>$\nu$: nuisance statistics (you are not interested in estimating this). </span>
<span id="cb86-129"><a href="#cb86-129" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>$\Psi_{\theta, \nu}(O)$: score function</span>
<span id="cb86-130"><a href="#cb86-130" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>$\alpha_i(x)$: weight given to observation $i$ when predicting at $X=x$. </span>
<span id="cb86-131"><a href="#cb86-131" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-132"><a href="#cb86-132" aria-hidden="true" tabindex="-1"></a>In general, GRF solves the following problem to find the estimate of $\theta$ conditional on $X_i= x$: </span>
<span id="cb86-133"><a href="#cb86-133" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-134"><a href="#cb86-134" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-135"><a href="#cb86-135" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb86-136"><a href="#cb86-136" aria-hidden="true" tabindex="-1"></a>\sum_{i=1}^n \alpha_i(x)\Psi_{\theta, \nu}(O_i) = 0</span>
<span id="cb86-137"><a href="#cb86-137" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb86-138"><a href="#cb86-138" aria-hidden="true" tabindex="-1"></a>$$ {#eq-opt}</span>
<span id="cb86-139"><a href="#cb86-139" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-140"><a href="#cb86-140" aria-hidden="true" tabindex="-1"></a>As we saw earlier, GRF is RF when $\Psi_{\theta, \nu}(O_i) = Y_i-\theta$. There is no nuisance statistics, $\nu(x)$, in the RF case. By changing how the score function ($\Psi_{\theta, \nu}(O_i)$) is defined, you can estimate <span class="kw">&lt;span</span> <span class="er">style</span><span class="ot">=</span><span class="st">"color:blue"</span><span class="kw">&gt;</span> a wide range of statistics using different approaches <span class="kw">&lt;/span&gt;</span>under the <span class="kw">&lt;span</span> <span class="er">style</span><span class="ot">=</span><span class="st">"color:blue"</span><span class="kw">&gt;</span> same<span class="kw">&lt;/span&gt;</span> estimation framework (this is why it is called <span class="kw">&lt;span</span> <span class="er">style</span><span class="ot">=</span><span class="st">"color:red"</span><span class="kw">&gt;</span>generalized <span class="kw">&lt;/span&gt;</span>random forest). Here are some of the statistics and estimation approaches that are under the GRF framework.</span>
<span id="cb86-141"><a href="#cb86-141" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-142"><a href="#cb86-142" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>Conditional expectation ($E<span class="co">[</span><span class="ot">Y|X</span><span class="co">]</span>$)</span>
<span id="cb86-143"><a href="#cb86-143" aria-hidden="true" tabindex="-1"></a><span class="ss">  * </span>Regression forest (Random forest for regression)</span>
<span id="cb86-144"><a href="#cb86-144" aria-hidden="true" tabindex="-1"></a><span class="ss">  * </span>Local linear forest</span>
<span id="cb86-145"><a href="#cb86-145" aria-hidden="true" tabindex="-1"></a><span class="ss">  * </span>Boosted regression forest</span>
<span id="cb86-146"><a href="#cb86-146" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>Conditional average treatment effect (CATE)</span>
<span id="cb86-147"><a href="#cb86-147" aria-hidden="true" tabindex="-1"></a><span class="ss">  * </span>Causal forest</span>
<span id="cb86-148"><a href="#cb86-148" aria-hidden="true" tabindex="-1"></a><span class="ss">  * </span>Instrumental forest</span>
<span id="cb86-149"><a href="#cb86-149" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>Conditional quantile</span>
<span id="cb86-150"><a href="#cb86-150" aria-hidden="true" tabindex="-1"></a><span class="ss">  * </span>Quantile forest</span>
<span id="cb86-151"><a href="#cb86-151" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-152"><a href="#cb86-152" aria-hidden="true" tabindex="-1"></a>:::{.callout-note}</span>
<span id="cb86-153"><a href="#cb86-153" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-154"><a href="#cb86-154" aria-hidden="true" tabindex="-1"></a><span class="fu">## Score function examples</span></span>
<span id="cb86-155"><a href="#cb86-155" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-156"><a href="#cb86-156" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>Random forest: $\Psi_{\theta, \nu}(O_i) = Y_i - \theta$</span>
<span id="cb86-157"><a href="#cb86-157" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-158"><a href="#cb86-158" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>Causal forest: $\Psi_{\theta, \nu}(O_i) = (\tilde{Y}_i- \theta\tilde{T}_i)\cdot\tilde{T}_i$, where $\tilde{v_i} = v_i - E<span class="co">[</span><span class="ot">v_i|X_i</span><span class="co">]</span>$</span>
<span id="cb86-159"><a href="#cb86-159" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-160"><a href="#cb86-160" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>Quantile forest: $\Psi_{\theta, \nu}(O_i) = qI<span class="sc">\{</span>Y_i &gt; \theta<span class="sc">\}</span> - (1-q)I<span class="sc">\{</span>Y_i \leq \theta<span class="sc">\}</span>$</span>
<span id="cb86-161"><a href="#cb86-161" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-162"><a href="#cb86-162" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb86-163"><a href="#cb86-163" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-164"><a href="#cb86-164" aria-hidden="true" tabindex="-1"></a>::: {.column-margin}</span>
<span id="cb86-165"><a href="#cb86-165" aria-hidden="true" tabindex="-1"></a>$I<span class="sc">\{\}</span>$ is an indicator function that takes 1 if the condition inside the curly brackets and 0 otherwise.</span>
<span id="cb86-166"><a href="#cb86-166" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb86-167"><a href="#cb86-167" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-168"><a href="#cb86-168" aria-hidden="true" tabindex="-1"></a>---</span>
<span id="cb86-169"><a href="#cb86-169" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-170"><a href="#cb86-170" aria-hidden="true" tabindex="-1"></a>So far, we have only talked about score functions so far, but not the weights. Do all the approaches listed above use the weights derived from the trees grown by the traditional RF in solving @eq-opt? You could (you are free to use any weights), but that would not be wise. As mentioned earlier, GRF adjusts the way trees are grown (thus weights) as well according to the score function so that weights are optimized for your objective. This makes sense. Right <span class="kw">&lt;span</span> <span class="er">style</span><span class="ot">=</span><span class="st">"color:blue"</span><span class="kw">&gt;</span> neighbors <span class="kw">&lt;/span&gt;</span>should be different based on what you are interesting in estimating.</span>
<span id="cb86-171"><a href="#cb86-171" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-172"><a href="#cb86-172" aria-hidden="true" tabindex="-1"></a>Specifically, GRF uses the random forest <span class="kw">&lt;span</span> <span class="er">style</span><span class="ot">=</span><span class="st">"color:blue"</span><span class="kw">&gt;</span>algorithm<span class="kw">&lt;/span&gt;</span> to grow trees based on <span class="kw">&lt;span</span> <span class="er">style</span><span class="ot">=</span><span class="st">"color:blue"</span><span class="kw">&gt;</span> pseudo outcome ($\rho_i$) <span class="kw">&lt;/span&gt;</span> derived from the score function that is specific to the type of regression you are running, but not on $Y$. Basically, you are using exactly the same algorithm as the traditional RF we saw in @sec-rf except that $Y$ is swapped with the pseudo outcome.</span>
<span id="cb86-173"><a href="#cb86-173" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-174"><a href="#cb86-174" aria-hidden="true" tabindex="-1"></a>::: {.column-margin}</span>
<span id="cb86-175"><a href="#cb86-175" aria-hidden="true" tabindex="-1"></a>See @athey2019generalized for how the pseudo outcome is derived from a score function in general. </span>
<span id="cb86-176"><a href="#cb86-176" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb86-177"><a href="#cb86-177" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-178"><a href="#cb86-178" aria-hidden="true" tabindex="-1"></a>:::{.callout-note}</span>
<span id="cb86-179"><a href="#cb86-179" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-180"><a href="#cb86-180" aria-hidden="true" tabindex="-1"></a><span class="fu">## Example pseudo outcomes</span></span>
<span id="cb86-181"><a href="#cb86-181" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-182"><a href="#cb86-182" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;span</span> <span class="er">style</span><span class="ot">=</span><span class="st">"color:blue"</span><span class="kw">&gt;</span>Random forest<span class="kw">&lt;/span&gt;</span>: </span>
<span id="cb86-183"><a href="#cb86-183" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-184"><a href="#cb86-184" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-185"><a href="#cb86-185" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb86-186"><a href="#cb86-186" aria-hidden="true" tabindex="-1"></a>\rho_i = Y_i - \hat{\theta}_P</span>
<span id="cb86-187"><a href="#cb86-187" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb86-188"><a href="#cb86-188" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-189"><a href="#cb86-189" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-190"><a href="#cb86-190" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;span</span> <span class="er">style</span><span class="ot">=</span><span class="st">"color:blue"</span><span class="kw">&gt;</span> Causal forest<span class="kw">&lt;/span&gt;</span>:</span>
<span id="cb86-191"><a href="#cb86-191" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-192"><a href="#cb86-192" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-193"><a href="#cb86-193" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb86-194"><a href="#cb86-194" aria-hidden="true" tabindex="-1"></a>\rho_i = (\tilde{Y}_i- \hat{\theta}_P \tilde{T}_i)\cdot\tilde{T}_i</span>
<span id="cb86-195"><a href="#cb86-195" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb86-196"><a href="#cb86-196" aria-hidden="true" tabindex="-1"></a>$$ </span>
<span id="cb86-197"><a href="#cb86-197" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-198"><a href="#cb86-198" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;span</span> <span class="er">style</span><span class="ot">=</span><span class="st">"color:blue"</span><span class="kw">&gt;</span> Quantile forest<span class="kw">&lt;/span&gt;</span>:</span>
<span id="cb86-199"><a href="#cb86-199" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-200"><a href="#cb86-200" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-201"><a href="#cb86-201" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb86-202"><a href="#cb86-202" aria-hidden="true" tabindex="-1"></a>\rho_i = I<span class="sc">\{</span>Y_i &gt; \hat{\theta}_P<span class="sc">\}</span></span>
<span id="cb86-203"><a href="#cb86-203" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb86-204"><a href="#cb86-204" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-205"><a href="#cb86-205" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-206"><a href="#cb86-206" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb86-207"><a href="#cb86-207" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-208"><a href="#cb86-208" aria-hidden="true" tabindex="-1"></a>Note that $\hat{\theta}_P$ in the pseudo outcomes presented above is the solution to @eq-opt with their respective score functions using the data in the parent node. For example, $\hat{\theta}_P = \bar{Y}_p$ for RF, which is the average value of $Y$ in the parent node. In quantile forest, $\hat{\theta}_P$ is the $q$th quantile of the parent node if you are estimating the $q$the conditional quantile.</span>
<span id="cb86-209"><a href="#cb86-209" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-210"><a href="#cb86-210" aria-hidden="true" tabindex="-1"></a>---</span>
<span id="cb86-211"><a href="#cb86-211" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-212"><a href="#cb86-212" aria-hidden="true" tabindex="-1"></a>Finally, here are the conceptual steps of GRF:</span>
<span id="cb86-213"><a href="#cb86-213" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-214"><a href="#cb86-214" aria-hidden="true" tabindex="-1"></a>:::{.callout-note}</span>
<span id="cb86-215"><a href="#cb86-215" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-216"><a href="#cb86-216" aria-hidden="true" tabindex="-1"></a><span class="fu">## GRF: conceptual steps </span></span>
<span id="cb86-217"><a href="#cb86-217" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-218"><a href="#cb86-218" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>Training (Forest Growing) Step:</span>
<span id="cb86-219"><a href="#cb86-219" aria-hidden="true" tabindex="-1"></a><span class="ss">  1. </span>Determine what statistics you are interested in ($\theta(X)$, e.g., CATE, conditional quantile)</span>
<span id="cb86-220"><a href="#cb86-220" aria-hidden="true" tabindex="-1"></a><span class="ss">  2. </span>Define the appropriate score function according to the statistics of interest</span>
<span id="cb86-221"><a href="#cb86-221" aria-hidden="true" tabindex="-1"></a><span class="ss">  3. </span>Derive the pseudo outcome based on the score function</span>
<span id="cb86-222"><a href="#cb86-222" aria-hidden="true" tabindex="-1"></a><span class="ss">  4. </span>Grow trees using the RF algorithm based on the pseudo outcome</span>
<span id="cb86-223"><a href="#cb86-223" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-224"><a href="#cb86-224" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>Prediction Step:</span>
<span id="cb86-225"><a href="#cb86-225" aria-hidden="true" tabindex="-1"></a><span class="ss">  1. </span>Determine what value of $X$ you would like to predict $\theta(X)$ (call it $X_0$)</span>
<span id="cb86-226"><a href="#cb86-226" aria-hidden="true" tabindex="-1"></a><span class="ss">  2. </span>Find the individual weights $\alpha(X_i, X_0)$ according to </span>
<span id="cb86-227"><a href="#cb86-227" aria-hidden="true" tabindex="-1"></a>  $$</span>
<span id="cb86-228"><a href="#cb86-228" aria-hidden="true" tabindex="-1"></a>  \begin{aligned}</span>
<span id="cb86-229"><a href="#cb86-229" aria-hidden="true" tabindex="-1"></a>  \alpha(X_i, X_0) = \frac{1}{T}\sum_{t=1}^T\frac{\eta_{i,t}(X_0)}{\sum_{i=1}^N\eta_{i,t}(X_0)}</span>
<span id="cb86-230"><a href="#cb86-230" aria-hidden="true" tabindex="-1"></a>  \end{aligned}</span>
<span id="cb86-231"><a href="#cb86-231" aria-hidden="true" tabindex="-1"></a>  $$</span>
<span id="cb86-232"><a href="#cb86-232" aria-hidden="true" tabindex="-1"></a>  , which measures how often observation $i$ belongs to the same node as the evaluation point $X_0$.</span>
<span id="cb86-233"><a href="#cb86-233" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-234"><a href="#cb86-234" aria-hidden="true" tabindex="-1"></a><span class="ss">  3. </span>Solve @eq-opt with the weights obtained above and the score function specified earlier</span>
<span id="cb86-235"><a href="#cb86-235" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-236"><a href="#cb86-236" aria-hidden="true" tabindex="-1"></a>::: </span>
<span id="cb86-237"><a href="#cb86-237" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-238"><a href="#cb86-238" aria-hidden="true" tabindex="-1"></a>The training step is done only once (trees are built only once). But, whenever you predict $\theta(X)$ at different values of $X$, you go through the prediction step.</span>
<span id="cb86-239"><a href="#cb86-239" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-240"><a href="#cb86-240" aria-hidden="true" tabindex="-1"></a>::: {.column-margin}</span>
<span id="cb86-241"><a href="#cb86-241" aria-hidden="true" tabindex="-1"></a>Orthogonal random forest (a forest-based heterogeneous treatment effect estimator like causal forest), on the other hand, build trees every time when predicting treatment effect $\theta(X)$ at particular values of $X$, which is why orthogonal random forest takes a very long time especially when there are many evaluation points. Orthogonal random forest is covered in @sec-cf-orf.</span>
<span id="cb86-242"><a href="#cb86-242" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb86-243"><a href="#cb86-243" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-244"><a href="#cb86-244" aria-hidden="true" tabindex="-1"></a><span class="fu">## Examples of GRF </span></span>
<span id="cb86-245"><a href="#cb86-245" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-246"><a href="#cb86-246" aria-hidden="true" tabindex="-1"></a><span class="fu">### Random forest as a GRF</span></span>
<span id="cb86-247"><a href="#cb86-247" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-248"><a href="#cb86-248" aria-hidden="true" tabindex="-1"></a>Here, we take a look at RF as a GRF as an illustration to understand the general GRF procedure better. When $\Psi_{\theta, \nu}(Y_i)$ is set to $Y_i - \theta$, then GRF is equivalent to the traditional RF. By plugging $Y_i - \theta$ into @eq-opt, the estimate of $E<span class="co">[</span><span class="ot">Y|X=X_0</span><span class="co">]</span>$, $\theta(X_0)$, is identified by solving </span>
<span id="cb86-249"><a href="#cb86-249" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-250"><a href="#cb86-250" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-251"><a href="#cb86-251" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb86-252"><a href="#cb86-252" aria-hidden="true" tabindex="-1"></a>\sum_{i=1}^n \alpha_i(X_0)(Y_i - \theta) = 0</span>
<span id="cb86-253"><a href="#cb86-253" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb86-254"><a href="#cb86-254" aria-hidden="true" tabindex="-1"></a>$$ {#eq-rf}</span>
<span id="cb86-255"><a href="#cb86-255" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-256"><a href="#cb86-256" aria-hidden="true" tabindex="-1"></a>The weights $\alpha_i(X_i, X_0)$ are defined as follows:</span>
<span id="cb86-257"><a href="#cb86-257" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-258"><a href="#cb86-258" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb86-259"><a href="#cb86-259" aria-hidden="true" tabindex="-1"></a>\alpha_i(X_i, X_0) = \frac{1}{T}\cdot\sum_{t=1}^T\frac{\eta_{i,t}(X_0)}{\sum_{i=1}^N\eta_{i,t}(X_0)}</span>
<span id="cb86-260"><a href="#cb86-260" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb86-261"><a href="#cb86-261" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-262"><a href="#cb86-262" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-263"><a href="#cb86-263" aria-hidden="true" tabindex="-1"></a>, where $\eta_{i,t}(X_0)$ is 1 if observation $i$ which has feature values $X_i$ belongs to the same leaf as $X_0$. </span>
<span id="cb86-264"><a href="#cb86-264" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-265"><a href="#cb86-265" aria-hidden="true" tabindex="-1"></a>---</span>
<span id="cb86-266"><a href="#cb86-266" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-267"><a href="#cb86-267" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;span</span> <span class="er">style</span><span class="ot">=</span><span class="st">"color:blue"</span><span class="kw">&gt;</span> Step 1: Grow trees<span class="kw">&lt;/span&gt;</span>:</span>
<span id="cb86-268"><a href="#cb86-268" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-269"><a href="#cb86-269" aria-hidden="true" tabindex="-1"></a>Now, let's consider growing trees to find $\alpha_i(X_0)$ in @eq-rf. For a given sample and set of variables randomly selected, GRF starts with solving the unweighted version of @eq-rf. </span>
<span id="cb86-270"><a href="#cb86-270" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-271"><a href="#cb86-271" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-272"><a href="#cb86-272" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb86-273"><a href="#cb86-273" aria-hidden="true" tabindex="-1"></a>\sum_{i=1}^n Y_i - \theta = 0</span>
<span id="cb86-274"><a href="#cb86-274" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb86-275"><a href="#cb86-275" aria-hidden="true" tabindex="-1"></a>$$ {#eq-rf-initial}</span>
<span id="cb86-276"><a href="#cb86-276" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-277"><a href="#cb86-277" aria-hidden="true" tabindex="-1"></a>The solution to this problem is simply the mean of $Y$, which will be denoted as $\bar{Y}_P$, where $P$ represents the parent node. Here, the parent node include all the data points as this is the first split.</span>
<span id="cb86-278"><a href="#cb86-278" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-279"><a href="#cb86-279" aria-hidden="true" tabindex="-1"></a>Then the pseudo outcome ($\rho_i$) that is used in splitting is</span>
<span id="cb86-280"><a href="#cb86-280" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-281"><a href="#cb86-281" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-282"><a href="#cb86-282" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb86-283"><a href="#cb86-283" aria-hidden="true" tabindex="-1"></a>\rho_i = Y_i - \bar{Y}_P</span>
<span id="cb86-284"><a href="#cb86-284" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb86-285"><a href="#cb86-285" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-286"><a href="#cb86-286" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-287"><a href="#cb86-287" aria-hidden="true" tabindex="-1"></a>Now, a standard CART regression split is applied on the pseudo outcomes. That is, the variable-threshold combination that maximizes the following criteria is found:</span>
<span id="cb86-288"><a href="#cb86-288" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-289"><a href="#cb86-289" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-290"><a href="#cb86-290" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb86-291"><a href="#cb86-291" aria-hidden="true" tabindex="-1"></a>\tilde{\Delta}(C_1, C_2) = \frac{(\sum_{i \in C_1} \rho_i)^2}{N_{C_1}} + \frac{(\sum_{i \in C_2} \rho_i)^2}{N_{C_2}}</span>
<span id="cb86-292"><a href="#cb86-292" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb86-293"><a href="#cb86-293" aria-hidden="true" tabindex="-1"></a>$$ {#eq-criteria}</span>
<span id="cb86-294"><a href="#cb86-294" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-295"><a href="#cb86-295" aria-hidden="true" tabindex="-1"></a>where $C_1$ and $C_2$ represent two child node candidates for a given split. Since $\bar{Y}_P$ is just a constant, it is equivalent to splitting on $Y_i$. So, this is exactly the same as how the traditional RF builds trees (see @sec-split-sim for the rationale behind maximizing the criteria presented in @eq-criteria).</span>
<span id="cb86-296"><a href="#cb86-296" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-297"><a href="#cb86-297" aria-hidden="true" tabindex="-1"></a>Once the best split is identified, each of the new child nodes is split following the exactly the same procedure. Splitting continues until one of the user-specified condition prevent a further splitting. Many trees from bootstrapped samples are created (just like the regular random forest) and they form a forest. This shows that the GRF with $\Psi_{\theta, \nu}(Y_i, X_i) = Y_i - \theta$ grows trees in the same manner as the traditional RF. RF in GRF is implemented by <span class="in">`regression_forest()`</span>. But, note that running <span class="in">`ranger()`</span> and <span class="in">`regression_forest()`</span> will not result in the same forest because of the randomness involved in resampling and random selection of variables. Only their algorithms are equivalent</span>
<span id="cb86-298"><a href="#cb86-298" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-299"><a href="#cb86-299" aria-hidden="true" tabindex="-1"></a>---</span>
<span id="cb86-300"><a href="#cb86-300" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;span</span> <span class="er">style</span><span class="ot">=</span><span class="st">"color:blue"</span><span class="kw">&gt;</span> Step 2: Predict<span class="kw">&lt;/span&gt;</span>:</span>
<span id="cb86-301"><a href="#cb86-301" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-302"><a href="#cb86-302" aria-hidden="true" tabindex="-1"></a>To predict $E<span class="co">[</span><span class="ot">Y|X=X_0</span><span class="co">]</span>$, @eq-rf is solved</span>
<span id="cb86-303"><a href="#cb86-303" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-304"><a href="#cb86-304" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-305"><a href="#cb86-305" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb86-306"><a href="#cb86-306" aria-hidden="true" tabindex="-1"></a>\sum_{i=1}^N \alpha_i(X_i, X_0)(Y_i-\theta) = 0</span>
<span id="cb86-307"><a href="#cb86-307" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb86-308"><a href="#cb86-308" aria-hidden="true" tabindex="-1"></a>$$  </span>
<span id="cb86-309"><a href="#cb86-309" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-310"><a href="#cb86-310" aria-hidden="true" tabindex="-1"></a>So,</span>
<span id="cb86-311"><a href="#cb86-311" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-312"><a href="#cb86-312" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-313"><a href="#cb86-313" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb86-314"><a href="#cb86-314" aria-hidden="true" tabindex="-1"></a>\theta(X_0) &amp; = \frac{\sum_{i=1}^N \alpha_i(X_0)Y_i}{\sum_{i=1}^N \alpha_i(X_0)}<span class="sc">\\</span></span>
<span id="cb86-315"><a href="#cb86-315" aria-hidden="true" tabindex="-1"></a>&amp; = \sum_{i=1}^N \alpha_i(X_0)Y_i \;\; \mbox{(since } \sum_{i=1}^N \alpha_i(X_0) = 1\mbox{)} <span class="sc">\\</span></span>
<span id="cb86-316"><a href="#cb86-316" aria-hidden="true" tabindex="-1"></a>&amp; = \sum_{i=1}^N \huge<span class="co">[</span><span class="ot">\normalsize \frac{1}{T}\cdot\sum_{t=1}^T\frac{\eta_{i,t}(X_0)}{\sum_{i=1}^N\eta_{i,t}(X_0)}\cdot y_i\huge</span><span class="co">]</span><span class="sc">\\</span></span>
<span id="cb86-317"><a href="#cb86-317" aria-hidden="true" tabindex="-1"></a>&amp; = \frac{1}{T}  \cdot\sum_{t=1}^T\sum_{i=1}^N \frac{\eta_{i,t}(X_0)}{\sum_{i=1}^N\eta_{i,t}(X_0)}\cdot y_i \;\; \mbox{(changing the order of the summations)} <span class="sc">\\</span></span>
<span id="cb86-318"><a href="#cb86-318" aria-hidden="true" tabindex="-1"></a>&amp; = \frac{1}{T} \cdot\sum_{t=1}^T \bar{Y}_t</span>
<span id="cb86-319"><a href="#cb86-319" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb86-320"><a href="#cb86-320" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-321"><a href="#cb86-321" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-322"><a href="#cb86-322" aria-hidden="true" tabindex="-1"></a>So, $\theta(X_0)$ from GRF is the average of tree-specific predictions for $X_0$, which is exactly how RF predicts $E<span class="co">[</span><span class="ot">Y|X=X_0</span><span class="co">]</span>$ as well.</span>
<span id="cb86-323"><a href="#cb86-323" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-324"><a href="#cb86-324" aria-hidden="true" tabindex="-1"></a>So, it has been shown that GRF with $\Psi_{\theta, \nu}(Y_i)=Y_i - \theta$ grows trees in the same manner as the traditional RF and also that GRF predicts $E<span class="co">[</span><span class="ot">Y|X=X_0</span><span class="co">]</span>$ in the same manner as the traditional RF. So, RF is a special case of GRF, where $\Psi_{\theta, \nu}(Y_i)=Y_i - \theta$.</span>
<span id="cb86-325"><a href="#cb86-325" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-326"><a href="#cb86-326" aria-hidden="true" tabindex="-1"></a><span class="fu">### Causal forest as a GRF</span></span>
<span id="cb86-327"><a href="#cb86-327" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-328"><a href="#cb86-328" aria-hidden="true" tabindex="-1"></a>Causal forest (with a single treatment variable) as an R-learner is a GRF with</span>
<span id="cb86-329"><a href="#cb86-329" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-330"><a href="#cb86-330" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-331"><a href="#cb86-331" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb86-332"><a href="#cb86-332" aria-hidden="true" tabindex="-1"></a>\Psi_{\theta, \nu}(O_i) = <span class="co">[</span><span class="ot">(Y_i - E[Y|X_i])- \theta(X_i)(T_i - E[T|X_i])</span><span class="co">](T_i - E[T|X_i])</span></span>
<span id="cb86-333"><a href="#cb86-333" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb86-334"><a href="#cb86-334" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-335"><a href="#cb86-335" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-336"><a href="#cb86-336" aria-hidden="true" tabindex="-1"></a>In practice $E<span class="co">[</span><span class="ot">Y|X_i</span><span class="co">]</span>$ and $E<span class="co">[</span><span class="ot">T|X_i</span><span class="co">]</span>$ are first estimated using appropriate machine learning methods (e.g., lasso, random forest, gradient boosted forest) in a cross-fitting manner and then the estimation of $Y_i - E<span class="co">[</span><span class="ot">Y|X_i</span><span class="co">]</span>$ and $T_i - E<span class="co">[</span><span class="ot">T|X_i</span><span class="co">]</span>$ are constructed. Let's denote them by $\tilde{Y}_i$ and $\tilde{T}_i$. Then the score function is written as</span>
<span id="cb86-337"><a href="#cb86-337" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-338"><a href="#cb86-338" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-339"><a href="#cb86-339" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb86-340"><a href="#cb86-340" aria-hidden="true" tabindex="-1"></a>\Psi_{\theta} = (\tilde{Y}_i- \theta\tilde{T}_i)\tilde{T}_i</span>
<span id="cb86-341"><a href="#cb86-341" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb86-342"><a href="#cb86-342" aria-hidden="true" tabindex="-1"></a>$$ {#eq-cf-score}</span>
<span id="cb86-343"><a href="#cb86-343" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-344"><a href="#cb86-344" aria-hidden="true" tabindex="-1"></a>So, the conditional treatment effect at $X=X_0$, $\theta(X_0)$, is estimated by solving the following problem:</span>
<span id="cb86-345"><a href="#cb86-345" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-346"><a href="#cb86-346" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-347"><a href="#cb86-347" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb86-348"><a href="#cb86-348" aria-hidden="true" tabindex="-1"></a>\hat{\theta}(X_0) = \sum_{i=1}^N \alpha_i(x)(\tilde{Y_i} - \theta\cdot \tilde{T_i})\tilde{T_i} = 0</span>
<span id="cb86-349"><a href="#cb86-349" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb86-350"><a href="#cb86-350" aria-hidden="true" tabindex="-1"></a>$$ {#eq-cf-solve}</span>
<span id="cb86-351"><a href="#cb86-351" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-352"><a href="#cb86-352" aria-hidden="true" tabindex="-1"></a>where $\alpha_i(x)$ is the weight obtained from the trees built using random forest on the pseudo outcomes that are derived from the score function (@eq-cf-score).</span>
<span id="cb86-353"><a href="#cb86-353" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-354"><a href="#cb86-354" aria-hidden="true" tabindex="-1"></a>For a given parent node $p$, the pseudo outcomes are defined as</span>
<span id="cb86-355"><a href="#cb86-355" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-356"><a href="#cb86-356" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-357"><a href="#cb86-357" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb86-358"><a href="#cb86-358" aria-hidden="true" tabindex="-1"></a>\rho_i = (\tilde{Y}_i- \hat{\theta}_P\tilde{T}_i)\tilde{T}_i</span>
<span id="cb86-359"><a href="#cb86-359" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb86-360"><a href="#cb86-360" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-361"><a href="#cb86-361" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-362"><a href="#cb86-362" aria-hidden="true" tabindex="-1"></a>where $\hat{\theta}_P$ is the solution of the following problem using the observations in the parent node:</span>
<span id="cb86-363"><a href="#cb86-363" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-364"><a href="#cb86-364" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-365"><a href="#cb86-365" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb86-366"><a href="#cb86-366" aria-hidden="true" tabindex="-1"></a>\sum_{i=1}^N \tilde{T_i}(\tilde{Y_i}-\theta \tilde{T_i}) = 0</span>
<span id="cb86-367"><a href="#cb86-367" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb86-368"><a href="#cb86-368" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-369"><a href="#cb86-369" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-370"><a href="#cb86-370" aria-hidden="true" tabindex="-1"></a>, which is the unweighted version of @eq-cf-solve.</span>
<span id="cb86-371"><a href="#cb86-371" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-372"><a href="#cb86-372" aria-hidden="true" tabindex="-1"></a>The standard CART splitting algorithm is applied on the pseudo outcomes to build trees.</span>
<span id="cb86-373"><a href="#cb86-373" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-374"><a href="#cb86-374" aria-hidden="true" tabindex="-1"></a><span class="fu">## Honesty {#sec-grf-honest}</span></span>
<span id="cb86-375"><a href="#cb86-375" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-376"><a href="#cb86-376" aria-hidden="true" tabindex="-1"></a>GRF applies <span class="kw">&lt;span</span> <span class="er">style</span><span class="ot">=</span><span class="st">"color:blue"</span><span class="kw">&gt;</span> honesty <span class="kw">&lt;/span&gt;</span>when it trains forests. Specifically, when building a tree, the subsample for the tree randomly selected without replacement is first split into two groups: subsample for <span class="kw">&lt;span</span> <span class="er">style</span><span class="ot">=</span><span class="st">"color:blue"</span><span class="kw">&gt;</span> splitting<span class="kw">&lt;/span&gt;</span> and <span class="kw">&lt;span</span> <span class="er">style</span><span class="ot">=</span><span class="st">"color:blue"</span><span class="kw">&gt;</span>prediction<span class="kw">&lt;/span&gt;</span>. Then, the a tree is trained on the subsample for splitting and then find the splitting rules using the standard CART algorithm. However, when predicting (say $E<span class="co">[</span><span class="ot">Y|X</span><span class="co">]</span>$ at $X=x$), the value of $Y$ from the subsample for splitting are NOT used. Rather, only the splitting rules are taken from the trained tree and then they are applied to the subsample for prediction (@fig-honest illustrates this process). </span>
<span id="cb86-377"><a href="#cb86-377" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-380"><a href="#cb86-380" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-381"><a href="#cb86-381" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb86-382"><a href="#cb86-382" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 2</span></span>
<span id="cb86-383"><a href="#cb86-383" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 4</span></span>
<span id="cb86-384"><a href="#cb86-384" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: Illustration of an honest tree</span></span>
<span id="cb86-385"><a href="#cb86-385" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-honest</span></span>
<span id="cb86-386"><a href="#cb86-386" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-387"><a href="#cb86-387" aria-hidden="true" tabindex="-1"></a>DiagrammeR<span class="sc">::</span><span class="fu">grViz</span>(</span>
<span id="cb86-388"><a href="#cb86-388" aria-hidden="true" tabindex="-1"></a><span class="st">"</span></span>
<span id="cb86-389"><a href="#cb86-389" aria-hidden="true" tabindex="-1"></a><span class="st">digraph {</span></span>
<span id="cb86-390"><a href="#cb86-390" aria-hidden="true" tabindex="-1"></a><span class="st">  graph [ranksep = 0.2, fontsize = 4]</span></span>
<span id="cb86-391"><a href="#cb86-391" aria-hidden="true" tabindex="-1"></a><span class="st">  node [shape = box]</span></span>
<span id="cb86-392"><a href="#cb86-392" aria-hidden="true" tabindex="-1"></a><span class="st">    SS [label = 'Subsample for splitting']</span></span>
<span id="cb86-393"><a href="#cb86-393" aria-hidden="true" tabindex="-1"></a><span class="st">    SP [label = 'Subsample for predicting']</span></span>
<span id="cb86-394"><a href="#cb86-394" aria-hidden="true" tabindex="-1"></a><span class="st">    BD [label = 'Sample for the tree']</span></span>
<span id="cb86-395"><a href="#cb86-395" aria-hidden="true" tabindex="-1"></a><span class="st">    TT [label = 'Trained tree']</span></span>
<span id="cb86-396"><a href="#cb86-396" aria-hidden="true" tabindex="-1"></a><span class="st">    PV [label = 'Predicted value']</span></span>
<span id="cb86-397"><a href="#cb86-397" aria-hidden="true" tabindex="-1"></a><span class="st">  edge [minlen = 2]</span></span>
<span id="cb86-398"><a href="#cb86-398" aria-hidden="true" tabindex="-1"></a><span class="st">    BD-&gt;SP</span></span>
<span id="cb86-399"><a href="#cb86-399" aria-hidden="true" tabindex="-1"></a><span class="st">    BD-&gt;SS</span></span>
<span id="cb86-400"><a href="#cb86-400" aria-hidden="true" tabindex="-1"></a><span class="st">    SS-&gt;TT</span></span>
<span id="cb86-401"><a href="#cb86-401" aria-hidden="true" tabindex="-1"></a><span class="st">    SP-&gt;PV</span></span>
<span id="cb86-402"><a href="#cb86-402" aria-hidden="true" tabindex="-1"></a><span class="st">    TT-&gt;SP [label='apply the splitting rules']</span></span>
<span id="cb86-403"><a href="#cb86-403" aria-hidden="true" tabindex="-1"></a><span class="st">  { rank = same; SS; SP}</span></span>
<span id="cb86-404"><a href="#cb86-404" aria-hidden="true" tabindex="-1"></a><span class="st">  { rank = same; TT}</span></span>
<span id="cb86-405"><a href="#cb86-405" aria-hidden="true" tabindex="-1"></a><span class="st">  { rank = same; PV}</span></span>
<span id="cb86-406"><a href="#cb86-406" aria-hidden="true" tabindex="-1"></a><span class="st">}</span></span>
<span id="cb86-407"><a href="#cb86-407" aria-hidden="true" tabindex="-1"></a><span class="st">"</span></span>
<span id="cb86-408"><a href="#cb86-408" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb86-409"><a href="#cb86-409" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-410"><a href="#cb86-410" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-411"><a href="#cb86-411" aria-hidden="true" tabindex="-1"></a>Let's demonstrate this using a very simple regression tree with two terminal nodes using the <span class="in">`mlb`</span> data from the <span class="in">`wooldridge`</span> package.</span>
<span id="cb86-412"><a href="#cb86-412" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-415"><a href="#cb86-415" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-416"><a href="#cb86-416" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(mlb1, <span class="at">package =</span> <span class="st">"wooldridge"</span>)</span>
<span id="cb86-417"><a href="#cb86-417" aria-hidden="true" tabindex="-1"></a>mlb1_dt <span class="ot">&lt;-</span> <span class="fu">data.table</span>(mlb1)</span>
<span id="cb86-418"><a href="#cb86-418" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-419"><a href="#cb86-419" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-420"><a href="#cb86-420" aria-hidden="true" tabindex="-1"></a>We would like to train a RF using this data where the dependent variable is logged salary (<span class="in">`lsalary`</span>). We will illustrate the honesty rule by working on building a single tree within the process of building a forest.</span>
<span id="cb86-421"><a href="#cb86-421" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-422"><a href="#cb86-422" aria-hidden="true" tabindex="-1"></a>We first sample the data without replacement. Here, we draw half of the entire sample randomly without replacement following the default value of <span class="in">`0.5`</span> for the <span class="in">`sample.fraction`</span> parameter.</span>
<span id="cb86-423"><a href="#cb86-423" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-426"><a href="#cb86-426" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-427"><a href="#cb86-427" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">358349</span>)</span>
<span id="cb86-428"><a href="#cb86-428" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-429"><a href="#cb86-429" aria-hidden="true" tabindex="-1"></a>num_obs <span class="ot">&lt;-</span> <span class="fu">nrow</span>(mlb1_dt)</span>
<span id="cb86-430"><a href="#cb86-430" aria-hidden="true" tabindex="-1"></a>row_indices <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="fu">seq_len</span>(num_obs), num_obs <span class="sc">/</span> <span class="dv">2</span>, <span class="at">replace =</span> <span class="cn">FALSE</span>)</span>
<span id="cb86-431"><a href="#cb86-431" aria-hidden="true" tabindex="-1"></a>boot_mlb1_dt <span class="ot">&lt;-</span> mlb1_dt[row_indices, ]</span>
<span id="cb86-432"><a href="#cb86-432" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-433"><a href="#cb86-433" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-434"><a href="#cb86-434" aria-hidden="true" tabindex="-1"></a><span class="in">`boot_mlb1_dt`</span> is the data that are involved in this tree. The rest of the samples in <span class="in">`mlb1_dt`</span> will have nothing to do with this tree whatsoever.</span>
<span id="cb86-435"><a href="#cb86-435" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-436"><a href="#cb86-436" aria-hidden="true" tabindex="-1"></a>We now split the sampled data into two groups: for splitting and prediction. </span>
<span id="cb86-437"><a href="#cb86-437" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-440"><a href="#cb86-440" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-441"><a href="#cb86-441" aria-hidden="true" tabindex="-1"></a>num_obs_sub <span class="ot">&lt;-</span> <span class="fu">nrow</span>(boot_mlb1_dt) </span>
<span id="cb86-442"><a href="#cb86-442" aria-hidden="true" tabindex="-1"></a>rows_split <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="fu">seq_len</span>(num_obs_sub), num_obs_sub <span class="sc">/</span> <span class="dv">2</span>, <span class="at">replace =</span> <span class="cn">FALSE</span>)</span>
<span id="cb86-443"><a href="#cb86-443" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-444"><a href="#cb86-444" aria-hidden="true" tabindex="-1"></a><span class="co">#=== data for splitting ===#</span></span>
<span id="cb86-445"><a href="#cb86-445" aria-hidden="true" tabindex="-1"></a>split_data <span class="ot">&lt;-</span> boot_mlb1_dt[rows_split, ]</span>
<span id="cb86-446"><a href="#cb86-446" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-447"><a href="#cb86-447" aria-hidden="true" tabindex="-1"></a><span class="co">#=== data for prediction ===#</span></span>
<span id="cb86-448"><a href="#cb86-448" aria-hidden="true" tabindex="-1"></a>eval_data <span class="ot">&lt;-</span> boot_mlb1_dt[<span class="sc">-</span>rows_split, ]</span>
<span id="cb86-449"><a href="#cb86-449" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-450"><a href="#cb86-450" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-451"><a href="#cb86-451" aria-hidden="true" tabindex="-1"></a>We then train a tree using the data for splitting (<span class="in">`split_data`</span>):</span>
<span id="cb86-452"><a href="#cb86-452" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-455"><a href="#cb86-455" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-456"><a href="#cb86-456" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: A simple regression tree using the subsamples for splitting</span></span>
<span id="cb86-457"><a href="#cb86-457" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-tree-sub</span></span>
<span id="cb86-458"><a href="#cb86-458" aria-hidden="true" tabindex="-1"></a><span class="co">#=== build a simple tree ===#</span></span>
<span id="cb86-459"><a href="#cb86-459" aria-hidden="true" tabindex="-1"></a>tree_trained <span class="ot">&lt;-</span></span>
<span id="cb86-460"><a href="#cb86-460" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rpart</span>(</span>
<span id="cb86-461"><a href="#cb86-461" aria-hidden="true" tabindex="-1"></a>    lsalary <span class="sc">~</span> hits <span class="sc">+</span> runsyr, </span>
<span id="cb86-462"><a href="#cb86-462" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> split_data, </span>
<span id="cb86-463"><a href="#cb86-463" aria-hidden="true" tabindex="-1"></a>    <span class="at">control =</span> <span class="fu">rpart.control</span>(<span class="at">minsplit =</span> <span class="dv">60</span>)</span>
<span id="cb86-464"><a href="#cb86-464" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb86-465"><a href="#cb86-465" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-466"><a href="#cb86-466" aria-hidden="true" tabindex="-1"></a><span class="fu">fancyRpartPlot</span>(tree_trained, <span class="at">digits =</span> <span class="dv">4</span>)</span>
<span id="cb86-467"><a href="#cb86-467" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-468"><a href="#cb86-468" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-469"><a href="#cb86-469" aria-hidden="true" tabindex="-1"></a>So the splitting rule is <span class="in">`hits &lt; 338`</span> as shown in @fig-tree-sub. At the terminal nodes, you see the prediction of <span class="in">`lsalary`</span>: $12.41$ for the left and $14.36$ for the right. These predictions are NOT honest. They are obtained from the observed values of <span class="in">`lsalary`</span> within the node using the splitting data (the data the tree is trained for). Instead of using these prediction values, an honest prediction applied the splitting rules (<span class="in">`hits &lt; 338`</span>) to the data reserved for prediction.</span>
<span id="cb86-470"><a href="#cb86-470" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-473"><a href="#cb86-473" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-474"><a href="#cb86-474" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb86-475"><a href="#cb86-475" aria-hidden="true" tabindex="-1"></a>honest_pred <span class="ot">&lt;-</span> eval_data[, <span class="fu">mean</span>(lsalary), <span class="at">by =</span> hits <span class="sc">&lt;</span> <span class="dv">338</span>]</span>
<span id="cb86-476"><a href="#cb86-476" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb86-477"><a href="#cb86-477" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-478"><a href="#cb86-478" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-479"><a href="#cb86-479" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-480"><a href="#cb86-480" aria-hidden="true" tabindex="-1"></a>So, instead of $12.41$ and $14.36$, the predicted values from the honest tree are $<span class="in">`r round(honest_pred[hits == TRUE, V1], digits = 2)`</span>$ and $<span class="in">`r round(honest_pred[hits == FALSE, V1], digits = 2)`</span>$ for the left and right nodes, respectively. Trees are built in this manner many times to form a forest.</span>
<span id="cb86-481"><a href="#cb86-481" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-482"><a href="#cb86-482" aria-hidden="true" tabindex="-1"></a>More generally, in GRF, honesty is applied by using the evaluation data to solve @eq-opt based on the weight $\alpha_i(x)$ derived from the trained forest using the splitting data. Honesty is required for the GRF estimator to be consistent and asymptotically normal <span class="co">[</span><span class="ot">@athey2019generalized</span><span class="co">]</span>. However, the application of honesty can do more damage than help when the sample size is small.</span>
<span id="cb86-483"><a href="#cb86-483" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-484"><a href="#cb86-484" aria-hidden="true" tabindex="-1"></a><span class="fu">## Understanding GRF better by example {#sec-understand-grf-example}</span></span>
<span id="cb86-485"><a href="#cb86-485" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-486"><a href="#cb86-486" aria-hidden="true" tabindex="-1"></a>This section goes through CF model training and CATE prediction step by step with codes to enhance our understanding of how GRF (and CF) works. In the process, we also learn the role of many of the hyper-parameters that are common across models that fall under GRF. We also learn the role of hyper-parameter that are specific to CF as well. </span>
<span id="cb86-487"><a href="#cb86-487" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-488"><a href="#cb86-488" aria-hidden="true" tabindex="-1"></a>While the process explained here is for CF, the vast majority of the process is exactly the same for other models under GRF. Here are some differences:</span>
<span id="cb86-489"><a href="#cb86-489" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-490"><a href="#cb86-490" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>Orthogonalization of the data at the beginning for CF </span>
<span id="cb86-491"><a href="#cb86-491" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>The definition of the pseudo outcome (but, the way they are used in the algorithm is identical once they are defined) </span>
<span id="cb86-492"><a href="#cb86-492" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>The role of <span class="in">`min.node.size`</span>, <span class="in">`alpha`</span>, and <span class="in">`imbalance.penalty`</span> used for safeguarding against extreme splits</span>
<span id="cb86-493"><a href="#cb86-493" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-494"><a href="#cb86-494" aria-hidden="true" tabindex="-1"></a>::: {.column-margin}</span>
<span id="cb86-495"><a href="#cb86-495" aria-hidden="true" tabindex="-1"></a>Obviously, the codes here are only for illustration and should not be used in practice.</span>
<span id="cb86-496"><a href="#cb86-496" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb86-497"><a href="#cb86-497" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-498"><a href="#cb86-498" aria-hidden="true" tabindex="-1"></a>We will use a synthetic data generated from the following DGP:</span>
<span id="cb86-499"><a href="#cb86-499" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-500"><a href="#cb86-500" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-501"><a href="#cb86-501" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb86-502"><a href="#cb86-502" aria-hidden="true" tabindex="-1"></a>y = \theta(X)\cdot T + g(X) + \mu</span>
<span id="cb86-503"><a href="#cb86-503" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb86-504"><a href="#cb86-504" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-505"><a href="#cb86-505" aria-hidden="true" tabindex="-1"></a>, where</span>
<span id="cb86-506"><a href="#cb86-506" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-507"><a href="#cb86-507" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-508"><a href="#cb86-508" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb86-509"><a href="#cb86-509" aria-hidden="true" tabindex="-1"></a>\theta(X) &amp; = 2 + 3 \times x_1^2 - \sqrt{3 \times  x_2^3 + 1}<span class="sc">\\</span></span>
<span id="cb86-510"><a href="#cb86-510" aria-hidden="true" tabindex="-1"></a>g(X) &amp; = 10 \cdot <span class="co">[</span><span class="ot">log(x_2+1) + 2\cdot x_3\times x_2</span><span class="co">]</span></span>
<span id="cb86-511"><a href="#cb86-511" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb86-512"><a href="#cb86-512" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-513"><a href="#cb86-513" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-514"><a href="#cb86-514" aria-hidden="true" tabindex="-1"></a>We have 10 feature variables $x_1, \dots, x_{10}$ and only $x_1$ through $x_3$ play a role.  </span>
<span id="cb86-515"><a href="#cb86-515" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-516"><a href="#cb86-516" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>$x_1, x_2 \sim U<span class="co">[</span><span class="ot">0, 3</span><span class="co">]</span>^2$</span>
<span id="cb86-517"><a href="#cb86-517" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>$x_3 \sim N(1, 1)$</span>
<span id="cb86-518"><a href="#cb86-518" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>$x_4, \dots, x_{10} \sim U<span class="co">[</span><span class="ot">0, 1</span><span class="co">]</span>^{7}$</span>
<span id="cb86-519"><a href="#cb86-519" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-520"><a href="#cb86-520" aria-hidden="true" tabindex="-1"></a>The error term and the treatment variable is defined as follows. </span>
<span id="cb86-521"><a href="#cb86-521" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-522"><a href="#cb86-522" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>$\mu \sim N(0, 1)$</span>
<span id="cb86-523"><a href="#cb86-523" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>$T \sim Ber(0.5)$</span>
<span id="cb86-524"><a href="#cb86-524" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-527"><a href="#cb86-527" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-528"><a href="#cb86-528" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">73843</span>)</span>
<span id="cb86-529"><a href="#cb86-529" aria-hidden="true" tabindex="-1"></a>N <span class="ot">&lt;-</span> <span class="dv">1000</span> <span class="co"># number of observations</span></span>
<span id="cb86-530"><a href="#cb86-530" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-531"><a href="#cb86-531" aria-hidden="true" tabindex="-1"></a>x4_x10 <span class="ot">&lt;-</span> </span>
<span id="cb86-532"><a href="#cb86-532" aria-hidden="true" tabindex="-1"></a>  <span class="fu">matrix</span>(<span class="fu">runif</span>(N <span class="sc">*</span> <span class="dv">7</span>), <span class="at">nrow =</span> N) <span class="sc">%&gt;%</span> </span>
<span id="cb86-533"><a href="#cb86-533" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.table</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb86-534"><a href="#cb86-534" aria-hidden="true" tabindex="-1"></a>  <span class="fu">setnames</span>(<span class="fu">names</span>(.), <span class="fu">paste0</span>(<span class="st">"x"</span>, <span class="dv">4</span><span class="sc">:</span><span class="dv">10</span>))</span>
<span id="cb86-535"><a href="#cb86-535" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-536"><a href="#cb86-536" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb86-537"><a href="#cb86-537" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> </span>
<span id="cb86-538"><a href="#cb86-538" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.table</span>(</span>
<span id="cb86-539"><a href="#cb86-539" aria-hidden="true" tabindex="-1"></a>    <span class="at">x1 =</span> <span class="dv">3</span> <span class="sc">*</span> <span class="fu">runif</span>(N),</span>
<span id="cb86-540"><a href="#cb86-540" aria-hidden="true" tabindex="-1"></a>    <span class="at">x2 =</span> <span class="dv">3</span> <span class="sc">*</span> <span class="fu">runif</span>(N),</span>
<span id="cb86-541"><a href="#cb86-541" aria-hidden="true" tabindex="-1"></a>    <span class="at">x3 =</span> <span class="fu">rnorm</span>(N, <span class="at">mean =</span> <span class="dv">1</span>, <span class="at">sd =</span> <span class="dv">1</span>),</span>
<span id="cb86-542"><a href="#cb86-542" aria-hidden="true" tabindex="-1"></a>    <span class="at">mu =</span> <span class="fu">rnorm</span>(N),</span>
<span id="cb86-543"><a href="#cb86-543" aria-hidden="true" tabindex="-1"></a>    <span class="at">T =</span> <span class="fu">runif</span>(N) <span class="sc">&gt;</span> <span class="fl">0.5</span></span>
<span id="cb86-544"><a href="#cb86-544" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span> </span>
<span id="cb86-545"><a href="#cb86-545" aria-hidden="true" tabindex="-1"></a>  <span class="co">#=== theta(X) ===#</span></span>
<span id="cb86-546"><a href="#cb86-546" aria-hidden="true" tabindex="-1"></a>  .[, <span class="at">theta_x :=</span> <span class="dv">2</span> <span class="sc">+</span> <span class="dv">3</span> <span class="sc">*</span> x1<span class="sc">^</span><span class="dv">2</span> <span class="sc">-</span> <span class="fu">sqrt</span>(<span class="dv">3</span><span class="sc">*</span> x2<span class="sc">^</span><span class="dv">3</span> <span class="sc">+</span> <span class="dv">2</span>)] <span class="sc">%&gt;%</span></span>
<span id="cb86-547"><a href="#cb86-547" aria-hidden="true" tabindex="-1"></a>  <span class="co">#=== g(X) ===#</span></span>
<span id="cb86-548"><a href="#cb86-548" aria-hidden="true" tabindex="-1"></a>  .[, <span class="at">g_x :=</span> <span class="dv">10</span> <span class="sc">*</span> (<span class="fu">log</span>(x2 <span class="sc">+</span> <span class="dv">1</span>) <span class="sc">+</span> <span class="dv">2</span> <span class="sc">*</span> x3 <span class="sc">*</span> x2)] <span class="sc">%&gt;%</span> </span>
<span id="cb86-549"><a href="#cb86-549" aria-hidden="true" tabindex="-1"></a>  .[, <span class="at">y :=</span> theta_x <span class="sc">*</span> T <span class="sc">+</span> g_x <span class="sc">+</span> mu] <span class="sc">%&gt;%</span> </span>
<span id="cb86-550"><a href="#cb86-550" aria-hidden="true" tabindex="-1"></a>  .[, .(y, T, x1, x2, x3)] <span class="sc">%&gt;%</span> </span>
<span id="cb86-551"><a href="#cb86-551" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cbind</span>(., x4_x10)</span>
<span id="cb86-552"><a href="#cb86-552" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb86-553"><a href="#cb86-553" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-554"><a href="#cb86-554" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-555"><a href="#cb86-555" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-556"><a href="#cb86-556" aria-hidden="true" tabindex="-1"></a><span class="fu">### Orthogonalization of $Y$ and $T$</span></span>
<span id="cb86-557"><a href="#cb86-557" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-558"><a href="#cb86-558" aria-hidden="true" tabindex="-1"></a>Causal forest (and instrumental forest) first orthogonalizes $Y$ and $T$. This is not the case for the other GRF methods. By default, <span class="in">`causal_forest()`</span> uses random forest to estimate $E<span class="co">[</span><span class="ot">Y|X</span><span class="co">]</span>$ and $E<span class="co">[</span><span class="ot">T|X</span><span class="co">]</span>$, and then use out-of-bag predictions. Since the treatment assignment is randomized, we could use $0.5$ for as the estimate for $E<span class="co">[</span><span class="ot">T|X</span><span class="co">]</span>$ for all the observations. But, we will estimate both in this example.</span>
<span id="cb86-559"><a href="#cb86-559" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-560"><a href="#cb86-560" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-563"><a href="#cb86-563" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-564"><a href="#cb86-564" aria-hidden="true" tabindex="-1"></a><span class="co">#--------------------------</span></span>
<span id="cb86-565"><a href="#cb86-565" aria-hidden="true" tabindex="-1"></a><span class="co"># E[Y|X]</span></span>
<span id="cb86-566"><a href="#cb86-566" aria-hidden="true" tabindex="-1"></a><span class="co">#--------------------------</span></span>
<span id="cb86-567"><a href="#cb86-567" aria-hidden="true" tabindex="-1"></a><span class="co">#=== train an RF on y ===#</span></span>
<span id="cb86-568"><a href="#cb86-568" aria-hidden="true" tabindex="-1"></a>y_rf_trained <span class="ot">&lt;-</span></span>
<span id="cb86-569"><a href="#cb86-569" aria-hidden="true" tabindex="-1"></a>  <span class="fu">regression_forest</span>(</span>
<span id="cb86-570"><a href="#cb86-570" aria-hidden="true" tabindex="-1"></a>    <span class="at">X =</span> data[, .(x1, x2, x3)],</span>
<span id="cb86-571"><a href="#cb86-571" aria-hidden="true" tabindex="-1"></a>    <span class="at">Y =</span> data[, y]</span>
<span id="cb86-572"><a href="#cb86-572" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb86-573"><a href="#cb86-573" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-574"><a href="#cb86-574" aria-hidden="true" tabindex="-1"></a><span class="co">#=== OOB estimates of E[Y|X] ===#</span></span>
<span id="cb86-575"><a href="#cb86-575" aria-hidden="true" tabindex="-1"></a>ey_x_hat <span class="ot">&lt;-</span> y_rf_trained<span class="sc">$</span>predictions</span>
<span id="cb86-576"><a href="#cb86-576" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-577"><a href="#cb86-577" aria-hidden="true" tabindex="-1"></a><span class="co">#--------------------------</span></span>
<span id="cb86-578"><a href="#cb86-578" aria-hidden="true" tabindex="-1"></a><span class="co"># E[T|X]</span></span>
<span id="cb86-579"><a href="#cb86-579" aria-hidden="true" tabindex="-1"></a><span class="co">#--------------------------</span></span>
<span id="cb86-580"><a href="#cb86-580" aria-hidden="true" tabindex="-1"></a><span class="co">#=== train an RF on T ===#</span></span>
<span id="cb86-581"><a href="#cb86-581" aria-hidden="true" tabindex="-1"></a>t_rf_trained <span class="ot">&lt;-</span></span>
<span id="cb86-582"><a href="#cb86-582" aria-hidden="true" tabindex="-1"></a>  <span class="fu">probability_forest</span>(</span>
<span id="cb86-583"><a href="#cb86-583" aria-hidden="true" tabindex="-1"></a>    <span class="at">X =</span> data[, .(x1, x2, x3)],</span>
<span id="cb86-584"><a href="#cb86-584" aria-hidden="true" tabindex="-1"></a>    <span class="at">Y =</span> data[, <span class="fu">factor</span>(T)]</span>
<span id="cb86-585"><a href="#cb86-585" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb86-586"><a href="#cb86-586" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-587"><a href="#cb86-587" aria-hidden="true" tabindex="-1"></a><span class="co">#=== OOB estimates of E[T|X] ===#</span></span>
<span id="cb86-588"><a href="#cb86-588" aria-hidden="true" tabindex="-1"></a>et_x_hat <span class="ot">&lt;-</span> t_rf_trained<span class="sc">$</span>predictions[, <span class="dv">2</span>] <span class="co"># get the probability of T = 1</span></span>
<span id="cb86-589"><a href="#cb86-589" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-590"><a href="#cb86-590" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-591"><a href="#cb86-591" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-592"><a href="#cb86-592" aria-hidden="true" tabindex="-1"></a>Let's now orthogonalize $Y$ and $T$,</span>
<span id="cb86-593"><a href="#cb86-593" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-596"><a href="#cb86-596" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-597"><a href="#cb86-597" aria-hidden="true" tabindex="-1"></a>data[, <span class="st">`</span><span class="at">:=</span><span class="st">`</span>(<span class="at">y_tilde =</span> y <span class="sc">-</span> ey_x_hat, <span class="at">t_tilde =</span> T <span class="sc">-</span> et_x_hat)]</span>
<span id="cb86-598"><a href="#cb86-598" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-599"><a href="#cb86-599" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-600"><a href="#cb86-600" aria-hidden="true" tabindex="-1"></a>Here is what the data looks like:</span>
<span id="cb86-601"><a href="#cb86-601" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-604"><a href="#cb86-604" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-605"><a href="#cb86-605" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(data)</span>
<span id="cb86-606"><a href="#cb86-606" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-607"><a href="#cb86-607" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-608"><a href="#cb86-608" aria-hidden="true" tabindex="-1"></a><span class="fu">### Building trees #{#sec-build-trees}</span></span>
<span id="cb86-609"><a href="#cb86-609" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-610"><a href="#cb86-610" aria-hidden="true" tabindex="-1"></a>---</span>
<span id="cb86-611"><a href="#cb86-611" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-612"><a href="#cb86-612" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;span</span> <span class="er">style</span><span class="ot">=</span><span class="st">"color:blue"</span><span class="kw">&gt;</span> Preparing Data<span class="kw">&lt;/span&gt;</span>:</span>
<span id="cb86-613"><a href="#cb86-613" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-614"><a href="#cb86-614" aria-hidden="true" tabindex="-1"></a>When building a tree, GRF use sampling without replacement instead of sampling with replacement as a default for RF implemented by <span class="in">`ranger()`</span>. <span class="in">`sample.fraction`</span> parameter determines the fraction of the entire sample drawn for each tree. Let's use the default value, which is $0.5$. </span>
<span id="cb86-615"><a href="#cb86-615" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-618"><a href="#cb86-618" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-619"><a href="#cb86-619" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb86-620"><a href="#cb86-620" aria-hidden="true" tabindex="-1"></a>data_for_the_first_tree <span class="ot">&lt;-</span> data[<span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span>N, <span class="fl">0.5</span> <span class="sc">*</span> N, <span class="at">replace =</span> <span class="cn">FALSE</span>), ] </span>
<span id="cb86-621"><a href="#cb86-621" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb86-622"><a href="#cb86-622" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-623"><a href="#cb86-623" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-624"><a href="#cb86-624" aria-hidden="true" tabindex="-1"></a>By default, honest splitting is implemented (<span class="in">`honesty = TRUE`</span>). The <span class="in">`honesty.fraction`</span> parameter determines the fraction of the randomly sampled data (<span class="in">`data_for_the_first_tree`</span>), which will be used for determining splitting decisions. We will use the default value of <span class="in">`honesty.fraction`</span>, which is $0.5$.</span>
<span id="cb86-625"><a href="#cb86-625" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-628"><a href="#cb86-628" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-629"><a href="#cb86-629" aria-hidden="true" tabindex="-1"></a><span class="co">#=== number of observations ===#</span></span>
<span id="cb86-630"><a href="#cb86-630" aria-hidden="true" tabindex="-1"></a>N_d <span class="ot">&lt;-</span> <span class="fu">nrow</span>(data_for_the_first_tree)</span>
<span id="cb86-631"><a href="#cb86-631" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-632"><a href="#cb86-632" aria-hidden="true" tabindex="-1"></a><span class="co">#=== indices ===#</span></span>
<span id="cb86-633"><a href="#cb86-633" aria-hidden="true" tabindex="-1"></a>J1_index <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span>N_d, <span class="fl">0.5</span> <span class="sc">*</span> N_d, <span class="at">replace =</span> <span class="cn">FALSE</span>)</span>
<span id="cb86-634"><a href="#cb86-634" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-635"><a href="#cb86-635" aria-hidden="true" tabindex="-1"></a><span class="co">#=== data for determining splitting rules ===#</span></span>
<span id="cb86-636"><a href="#cb86-636" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb86-637"><a href="#cb86-637" aria-hidden="true" tabindex="-1"></a>data_J1 <span class="ot">&lt;-</span> data_for_the_first_tree[J1_index, ]</span>
<span id="cb86-638"><a href="#cb86-638" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb86-639"><a href="#cb86-639" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-640"><a href="#cb86-640" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-641"><a href="#cb86-641" aria-hidden="true" tabindex="-1"></a>As you can see, we are just using a quarter of the original dataset (<span class="in">`sample.fraction`</span> $\times$ <span class="in">`honesty.fraction`</span> = $0.5 \times 0.5$). Here is the data for repopulating the tree once the tree is built.</span>
<span id="cb86-642"><a href="#cb86-642" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-645"><a href="#cb86-645" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-646"><a href="#cb86-646" aria-hidden="true" tabindex="-1"></a><span class="co">#=== data for repopulate the tree ===#</span></span>
<span id="cb86-647"><a href="#cb86-647" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb86-648"><a href="#cb86-648" aria-hidden="true" tabindex="-1"></a>data_J2 <span class="ot">&lt;-</span> data_for_the_first_tree[<span class="sc">-</span>J1_index, ]</span>
<span id="cb86-649"><a href="#cb86-649" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb86-650"><a href="#cb86-650" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-651"><a href="#cb86-651" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-652"><a href="#cb86-652" aria-hidden="true" tabindex="-1"></a>---</span>
<span id="cb86-653"><a href="#cb86-653" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-654"><a href="#cb86-654" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;span</span> <span class="er">style</span><span class="ot">=</span><span class="st">"color:blue"</span><span class="kw">&gt;</span> Determining splitting rules<span class="kw">&lt;/span&gt;</span>:</span>
<span id="cb86-655"><a href="#cb86-655" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-656"><a href="#cb86-656" aria-hidden="true" tabindex="-1"></a>Now let's find the splitting rules using <span class="in">`data_J1`</span> (Remember, <span class="in">`data_J2`</span> is not used).</span>
<span id="cb86-657"><a href="#cb86-657" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-658"><a href="#cb86-658" aria-hidden="true" tabindex="-1"></a>We are at the root node to which all the observations belong. Let's first calculate the pseudo outcome. Finding pseudo outcomes starts from solving the unweighted version of @eq-opt using the samples in the parent node (root node here) in general. For CF, it is</span>
<span id="cb86-659"><a href="#cb86-659" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-660"><a href="#cb86-660" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-661"><a href="#cb86-661" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb86-662"><a href="#cb86-662" aria-hidden="true" tabindex="-1"></a>\sum_{i=1}^n (\tilde{Y}_i - \theta\tilde{T}_i)\cdot \tilde{T}_i = 0</span>
<span id="cb86-663"><a href="#cb86-663" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb86-664"><a href="#cb86-664" aria-hidden="true" tabindex="-1"></a>$$ {#eq-theta-p}</span>
<span id="cb86-665"><a href="#cb86-665" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-666"><a href="#cb86-666" aria-hidden="true" tabindex="-1"></a>because the score function for CF is $(\tilde{Y}_i - \theta\tilde{T}_i)\cdot \tilde{T}_i$. So, </span>
<span id="cb86-667"><a href="#cb86-667" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-668"><a href="#cb86-668" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-669"><a href="#cb86-669" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb86-670"><a href="#cb86-670" aria-hidden="true" tabindex="-1"></a>\theta_P = \frac{\sum_{i=1}^n \tilde{Y}_i\cdot \tilde{T}_i}{\sum_{i=1}^n\tilde{T}_i\cdot \tilde{T}_i}</span>
<span id="cb86-671"><a href="#cb86-671" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb86-672"><a href="#cb86-672" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-673"><a href="#cb86-673" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-676"><a href="#cb86-676" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-677"><a href="#cb86-677" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb86-678"><a href="#cb86-678" aria-hidden="true" tabindex="-1"></a>theta_p <span class="ot">&lt;-</span> data_J1[, <span class="fu">sum</span>(y_tilde <span class="sc">*</span> t_tilde)<span class="sc">/</span><span class="fu">sum</span>(t_tilde <span class="sc">*</span> t_tilde)]</span>
<span id="cb86-679"><a href="#cb86-679" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb86-680"><a href="#cb86-680" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-681"><a href="#cb86-681" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-682"><a href="#cb86-682" aria-hidden="true" tabindex="-1"></a>The pseudo outcome for CF is $(\tilde{Y}_i - \theta_P\tilde{T}_i)\cdot \tilde{T}_i$. Let's define that in <span class="in">`data_J1`</span>.</span>
<span id="cb86-683"><a href="#cb86-683" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-686"><a href="#cb86-686" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-687"><a href="#cb86-687" aria-hidden="true" tabindex="-1"></a>data_J1[, rho <span class="sc">:</span><span class="er">=</span> (y_tilde <span class="sc">-</span> theta_p <span class="sc">*</span> t_tilde) <span class="sc">*</span> t_tilde]</span>
<span id="cb86-688"><a href="#cb86-688" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-689"><a href="#cb86-689" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-690"><a href="#cb86-690" aria-hidden="true" tabindex="-1"></a>Now, the number of variables to use for splitting is determined by the <span class="in">`mtry`</span> parameter. Its default is <span class="in">`min(ceiling(sqrt(ncol(X))+20), ncol(X))`</span>, where <span class="in">`X`</span> is the feature matrix. Here <span class="in">`ncol(X)`</span> is 10. So, <span class="in">`mtry`</span> is set to <span class="in">`r min(ceiling(sqrt(10)+20), 10)`</span>. Following this, we will use all three variables for splitting here. Let $C_j$ ($j = 1, 2$) and $N_j$ denote the child node $j$ and the number of observations in child node $j$, respectively. We will write a function that works on a single feature variable to find all the threshold values that would result in unique splits, and then return the following information:</span>
<span id="cb86-691"><a href="#cb86-691" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-692"><a href="#cb86-692" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>$(\sum_{i\in C_j} \rho_i)^2$</span>
<span id="cb86-693"><a href="#cb86-693" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>$N_j$</span>
<span id="cb86-694"><a href="#cb86-694" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>$\sum_{i=1}^{N_j} (T_i - \bar{T})^2$ (we will call this <span class="in">`info_size`</span>)</span>
<span id="cb86-695"><a href="#cb86-695" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>the number of treated and control units</span>
<span id="cb86-696"><a href="#cb86-696" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-697"><a href="#cb86-697" aria-hidden="true" tabindex="-1"></a>The output allows us to calculate the heterogeneity score of the split defined as</span>
<span id="cb86-698"><a href="#cb86-698" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-699"><a href="#cb86-699" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-700"><a href="#cb86-700" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb86-701"><a href="#cb86-701" aria-hidden="true" tabindex="-1"></a>\sum_{j=1}^2 \frac{(\sum_{i\in C_j} \rho_i)^2}{N_j}.</span>
<span id="cb86-702"><a href="#cb86-702" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb86-703"><a href="#cb86-703" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-704"><a href="#cb86-704" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-705"><a href="#cb86-705" aria-hidden="true" tabindex="-1"></a>It also allows us to eliminate some of the splits based on <span class="in">`mtry`</span>, <span class="in">`alpha`</span>, and <span class="in">`imbalance.penalty`</span> parameters. </span>
<span id="cb86-706"><a href="#cb86-706" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-709"><a href="#cb86-709" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-710"><a href="#cb86-710" aria-hidden="true" tabindex="-1"></a>get_ss_by_var <span class="ot">&lt;-</span> <span class="cf">function</span>(feature_var, outcome_var, parent_data)</span>
<span id="cb86-711"><a href="#cb86-711" aria-hidden="true" tabindex="-1"></a>{</span>
<span id="cb86-712"><a href="#cb86-712" aria-hidden="true" tabindex="-1"></a>  temp_data <span class="ot">&lt;-</span> </span>
<span id="cb86-713"><a href="#cb86-713" aria-hidden="true" tabindex="-1"></a>    <span class="fu">copy</span>(parent_data) <span class="sc">%&gt;%</span> </span>
<span id="cb86-714"><a href="#cb86-714" aria-hidden="true" tabindex="-1"></a>    <span class="fu">setnames</span>(feature_var, <span class="st">"temp_var"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb86-715"><a href="#cb86-715" aria-hidden="true" tabindex="-1"></a>    <span class="fu">setnames</span>(outcome_var, <span class="st">"outcome"</span>)  </span>
<span id="cb86-716"><a href="#cb86-716" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-717"><a href="#cb86-717" aria-hidden="true" tabindex="-1"></a>  <span class="co">#=== define a sequence of values of hruns ===#</span></span>
<span id="cb86-718"><a href="#cb86-718" aria-hidden="true" tabindex="-1"></a>  thr_seq <span class="ot">&lt;-</span></span>
<span id="cb86-719"><a href="#cb86-719" aria-hidden="true" tabindex="-1"></a>    temp_data[<span class="fu">order</span>(temp_var), <span class="fu">unique</span>(temp_var)] <span class="sc">%&gt;%</span></span>
<span id="cb86-720"><a href="#cb86-720" aria-hidden="true" tabindex="-1"></a>    <span class="co">#=== get the rolling mean ===#</span></span>
<span id="cb86-721"><a href="#cb86-721" aria-hidden="true" tabindex="-1"></a>    <span class="fu">frollmean</span>(<span class="dv">2</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb86-722"><a href="#cb86-722" aria-hidden="true" tabindex="-1"></a>    .[<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb86-723"><a href="#cb86-723" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-724"><a href="#cb86-724" aria-hidden="true" tabindex="-1"></a>  <span class="co">#=== get RSS ===#</span></span>
<span id="cb86-725"><a href="#cb86-725" aria-hidden="true" tabindex="-1"></a>  ss_value <span class="ot">&lt;-</span></span>
<span id="cb86-726"><a href="#cb86-726" aria-hidden="true" tabindex="-1"></a>    <span class="fu">lapply</span>(</span>
<span id="cb86-727"><a href="#cb86-727" aria-hidden="true" tabindex="-1"></a>      thr_seq,</span>
<span id="cb86-728"><a href="#cb86-728" aria-hidden="true" tabindex="-1"></a>      <span class="cf">function</span>(x){</span>
<span id="cb86-729"><a href="#cb86-729" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-730"><a href="#cb86-730" aria-hidden="true" tabindex="-1"></a>        return_data <span class="ot">&lt;-</span> </span>
<span id="cb86-731"><a href="#cb86-731" aria-hidden="true" tabindex="-1"></a>          temp_data[, .(</span>
<span id="cb86-732"><a href="#cb86-732" aria-hidden="true" tabindex="-1"></a>            <span class="at">het_score =</span> <span class="fu">sum</span>(outcome)<span class="sc">^</span><span class="dv">2</span><span class="sc">/</span>.N, </span>
<span id="cb86-733"><a href="#cb86-733" aria-hidden="true" tabindex="-1"></a>            <span class="at">nobs =</span> .N,</span>
<span id="cb86-734"><a href="#cb86-734" aria-hidden="true" tabindex="-1"></a>            <span class="at">info_size =</span> <span class="fu">sum</span>((T <span class="sc">-</span> <span class="fu">mean</span>(T))<span class="sc">^</span><span class="dv">2</span>),</span>
<span id="cb86-735"><a href="#cb86-735" aria-hidden="true" tabindex="-1"></a>            <span class="at">num_treated =</span> <span class="fu">sum</span>(T),</span>
<span id="cb86-736"><a href="#cb86-736" aria-hidden="true" tabindex="-1"></a>            <span class="at">num_ctrl =</span> <span class="fu">sum</span>(<span class="sc">!</span>T)</span>
<span id="cb86-737"><a href="#cb86-737" aria-hidden="true" tabindex="-1"></a>          ), by <span class="ot">=</span> temp_var <span class="sc">&lt;</span> x] <span class="sc">%&gt;%</span> </span>
<span id="cb86-738"><a href="#cb86-738" aria-hidden="true" tabindex="-1"></a>          <span class="fu">setnames</span>(<span class="st">"temp_var"</span>, <span class="st">"child"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb86-739"><a href="#cb86-739" aria-hidden="true" tabindex="-1"></a>          .[, child <span class="sc">:</span><span class="er">=</span> <span class="fu">ifelse</span>(child <span class="sc">==</span> <span class="cn">TRUE</span>, <span class="st">"Child 1"</span>, <span class="st">"Child 2"</span>)] <span class="sc">%&gt;%</span> </span>
<span id="cb86-740"><a href="#cb86-740" aria-hidden="true" tabindex="-1"></a>          .[, thr <span class="sc">:</span><span class="er">=</span> x]</span>
<span id="cb86-741"><a href="#cb86-741" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb86-742"><a href="#cb86-742" aria-hidden="true" tabindex="-1"></a>        <span class="fu">return</span>(return_data)</span>
<span id="cb86-743"><a href="#cb86-743" aria-hidden="true" tabindex="-1"></a>      } </span>
<span id="cb86-744"><a href="#cb86-744" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">%&gt;%</span> </span>
<span id="cb86-745"><a href="#cb86-745" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rbindlist</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb86-746"><a href="#cb86-746" aria-hidden="true" tabindex="-1"></a>    .[, var <span class="sc">:</span><span class="er">=</span> feature_var] <span class="sc">%&gt;%</span> </span>
<span id="cb86-747"><a href="#cb86-747" aria-hidden="true" tabindex="-1"></a>    <span class="fu">relocate</span>(var, thr, child)</span>
<span id="cb86-748"><a href="#cb86-748" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-749"><a href="#cb86-749" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(ss_value)</span>
<span id="cb86-750"><a href="#cb86-750" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb86-751"><a href="#cb86-751" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-752"><a href="#cb86-752" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-753"><a href="#cb86-753" aria-hidden="true" tabindex="-1"></a>For example, here is the output for <span class="in">`x1`</span>.</span>
<span id="cb86-754"><a href="#cb86-754" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-757"><a href="#cb86-757" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-758"><a href="#cb86-758" aria-hidden="true" tabindex="-1"></a><span class="fu">get_ss_by_var</span>(<span class="st">"x1"</span>, <span class="st">"rho"</span>, data_J1)[]</span>
<span id="cb86-759"><a href="#cb86-759" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-760"><a href="#cb86-760" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-761"><a href="#cb86-761" aria-hidden="true" tabindex="-1"></a>Repeating this for all the feature variables,</span>
<span id="cb86-762"><a href="#cb86-762" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-765"><a href="#cb86-765" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-766"><a href="#cb86-766" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb86-767"><a href="#cb86-767" aria-hidden="true" tabindex="-1"></a>thr_score_data <span class="ot">&lt;-</span></span>
<span id="cb86-768"><a href="#cb86-768" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(</span>
<span id="cb86-769"><a href="#cb86-769" aria-hidden="true" tabindex="-1"></a>    <span class="fu">paste0</span>(<span class="st">"x"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>),</span>
<span id="cb86-770"><a href="#cb86-770" aria-hidden="true" tabindex="-1"></a>    <span class="cf">function</span>(x) <span class="fu">get_ss_by_var</span>(x, <span class="st">"rho"</span>, data_J1)</span>
<span id="cb86-771"><a href="#cb86-771" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span> </span>
<span id="cb86-772"><a href="#cb86-772" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rbindlist</span>()</span>
<span id="cb86-773"><a href="#cb86-773" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb86-774"><a href="#cb86-774" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-775"><a href="#cb86-775" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-776"><a href="#cb86-776" aria-hidden="true" tabindex="-1"></a>Now, we ensure that we have at least as many <span class="in">`min.node.size`</span> numbers of treated and control units in both child nodes. This provides a safeguard against having a very inaccurate treatment effect estimation. The default value of <span class="in">`min.node.size`</span> is 5. </span>
<span id="cb86-777"><a href="#cb86-777" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-780"><a href="#cb86-780" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-781"><a href="#cb86-781" aria-hidden="true" tabindex="-1"></a>min.node.size <span class="ot">&lt;-</span> <span class="dv">5</span> <span class="co"># referred to as mns</span></span>
<span id="cb86-782"><a href="#cb86-782" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-783"><a href="#cb86-783" aria-hidden="true" tabindex="-1"></a><span class="co">#=== check if both child nodes have at least mns control and treatment units  ===#</span></span>
<span id="cb86-784"><a href="#cb86-784" aria-hidden="true" tabindex="-1"></a>thr_score_data[, mns_met_grup <span class="sc">:</span><span class="er">=</span> <span class="fu">all</span>(num_ctrl <span class="sc">&gt;=</span> min.node.size <span class="sc">&amp;</span> num_treated <span class="sc">&gt;=</span> min.node.size), by <span class="ot">=</span> .(var, thr)]</span>
<span id="cb86-785"><a href="#cb86-785" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-786"><a href="#cb86-786" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb86-787"><a href="#cb86-787" aria-hidden="true" tabindex="-1"></a>msn_met_data <span class="ot">&lt;-</span> thr_score_data[mns_met_grup <span class="sc">==</span> <span class="cn">TRUE</span>, ]</span>
<span id="cb86-788"><a href="#cb86-788" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb86-789"><a href="#cb86-789" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-790"><a href="#cb86-790" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-791"><a href="#cb86-791" aria-hidden="true" tabindex="-1"></a>Now, we consider how <span class="in">`alpha`</span> and <span class="in">`imbalance.penalty`</span> affect the potential pool of feature-threshold combinations. Each child node needs to have at least as large <span class="in">`info_size`</span> as <span class="in">`alpha`</span> $\times$ <span class="in">`info_size`</span> of the parent node. The <span class="in">`info_size`</span> of the parent node is </span>
<span id="cb86-792"><a href="#cb86-792" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-795"><a href="#cb86-795" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-796"><a href="#cb86-796" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb86-797"><a href="#cb86-797" aria-hidden="true" tabindex="-1"></a>info_size_p <span class="ot">&lt;-</span> data_J1[, <span class="fu">sum</span>((T <span class="sc">-</span> <span class="fu">mean</span>(T))<span class="sc">^</span><span class="dv">2</span>)]</span>
<span id="cb86-798"><a href="#cb86-798" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb86-799"><a href="#cb86-799" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-800"><a href="#cb86-800" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-801"><a href="#cb86-801" aria-hidden="true" tabindex="-1"></a>By default, <span class="in">`alpha`</span> is set to 0.05. We use this number.</span>
<span id="cb86-802"><a href="#cb86-802" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-803"><a href="#cb86-803" aria-hidden="true" tabindex="-1"></a><span class="in">```{r alpha}</span></span>
<span id="cb86-804"><a href="#cb86-804" aria-hidden="true" tabindex="-1"></a>alpha <span class="ot">&lt;-</span> <span class="fl">0.05</span></span>
<span id="cb86-805"><a href="#cb86-805" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-806"><a href="#cb86-806" aria-hidden="true" tabindex="-1"></a>msn_met_data[, info_size_met <span class="sc">:</span><span class="er">=</span> <span class="fu">all</span>(info_size <span class="sc">&gt;</span> (alpha <span class="sc">*</span> info_size_p)), by <span class="ot">=</span> .(var, thr)]</span>
<span id="cb86-807"><a href="#cb86-807" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-808"><a href="#cb86-808" aria-hidden="true" tabindex="-1"></a>msn_ifs_met <span class="ot">&lt;-</span> msn_met_data[info_size_met <span class="sc">==</span> <span class="cn">TRUE</span>, ]</span>
<span id="cb86-809"><a href="#cb86-809" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-810"><a href="#cb86-810" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-811"><a href="#cb86-811" aria-hidden="true" tabindex="-1"></a><span class="in">`imbalance.penalty`</span> is used to further punish splits that introduce imbalance. For example, for feature <span class="in">`x1`</span> and <span class="in">`thr`</span> of <span class="in">`r msn_ifs_met[1, thr]`</span>, </span>
<span id="cb86-812"><a href="#cb86-812" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-815"><a href="#cb86-815" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-816"><a href="#cb86-816" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb86-817"><a href="#cb86-817" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb86-818"><a href="#cb86-818" aria-hidden="true" tabindex="-1"></a>ex_score <span class="ot">&lt;-</span> </span>
<span id="cb86-819"><a href="#cb86-819" aria-hidden="true" tabindex="-1"></a>  msn_met_data[var <span class="sc">==</span> <span class="st">"x1"</span>, ] <span class="sc">%&gt;%</span> </span>
<span id="cb86-820"><a href="#cb86-820" aria-hidden="true" tabindex="-1"></a>  .[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>,]</span>
<span id="cb86-821"><a href="#cb86-821" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb86-822"><a href="#cb86-822" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-823"><a href="#cb86-823" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-824"><a href="#cb86-824" aria-hidden="true" tabindex="-1"></a>the unpunished heterogeneity score is <span class="in">`r ex_score[, sum(het_score)]`</span>. With a non-zero value of <span class="in">`imbalance.penalty`</span>, the following penalty will be subtracted from the unpunished heterogeneity score: <span class="in">`imbalance.penalty * (1/info_size_1 + 1/info_size_2)`</span>. By default, <span class="in">`imbalance.penalty`</span> is 0. But, let's use <span class="in">`imbalance.penalty`</span> $= 1$ here.</span>
<span id="cb86-825"><a href="#cb86-825" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-828"><a href="#cb86-828" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-829"><a href="#cb86-829" aria-hidden="true" tabindex="-1"></a>imbalance.penalty <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb86-830"><a href="#cb86-830" aria-hidden="true" tabindex="-1"></a>msn_ifs_met[, imb_penalty <span class="sc">:</span><span class="er">=</span> imbalance.penalty <span class="sc">*</span> (<span class="dv">1</span><span class="sc">/</span>info_size)]</span>
<span id="cb86-831"><a href="#cb86-831" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-832"><a href="#cb86-832" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-833"><a href="#cb86-833" aria-hidden="true" tabindex="-1"></a>Now, we calculate the heterogeneity score for each feature-threshold with imbalance penalty.</span>
<span id="cb86-834"><a href="#cb86-834" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-835"><a href="#cb86-835" aria-hidden="true" tabindex="-1"></a><span class="in">```{r imbp-score}</span></span>
<span id="cb86-836"><a href="#cb86-836" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb86-837"><a href="#cb86-837" aria-hidden="true" tabindex="-1"></a>het_score_data <span class="ot">&lt;-</span> msn_ifs_met[, .(<span class="at">het_with_imbp =</span> <span class="fu">sum</span>(het_score <span class="sc">-</span> imb_penalty)), <span class="at">by =</span> .(var, thr)]</span>
<span id="cb86-838"><a href="#cb86-838" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb86-839"><a href="#cb86-839" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-840"><a href="#cb86-840" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-841"><a href="#cb86-841" aria-hidden="true" tabindex="-1"></a>We now find the feature-threshold combination that maximizes the score,</span>
<span id="cb86-842"><a href="#cb86-842" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-845"><a href="#cb86-845" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-846"><a href="#cb86-846" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb86-847"><a href="#cb86-847" aria-hidden="true" tabindex="-1"></a>best_split <span class="ot">&lt;-</span> het_score_data[<span class="fu">which.max</span>(het_with_imbp), ]</span>
<span id="cb86-848"><a href="#cb86-848" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb86-849"><a href="#cb86-849" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-850"><a href="#cb86-850" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-851"><a href="#cb86-851" aria-hidden="true" tabindex="-1"></a>So, we are splitting the root node using <span class="in">`x1`</span> with threshold of <span class="in">`r best_split[, round(thr, digits = 4)]`</span>. Here is a visualization of the split,</span>
<span id="cb86-852"><a href="#cb86-852" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-853"><a href="#cb86-853" aria-hidden="true" tabindex="-1"></a><span class="in">```{r viz-first-split}</span></span>
<span id="cb86-854"><a href="#cb86-854" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb86-855"><a href="#cb86-855" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(data_J1) <span class="sc">+</span></span>
<span id="cb86-856"><a href="#cb86-856" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">y =</span> rho, <span class="at">x =</span> x1, <span class="at">color =</span> (x1 <span class="sc">&lt;</span> best_split[, thr]))) <span class="sc">+</span></span>
<span id="cb86-857"><a href="#cb86-857" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_discrete</span>(<span class="st">"Less than the threshold"</span>) <span class="sc">+</span></span>
<span id="cb86-858"><a href="#cb86-858" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb86-859"><a href="#cb86-859" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"bottom"</span>)</span>
<span id="cb86-860"><a href="#cb86-860" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-861"><a href="#cb86-861" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-862"><a href="#cb86-862" aria-hidden="true" tabindex="-1"></a>Now, let's look at the pseudo outcome and elaborate more on what we are seeing here.</span>
<span id="cb86-863"><a href="#cb86-863" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-864"><a href="#cb86-864" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-865"><a href="#cb86-865" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb86-866"><a href="#cb86-866" aria-hidden="true" tabindex="-1"></a>\rho_i = (\tilde{Y}_i- \hat{\theta}_P\tilde{T}_i)\tilde{T}_i</span>
<span id="cb86-867"><a href="#cb86-867" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb86-868"><a href="#cb86-868" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-869"><a href="#cb86-869" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-870"><a href="#cb86-870" aria-hidden="true" tabindex="-1"></a>Note that $\hat{\theta}_P$ is just a constant and represents the average treatment effect for the parent node. So, $\rho_i$ tends to be higher for the observations whose CATE is higher (above-average treatment effect). So, splitting on $\rho$ (instead of $Y$) made it possible to pick the right feature variable <span class="in">`x1`</span>. This is because <span class="in">`x1`</span> is influential in determining CATE and higher values of <span class="in">`x1`</span> lead to more positive CATE.</span>
<span id="cb86-871"><a href="#cb86-871" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-872"><a href="#cb86-872" aria-hidden="true" tabindex="-1"></a>Note that the checks implemented with <span class="in">`mtry`</span>, <span class="in">`alpha`</span>, and  <span class="in">`imbalance.penalty`</span> is more important further down a tree. But, the examples above should have illustrated how they are used. </span>
<span id="cb86-873"><a href="#cb86-873" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-874"><a href="#cb86-874" aria-hidden="true" tabindex="-1"></a>Okay, now let's split the root node into two child nodes according to the best split on $\rho$ we found earlier.</span>
<span id="cb86-875"><a href="#cb86-875" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-878"><a href="#cb86-878" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-879"><a href="#cb86-879" aria-hidden="true" tabindex="-1"></a>depth_2_node_1 <span class="ot">&lt;-</span> data_J1[x1 <span class="sc">&lt;</span> best_split[, thr], ]</span>
<span id="cb86-880"><a href="#cb86-880" aria-hidden="true" tabindex="-1"></a>depth_2_node_2 <span class="ot">&lt;-</span> data_J1[x1 <span class="sc">&gt;=</span> best_split[, thr], ]</span>
<span id="cb86-881"><a href="#cb86-881" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-882"><a href="#cb86-882" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-883"><a href="#cb86-883" aria-hidden="true" tabindex="-1"></a>Let's consider splitting the second node, <span class="in">`depth_2_node_1`</span>. We basically follow exactly the same process the we just did. First, find $\hat{\theta}_P$ for this (parent) node using @eq-theta-p and define the pseudo outcomes.</span>
<span id="cb86-884"><a href="#cb86-884" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-887"><a href="#cb86-887" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-888"><a href="#cb86-888" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb86-889"><a href="#cb86-889" aria-hidden="true" tabindex="-1"></a>theta_p <span class="ot">&lt;-</span> depth_2_node_1[, <span class="fu">sum</span>(y_tilde <span class="sc">*</span> t_tilde)<span class="sc">/</span><span class="fu">sum</span>(t_tilde <span class="sc">*</span> t_tilde)]</span>
<span id="cb86-890"><a href="#cb86-890" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb86-891"><a href="#cb86-891" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-892"><a href="#cb86-892" aria-hidden="true" tabindex="-1"></a>depth_2_node_1[, rho <span class="sc">:</span><span class="er">=</span> (y_tilde <span class="sc">-</span> theta_p <span class="sc">*</span> t_tilde) <span class="sc">*</span> t_tilde]</span>
<span id="cb86-893"><a href="#cb86-893" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-894"><a href="#cb86-894" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-895"><a href="#cb86-895" aria-hidden="true" tabindex="-1"></a>Now, we find the best split (here we ignore various checks),</span>
<span id="cb86-896"><a href="#cb86-896" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-899"><a href="#cb86-899" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-900"><a href="#cb86-900" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb86-901"><a href="#cb86-901" aria-hidden="true" tabindex="-1"></a>best_split_d2_n1 <span class="ot">&lt;-</span></span>
<span id="cb86-902"><a href="#cb86-902" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(</span>
<span id="cb86-903"><a href="#cb86-903" aria-hidden="true" tabindex="-1"></a>    <span class="fu">paste0</span>(<span class="st">"x"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>),</span>
<span id="cb86-904"><a href="#cb86-904" aria-hidden="true" tabindex="-1"></a>    <span class="cf">function</span>(x) <span class="fu">get_ss_by_var</span>(x, <span class="st">"rho"</span>, depth_2_node_1)</span>
<span id="cb86-905"><a href="#cb86-905" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span> </span>
<span id="cb86-906"><a href="#cb86-906" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rbindlist</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb86-907"><a href="#cb86-907" aria-hidden="true" tabindex="-1"></a>  .[, .(<span class="at">het_score =</span> <span class="fu">sum</span>(het_score)), <span class="at">by =</span> .(var, thr)] <span class="sc">%&gt;%</span> </span>
<span id="cb86-908"><a href="#cb86-908" aria-hidden="true" tabindex="-1"></a>  .[<span class="fu">which.max</span>(het_score), ]</span>
<span id="cb86-909"><a href="#cb86-909" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb86-910"><a href="#cb86-910" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-911"><a href="#cb86-911" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-912"><a href="#cb86-912" aria-hidden="true" tabindex="-1"></a>We keep splitting nodes until no splits is possible any more based on the value of <span class="in">`min.node.size`</span>, <span class="in">`alpha`</span>, and <span class="in">`imbalance.penalty`</span>. </span>
<span id="cb86-913"><a href="#cb86-913" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-914"><a href="#cb86-914" aria-hidden="true" tabindex="-1"></a>---</span>
<span id="cb86-915"><a href="#cb86-915" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-916"><a href="#cb86-916" aria-hidden="true" tabindex="-1"></a>To see the role of <span class="in">`honesty.prune.leaves`</span>, let's suppose we stopped splitting after the first split based on <span class="in">`x`</span> with threshold of <span class="in">`r best_split[, round(thr, digits = 4)]`</span>. So, we just two terminal nodes. As stated in @sec-grf-honest, when we do prediction, <span class="in">`data_J1`</span> (which is used for determining the splitting rule) is not used. Rather, <span class="in">`data_J2`</span> (the data that was set aside) is used. When <span class="in">`honesty.prune.leaves = TRUE`</span> (default), the tree is pruned so that no leaves are empty. Yes, we made sure that we have certain number of observations in each node with <span class="in">`min.node.size`</span>, however, that is only for <span class="in">`data_J1`</span>, not <span class="in">`data_J2`</span>. </span>
<span id="cb86-917"><a href="#cb86-917" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-920"><a href="#cb86-920" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-921"><a href="#cb86-921" aria-hidden="true" tabindex="-1"></a>data_J2[, .N, by <span class="ot">=</span> x1 <span class="sc">&lt;</span> best_split[, thr]]</span>
<span id="cb86-922"><a href="#cb86-922" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-923"><a href="#cb86-923" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-924"><a href="#cb86-924" aria-hidden="true" tabindex="-1"></a>Okay, so, each child has at least one observation. So, we are not going to prune the tree. If either of the child nodes was empty, then we would have removed the leaves and had the root node as the tree. </span>
<span id="cb86-925"><a href="#cb86-925" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-926"><a href="#cb86-926" aria-hidden="true" tabindex="-1"></a>As you can see, while <span class="in">`data_J1`</span> plays a central role in the tree building process, the other half, <span class="in">`data_J2`</span>, plays a very small role. However, <span class="in">`data_J2`</span> plays the central role in prediction, which will be seen in the next section.</span>
<span id="cb86-927"><a href="#cb86-927" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-928"><a href="#cb86-928" aria-hidden="true" tabindex="-1"></a><span class="fu">### Prediction</span></span>
<span id="cb86-929"><a href="#cb86-929" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-930"><a href="#cb86-930" aria-hidden="true" tabindex="-1"></a>Suppose you have built $T$ trees already. In general for GRF, statistics of interest, $\theta(X)$, at $X = X_0$ is found by solving the following equation:</span>
<span id="cb86-931"><a href="#cb86-931" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-932"><a href="#cb86-932" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-933"><a href="#cb86-933" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb86-934"><a href="#cb86-934" aria-hidden="true" tabindex="-1"></a>\sum_{i=1}^N \alpha_i(X_i, X_0)\psi_{\theta,\nu}(O_i) = 0 </span>
<span id="cb86-935"><a href="#cb86-935" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb86-936"><a href="#cb86-936" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-937"><a href="#cb86-937" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-938"><a href="#cb86-938" aria-hidden="true" tabindex="-1"></a>where $\alpha_i(X_i, X_0)$ is the weight given to each $i$ calculated based on the trees that have been built. Let $\eta_{i,t}(X_i, X_0)$ be 1 if observation $i$ belongs to the same leaf as $X_0$ in tree $t$. Then, </span>
<span id="cb86-939"><a href="#cb86-939" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-940"><a href="#cb86-940" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-941"><a href="#cb86-941" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb86-942"><a href="#cb86-942" aria-hidden="true" tabindex="-1"></a>\alpha_i(X_i, X_0) = \frac{1}{T}\sum_{t=1}^T\frac{\eta_{i,t}(X_i, X_0)}{\sum_{i=1}^N\eta_{i,t}(X_i, X_0)}</span>
<span id="cb86-943"><a href="#cb86-943" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb86-944"><a href="#cb86-944" aria-hidden="true" tabindex="-1"></a>$$ {#eq-weight-cf}</span>
<span id="cb86-945"><a href="#cb86-945" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-946"><a href="#cb86-946" aria-hidden="true" tabindex="-1"></a>So, the weight given to observation $i$ is higher if observation $i$ belongs to the same leaf as the evaluation point $X_0$ in more trees. </span>
<span id="cb86-947"><a href="#cb86-947" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-948"><a href="#cb86-948" aria-hidden="true" tabindex="-1"></a>For CF, since the score function is defined as $\psi_{\theta,\nu}(O_i) = \tilde{Y}_i - \theta\tilde{T}_i)\cdot \tilde{T}_i$, the equation can be written as</span>
<span id="cb86-949"><a href="#cb86-949" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-950"><a href="#cb86-950" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-951"><a href="#cb86-951" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb86-952"><a href="#cb86-952" aria-hidden="true" tabindex="-1"></a>\sum_{i=1}^N \alpha_i(X_i, X_0)<span class="co">[</span><span class="ot">\tilde{Y_i} - \theta\cdot \tilde{T_i}</span><span class="co">]</span>\hat{\tilde{T_i}} = 0 </span>
<span id="cb86-953"><a href="#cb86-953" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb86-954"><a href="#cb86-954" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-955"><a href="#cb86-955" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-956"><a href="#cb86-956" aria-hidden="true" tabindex="-1"></a>So,</span>
<span id="cb86-957"><a href="#cb86-957" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-958"><a href="#cb86-958" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-959"><a href="#cb86-959" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb86-960"><a href="#cb86-960" aria-hidden="true" tabindex="-1"></a>\theta_P = \frac{\sum_{i=1}^n \alpha_i(X_i, X_0)\tilde{Y}_i\cdot \tilde{T}_i}{\sum_{i=1}^n\alpha_i(X_i, X_0)\tilde{T}_i\cdot \tilde{T}_i}</span>
<span id="cb86-961"><a href="#cb86-961" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb86-962"><a href="#cb86-962" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-963"><a href="#cb86-963" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-964"><a href="#cb86-964" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-965"><a href="#cb86-965" aria-hidden="true" tabindex="-1"></a>---</span>
<span id="cb86-966"><a href="#cb86-966" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-967"><a href="#cb86-967" aria-hidden="true" tabindex="-1"></a>We now illustrate how predictions are done once trees have been built. Let's first build trees using <span class="in">`causal_forest()`</span>. <span class="in">`min.node.size`</span> is set deliberately high so we can work with very simple trees. </span>
<span id="cb86-968"><a href="#cb86-968" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-969"><a href="#cb86-969" aria-hidden="true" tabindex="-1"></a>::: {.column-margin}</span>
<span id="cb86-970"><a href="#cb86-970" aria-hidden="true" tabindex="-1"></a>You do not need to know how <span class="in">`causal_forest()`</span> works. You only need to know that <span class="in">`causal_forest()`</span> build trees for CATE estimation.</span>
<span id="cb86-971"><a href="#cb86-971" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb86-972"><a href="#cb86-972" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-975"><a href="#cb86-975" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-976"><a href="#cb86-976" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">443784</span>)</span>
<span id="cb86-977"><a href="#cb86-977" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-978"><a href="#cb86-978" aria-hidden="true" tabindex="-1"></a>cf_trained <span class="ot">&lt;-</span></span>
<span id="cb86-979"><a href="#cb86-979" aria-hidden="true" tabindex="-1"></a>  <span class="fu">causal_forest</span>(</span>
<span id="cb86-980"><a href="#cb86-980" aria-hidden="true" tabindex="-1"></a>    <span class="at">X =</span> <span class="fu">select</span>(data, <span class="fu">starts_with</span>(<span class="st">"x"</span>)),</span>
<span id="cb86-981"><a href="#cb86-981" aria-hidden="true" tabindex="-1"></a>    <span class="at">Y =</span> data[, y],</span>
<span id="cb86-982"><a href="#cb86-982" aria-hidden="true" tabindex="-1"></a>    <span class="at">W =</span> data[, T],</span>
<span id="cb86-983"><a href="#cb86-983" aria-hidden="true" tabindex="-1"></a>    <span class="at">min.node.size =</span> <span class="dv">50</span></span>
<span id="cb86-984"><a href="#cb86-984" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb86-985"><a href="#cb86-985" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-986"><a href="#cb86-986" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-987"><a href="#cb86-987" aria-hidden="true" tabindex="-1"></a>After training a causal forest model, we have trees like the one shown in @fig-cf-tree, which is the first of the $2000$ trees.</span>
<span id="cb86-988"><a href="#cb86-988" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-991"><a href="#cb86-991" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-992"><a href="#cb86-992" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: Example trees built in causal forest estimation</span></span>
<span id="cb86-993"><a href="#cb86-993" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-cf-tree </span></span>
<span id="cb86-994"><a href="#cb86-994" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false </span></span>
<span id="cb86-995"><a href="#cb86-995" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 4</span></span>
<span id="cb86-996"><a href="#cb86-996" aria-hidden="true" tabindex="-1"></a><span class="fu">get_tree</span>(cf_trained, <span class="dv">1</span>) <span class="sc">%&gt;%</span> <span class="fu">plot</span>()</span>
<span id="cb86-997"><a href="#cb86-997" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-998"><a href="#cb86-998" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-999"><a href="#cb86-999" aria-hidden="true" tabindex="-1"></a>You probably noticed that the total number of samples in the leaves is only $250$ instead of $1000$, which is the total number of observations in <span class="in">`data`</span>. When causal forest was trained on this dataset, only half of the entire sample are randomly selected for building each tree (due to the default setting of <span class="in">`sample.fraction = 0.5`</span>). The halved sample is further split into two groups, each containing $250$ observations (due to the default setting of <span class="in">`honesty = TRUE`</span> and <span class="in">`honest.fraction = 0.5`</span>). Let's call them $J_1$ and $J_2$. Then, $J_1$ is used to train a tree to find the splitting rules (e.g., $x_1 \leq 1.44$ for the first tree). See @sec-build-trees for how only a subset of the drawn samples was used to determine the splitting rules. Once the splitting rules are determined (tree building process is complete), then $J_1$ "vacates" the tree. Then, $J_2$ repopulates the tree. In other words, the splitting rule associated with the tree is applied to each of the samples in $J_2$ and place it in the node they belong to. So, the size value in each of the terminal nodes represents how many samples in $J_2$ are in each node. The statistics for each terminal node are from the samples in $J_2$ within the node. This will become clearer when we talk about finding individual weights.</span>
<span id="cb86-1000"><a href="#cb86-1000" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1001"><a href="#cb86-1001" aria-hidden="true" tabindex="-1"></a>Let's take a look at a tree to see what happened. We can use <span class="in">`get_tree()`</span> to access individual trees from <span class="in">`cf_trained`</span>. </span>
<span id="cb86-1002"><a href="#cb86-1002" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1005"><a href="#cb86-1005" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-1006"><a href="#cb86-1006" aria-hidden="true" tabindex="-1"></a><span class="co">#=== get the first tree ===#</span></span>
<span id="cb86-1007"><a href="#cb86-1007" aria-hidden="true" tabindex="-1"></a>a_tree <span class="ot">&lt;-</span> <span class="fu">get_tree</span>(cf_trained, <span class="dv">1</span>)</span>
<span id="cb86-1008"><a href="#cb86-1008" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-1009"><a href="#cb86-1009" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1010"><a href="#cb86-1010" aria-hidden="true" tabindex="-1"></a><span class="in">`drawn_samples`</span> attribute of the tree contains row indices that are selected randomly for this tree.</span>
<span id="cb86-1011"><a href="#cb86-1011" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1014"><a href="#cb86-1014" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-1015"><a href="#cb86-1015" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(a_tree<span class="sc">$</span>drawn_samples)</span>
<span id="cb86-1016"><a href="#cb86-1016" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1017"><a href="#cb86-1017" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(a_tree<span class="sc">$</span>drawn_samples)</span>
<span id="cb86-1018"><a href="#cb86-1018" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-1019"><a href="#cb86-1019" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1020"><a href="#cb86-1020" aria-hidden="true" tabindex="-1"></a>As you can see, there are 500 samples (due to <span class="in">`sample.fraction = 0.5`</span>). The rest of the observations were not used for this tree at all. Accessing <span class="in">`nodes`</span> attribute will give you the splitting rules for the tree built and which samples are in what node.</span>
<span id="cb86-1021"><a href="#cb86-1021" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1024"><a href="#cb86-1024" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-1025"><a href="#cb86-1025" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb86-1026"><a href="#cb86-1026" aria-hidden="true" tabindex="-1"></a>nodes <span class="ot">&lt;-</span> a_tree<span class="sc">$</span>nodes</span>
<span id="cb86-1027"><a href="#cb86-1027" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb86-1028"><a href="#cb86-1028" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-1029"><a href="#cb86-1029" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1030"><a href="#cb86-1030" aria-hidden="true" tabindex="-1"></a><span class="in">`nodes`</span> is a list of three elements (one root node and two terminal nodes here). The <span class="in">`samples`</span> attribute gives you row indices of the samples that belong to the terminal nodes. </span>
<span id="cb86-1031"><a href="#cb86-1031" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1034"><a href="#cb86-1034" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-1035"><a href="#cb86-1035" aria-hidden="true" tabindex="-1"></a><span class="co">#=== left node ===#  </span></span>
<span id="cb86-1036"><a href="#cb86-1036" aria-hidden="true" tabindex="-1"></a>nodes[[<span class="dv">2</span>]]<span class="sc">$</span>samples</span>
<span id="cb86-1037"><a href="#cb86-1037" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1038"><a href="#cb86-1038" aria-hidden="true" tabindex="-1"></a><span class="co">#=== right node ===#  </span></span>
<span id="cb86-1039"><a href="#cb86-1039" aria-hidden="true" tabindex="-1"></a>nodes[[<span class="dv">3</span>]]<span class="sc">$</span>samples</span>
<span id="cb86-1040"><a href="#cb86-1040" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-1041"><a href="#cb86-1041" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1042"><a href="#cb86-1042" aria-hidden="true" tabindex="-1"></a>It is important to keep in mind that these observations with these row indices belong to $J_2$. These observations were <span class="kw">&lt;span</span> <span class="er">style</span><span class="ot">=</span><span class="st">"color:red"</span><span class="kw">&gt;</span>NOT<span class="kw">&lt;/span&gt;</span> used in determining the splitting rule of $x_1 \leq 1.44$. They were populating the terminal nodes by simply following the splitting rule, which is determined using the data in $J_1$ following the process described in @sec-build-trees. The difference in <span class="in">`a_tree$drawn_samples`</span> and the combination of <span class="in">`nodes[[2]]$samples`</span> and <span class="in">`nodes[[3]]$samples`</span> is $J_1$.</span>
<span id="cb86-1043"><a href="#cb86-1043" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1046"><a href="#cb86-1046" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-1047"><a href="#cb86-1047" aria-hidden="true" tabindex="-1"></a>J2_rows <span class="ot">&lt;-</span> <span class="fu">c</span>(nodes[[<span class="dv">2</span>]]<span class="sc">$</span>samples, nodes[[<span class="dv">3</span>]]<span class="sc">$</span>samples)</span>
<span id="cb86-1048"><a href="#cb86-1048" aria-hidden="true" tabindex="-1"></a>J1_J2_rows <span class="ot">&lt;-</span> a_tree<span class="sc">$</span>drawn_samples</span>
<span id="cb86-1049"><a href="#cb86-1049" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1050"><a href="#cb86-1050" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb86-1051"><a href="#cb86-1051" aria-hidden="true" tabindex="-1"></a>J1_rows <span class="ot">&lt;-</span> J1_J2_rows[J1_J2_rows <span class="sc">%in%</span> J2_rows]</span>
<span id="cb86-1052"><a href="#cb86-1052" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb86-1053"><a href="#cb86-1053" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-1054"><a href="#cb86-1054" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1055"><a href="#cb86-1055" aria-hidden="true" tabindex="-1"></a>As you can see, there are 250 samples in $J_1$ as well. They were used to find the splitting rule. </span>
<span id="cb86-1056"><a href="#cb86-1056" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1057"><a href="#cb86-1057" aria-hidden="true" tabindex="-1"></a>For all the trees, the information we went through here is stored and can be used for predicting $\theta(X)$.</span>
<span id="cb86-1058"><a href="#cb86-1058" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1059"><a href="#cb86-1059" aria-hidden="true" tabindex="-1"></a>---</span>
<span id="cb86-1060"><a href="#cb86-1060" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1061"><a href="#cb86-1061" aria-hidden="true" tabindex="-1"></a>Suppose you are interested in predicting $\hat{\theta}$ (CATE, here) at $X_0 = <span class="sc">\{</span>x_1 = 2, x_2,\dots, x_{10} = 1<span class="sc">\}</span>$. We can simply use <span class="in">`predict()`</span> function in practice. But, we look into what is happening in <span class="in">`prediction()`</span> behind the scene here. </span>
<span id="cb86-1062"><a href="#cb86-1062" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1063"><a href="#cb86-1063" aria-hidden="true" tabindex="-1"></a>For a given tree, we give 1 to the observations in <span class="kw">&lt;span</span> <span class="er">style</span><span class="ot">=</span><span class="st">"color:red"</span><span class="kw">&gt;</span>$J_2$<span class="kw">&lt;/span&gt;</span> that belong to the same leaf as $X_0$. For example, for the first tree, $X_0$ belongs to the right leaf because $x1 = 2 &gt; 1.44$ for $X_0$. We can tell which node $X_0$ belongs to by supplying $X_0$ to <span class="in">`get_leaf_node()`</span>.</span>
<span id="cb86-1064"><a href="#cb86-1064" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1067"><a href="#cb86-1067" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-1068"><a href="#cb86-1068" aria-hidden="true" tabindex="-1"></a>X_0 <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="dv">2</span>, <span class="fu">rep</span>(<span class="dv">1</span>, <span class="dv">9</span>)), <span class="at">nrow =</span> <span class="dv">1</span>)</span>
<span id="cb86-1069"><a href="#cb86-1069" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1070"><a href="#cb86-1070" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb86-1071"><a href="#cb86-1071" aria-hidden="true" tabindex="-1"></a>which_tree_is_X0_in <span class="ot">&lt;-</span> <span class="fu">get_leaf_node</span>(a_tree, X_0)</span>
<span id="cb86-1072"><a href="#cb86-1072" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb86-1073"><a href="#cb86-1073" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-1074"><a href="#cb86-1074" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1075"><a href="#cb86-1075" aria-hidden="true" tabindex="-1"></a>So, we give $1/N_t(X_0)$ to all those in the right leaf (the third node in <span class="in">`nodes`</span>) and 0 to those in the left leaf, where $N_t(X_0)$ is the number of observations that belong to the same leaf as $X_0$. All the other observations are assigned 0. </span>
<span id="cb86-1076"><a href="#cb86-1076" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1079"><a href="#cb86-1079" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-1080"><a href="#cb86-1080" aria-hidden="true" tabindex="-1"></a><span class="co">#=== which row numbers in the same leaf as X_0? ===#</span></span>
<span id="cb86-1081"><a href="#cb86-1081" aria-hidden="true" tabindex="-1"></a>rows_1 <span class="ot">&lt;-</span> nodes[[which_tree_is_X0_in]]<span class="sc">$</span>samples</span>
<span id="cb86-1082"><a href="#cb86-1082" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1083"><a href="#cb86-1083" aria-hidden="true" tabindex="-1"></a><span class="co">#=== define eta for tree 1  ===#</span></span>
<span id="cb86-1084"><a href="#cb86-1084" aria-hidden="true" tabindex="-1"></a>data[, eta_t1 <span class="sc">:</span><span class="er">=</span> <span class="dv">0</span>] <span class="co"># first set eta to 0 for all the observations</span></span>
<span id="cb86-1085"><a href="#cb86-1085" aria-hidden="true" tabindex="-1"></a>data[rows_1, eta_t1 <span class="sc">:</span><span class="er">=</span> <span class="dv">1</span> <span class="sc">/</span> <span class="fu">length</span>(rows_1)] <span class="co"># replace eta with 1/N_t(X_0) if in the right node</span></span>
<span id="cb86-1086"><a href="#cb86-1086" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1087"><a href="#cb86-1087" aria-hidden="true" tabindex="-1"></a><span class="co">#=== see the data ===#</span></span>
<span id="cb86-1088"><a href="#cb86-1088" aria-hidden="true" tabindex="-1"></a>data</span>
<span id="cb86-1089"><a href="#cb86-1089" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-1090"><a href="#cb86-1090" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1091"><a href="#cb86-1091" aria-hidden="true" tabindex="-1"></a>We repeat this for all the trees and use @eq-weight-cf to calculate the weights for the individual observations. The following function gets $\eta_{i,t}(X_i, X_0)$ for a given tree for all the observations.</span>
<span id="cb86-1092"><a href="#cb86-1092" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1095"><a href="#cb86-1095" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-1096"><a href="#cb86-1096" aria-hidden="true" tabindex="-1"></a>get_eta <span class="ot">&lt;-</span> <span class="cf">function</span>(t, X_0) {</span>
<span id="cb86-1097"><a href="#cb86-1097" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1098"><a href="#cb86-1098" aria-hidden="true" tabindex="-1"></a>  w_tree <span class="ot">&lt;-</span> <span class="fu">get_tree</span>(cf_trained, t)</span>
<span id="cb86-1099"><a href="#cb86-1099" aria-hidden="true" tabindex="-1"></a>  which_tree_is_X0_in <span class="ot">&lt;-</span> <span class="fu">get_leaf_node</span>(w_tree, X_0)</span>
<span id="cb86-1100"><a href="#cb86-1100" aria-hidden="true" tabindex="-1"></a>  rows <span class="ot">&lt;-</span> w_tree<span class="sc">$</span>nodes[[which_tree_is_X0_in]]<span class="sc">$</span>samples</span>
<span id="cb86-1101"><a href="#cb86-1101" aria-hidden="true" tabindex="-1"></a>  eta_data <span class="ot">&lt;-</span> </span>
<span id="cb86-1102"><a href="#cb86-1102" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.table</span>(</span>
<span id="cb86-1103"><a href="#cb86-1103" aria-hidden="true" tabindex="-1"></a>      <span class="at">row_id =</span> <span class="fu">seq_len</span>(<span class="fu">nrow</span>(data)),</span>
<span id="cb86-1104"><a href="#cb86-1104" aria-hidden="true" tabindex="-1"></a>      <span class="at">eta =</span> <span class="fu">rep</span>(<span class="dv">0</span>, <span class="fu">nrow</span>(data))</span>
<span id="cb86-1105"><a href="#cb86-1105" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">%&gt;%</span> </span>
<span id="cb86-1106"><a href="#cb86-1106" aria-hidden="true" tabindex="-1"></a>    .[rows, eta <span class="sc">:</span><span class="er">=</span> <span class="dv">1</span> <span class="sc">/</span> <span class="fu">length</span>(rows)]</span>
<span id="cb86-1107"><a href="#cb86-1107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1108"><a href="#cb86-1108" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(eta_data)</span>
<span id="cb86-1109"><a href="#cb86-1109" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb86-1110"><a href="#cb86-1110" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1111"><a href="#cb86-1111" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-1112"><a href="#cb86-1112" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1113"><a href="#cb86-1113" aria-hidden="true" tabindex="-1"></a>We apply <span class="in">`get_eta()`</span> to each of the 2000 trees.</span>
<span id="cb86-1114"><a href="#cb86-1114" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1117"><a href="#cb86-1117" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-1118"><a href="#cb86-1118" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb86-1119"><a href="#cb86-1119" aria-hidden="true" tabindex="-1"></a>eta_all <span class="ot">&lt;-</span></span>
<span id="cb86-1120"><a href="#cb86-1120" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(</span>
<span id="cb86-1121"><a href="#cb86-1121" aria-hidden="true" tabindex="-1"></a>    <span class="dv">1</span><span class="sc">:</span><span class="dv">2000</span>,</span>
<span id="cb86-1122"><a href="#cb86-1122" aria-hidden="true" tabindex="-1"></a>    <span class="cf">function</span>(x) <span class="fu">get_eta</span>(x, X_0)</span>
<span id="cb86-1123"><a href="#cb86-1123" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span> </span>
<span id="cb86-1124"><a href="#cb86-1124" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rbindlist</span>(<span class="at">idcol =</span> <span class="st">"t"</span>)</span>
<span id="cb86-1125"><a href="#cb86-1125" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb86-1126"><a href="#cb86-1126" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-1127"><a href="#cb86-1127" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1128"><a href="#cb86-1128" aria-hidden="true" tabindex="-1"></a>Calculate the mean of $\eta_{i,t}$ by <span class="in">`row_id`</span> (observation) to calculate $\alpha(X_i, X_0)$.</span>
<span id="cb86-1129"><a href="#cb86-1129" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1132"><a href="#cb86-1132" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-1133"><a href="#cb86-1133" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb86-1134"><a href="#cb86-1134" aria-hidden="true" tabindex="-1"></a>weights <span class="ot">&lt;-</span> </span>
<span id="cb86-1135"><a href="#cb86-1135" aria-hidden="true" tabindex="-1"></a>  eta_all <span class="sc">%&gt;%</span> </span>
<span id="cb86-1136"><a href="#cb86-1136" aria-hidden="true" tabindex="-1"></a>  .[, .(<span class="at">weight =</span> <span class="fu">mean</span>(eta)), <span class="at">by =</span> row_id]</span>
<span id="cb86-1137"><a href="#cb86-1137" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb86-1138"><a href="#cb86-1138" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1139"><a href="#cb86-1139" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-1140"><a href="#cb86-1140" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1141"><a href="#cb86-1141" aria-hidden="true" tabindex="-1"></a>Here is the observations that was given the highest and lowest weights.</span>
<span id="cb86-1142"><a href="#cb86-1142" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1145"><a href="#cb86-1145" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-1146"><a href="#cb86-1146" aria-hidden="true" tabindex="-1"></a>data_with_wights <span class="ot">&lt;-</span> <span class="fu">cbind</span>(data, weights)</span>
<span id="cb86-1147"><a href="#cb86-1147" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1148"><a href="#cb86-1148" aria-hidden="true" tabindex="-1"></a><span class="co">#=== highest (1st) and lowest (2nd) ===#</span></span>
<span id="cb86-1149"><a href="#cb86-1149" aria-hidden="true" tabindex="-1"></a>data_with_wights[weight <span class="sc">%in%</span> <span class="fu">c</span>(<span class="fu">max</span>(weight), <span class="fu">min</span>(weight)), ]</span>
<span id="cb86-1150"><a href="#cb86-1150" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-1151"><a href="#cb86-1151" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1152"><a href="#cb86-1152" aria-hidden="true" tabindex="-1"></a>Then, we can use @eq-theta-solution to calculate $\hat{\theta}(X_0)$.</span>
<span id="cb86-1153"><a href="#cb86-1153" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1156"><a href="#cb86-1156" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-1157"><a href="#cb86-1157" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb86-1158"><a href="#cb86-1158" aria-hidden="true" tabindex="-1"></a>theta_X0 <span class="ot">&lt;-</span> <span class="fu">sum</span>(data_with_wights[, weight <span class="sc">*</span> (T<span class="sc">-</span>cf_trained<span class="sc">$</span>W.hat) <span class="sc">*</span> (y<span class="sc">-</span>cf_trained<span class="sc">$</span>Y.hat)]) <span class="sc">/</span> <span class="fu">sum</span>(data_with_wights[, weight <span class="sc">*</span> (T<span class="sc">-</span>cf_trained<span class="sc">$</span>W.hat)<span class="sc">^</span><span class="dv">2</span>])</span>
<span id="cb86-1159"><a href="#cb86-1159" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb86-1160"><a href="#cb86-1160" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-1161"><a href="#cb86-1161" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1162"><a href="#cb86-1162" aria-hidden="true" tabindex="-1"></a>Note that $Y.hat$ and $W.hat$ attributes of <span class="in">`cf_trained`</span> are the estimates of $E<span class="co">[</span><span class="ot">Y|X</span><span class="co">]</span>$ and $E<span class="co">[</span><span class="ot">T|X</span><span class="co">]</span>$, respectively. By subtracting them from $Y$ and $T$, $\tilde{Y}$ and $\tilde{T}$ are calculated in the above code.</span>
<span id="cb86-1163"><a href="#cb86-1163" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1164"><a href="#cb86-1164" aria-hidden="true" tabindex="-1"></a>To get the weights, we could have just used <span class="in">`get_forest_weights()`</span> like below:</span>
<span id="cb86-1165"><a href="#cb86-1165" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1168"><a href="#cb86-1168" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-1169"><a href="#cb86-1169" aria-hidden="true" tabindex="-1"></a>weights <span class="ot">&lt;-</span> <span class="fu">get_forest_weights</span>(cf_trained, X_0)</span>
<span id="cb86-1170"><a href="#cb86-1170" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-1171"><a href="#cb86-1171" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1172"><a href="#cb86-1172" aria-hidden="true" tabindex="-1"></a>Let's compare the weights and see if we did it right.</span>
<span id="cb86-1173"><a href="#cb86-1173" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1176"><a href="#cb86-1176" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-1177"><a href="#cb86-1177" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(<span class="fu">abs</span>(<span class="fu">as.matrix</span>(weights) <span class="sc">-</span> data_with_wights<span class="sc">$</span>weight))</span>
<span id="cb86-1178"><a href="#cb86-1178" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-1179"><a href="#cb86-1179" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1180"><a href="#cb86-1180" aria-hidden="true" tabindex="-1"></a>Great, virtually the same.</span>
<span id="cb86-1181"><a href="#cb86-1181" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1182"><a href="#cb86-1182" aria-hidden="true" tabindex="-1"></a><span class="fu">## Clustering</span></span>
<span id="cb86-1183"><a href="#cb86-1183" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1184"><a href="#cb86-1184" aria-hidden="true" tabindex="-1"></a>In this section, we will conduct MC simulations to see how cluserting affects CF estimation of CATE ($\theta(X)$) and the standard error of $\hat{\theta}(X)$. We use the folloinwg data generating process:</span>
<span id="cb86-1185"><a href="#cb86-1185" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1186"><a href="#cb86-1186" aria-hidden="true" tabindex="-1"></a>:::{.callout-note}</span>
<span id="cb86-1187"><a href="#cb86-1187" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1188"><a href="#cb86-1188" aria-hidden="true" tabindex="-1"></a><span class="fu">## DGP 1</span></span>
<span id="cb86-1189"><a href="#cb86-1189" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1190"><a href="#cb86-1190" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-1191"><a href="#cb86-1191" aria-hidden="true" tabindex="-1"></a>\begin{aligned}</span>
<span id="cb86-1192"><a href="#cb86-1192" aria-hidden="true" tabindex="-1"></a>Y_i &amp; = (3 x_{i,1} + 0.5(x_{i,2} x_{i,3}))\cdot T_i + x_{i,3} + 2 x_{i,3} x_{i,4} + \mu_i <span class="sc">\\</span></span>
<span id="cb86-1193"><a href="#cb86-1193" aria-hidden="true" tabindex="-1"></a>T_i|X_i &amp; = Bernouli((1+x_{i,1})/2) <span class="sc">\\</span></span>
<span id="cb86-1194"><a href="#cb86-1194" aria-hidden="true" tabindex="-1"></a>\mu_i|X_i &amp; = N(0,1)</span>
<span id="cb86-1195"><a href="#cb86-1195" aria-hidden="true" tabindex="-1"></a>\end{aligned}</span>
<span id="cb86-1196"><a href="#cb86-1196" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb86-1197"><a href="#cb86-1197" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1198"><a href="#cb86-1198" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>$x_1 \sim U<span class="co">[</span><span class="ot">0,1</span><span class="co">]</span>$</span>
<span id="cb86-1199"><a href="#cb86-1199" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>$x_2\sim U<span class="co">[</span><span class="ot">0,2</span><span class="co">]</span>$</span>
<span id="cb86-1200"><a href="#cb86-1200" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>$x_3\sim U<span class="co">[</span><span class="ot">0,5</span><span class="co">]</span>$</span>
<span id="cb86-1201"><a href="#cb86-1201" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>$x_4\dots, x_10\sim N(0,1)^7$</span>
<span id="cb86-1202"><a href="#cb86-1202" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>$\mu\sim N(0, 9)$</span>
<span id="cb86-1203"><a href="#cb86-1203" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1204"><a href="#cb86-1204" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb86-1205"><a href="#cb86-1205" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1208"><a href="#cb86-1208" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-1209"><a href="#cb86-1209" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb86-1210"><a href="#cb86-1210" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1211"><a href="#cb86-1211" aria-hidden="true" tabindex="-1"></a>gen_cl_var <span class="ot">&lt;-</span>  <span class="cf">function</span>(G, Ng){</span>
<span id="cb86-1212"><a href="#cb86-1212" aria-hidden="true" tabindex="-1"></a>  var <span class="ot">&lt;-</span></span>
<span id="cb86-1213"><a href="#cb86-1213" aria-hidden="true" tabindex="-1"></a>    MASS<span class="sc">::</span><span class="fu">mvrnorm</span>(</span>
<span id="cb86-1214"><a href="#cb86-1214" aria-hidden="true" tabindex="-1"></a>      G, <span class="at">mu =</span> <span class="fu">rep</span>(<span class="dv">0</span>, Ng), </span>
<span id="cb86-1215"><a href="#cb86-1215" aria-hidden="true" tabindex="-1"></a>      <span class="at">Sigma =</span> <span class="fu">matrix</span>(<span class="fl">0.75</span>, <span class="at">nrow =</span> Ng, <span class="at">ncol =</span> Ng) <span class="sc">+</span> <span class="fl">0.25</span> <span class="sc">*</span> <span class="fu">diag</span>(Ng)</span>
<span id="cb86-1216"><a href="#cb86-1216" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">%&gt;%</span> <span class="fu">t</span>() <span class="sc">%&gt;%</span> <span class="fu">c</span>()</span>
<span id="cb86-1217"><a href="#cb86-1217" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1218"><a href="#cb86-1218" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(var)</span>
<span id="cb86-1219"><a href="#cb86-1219" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb86-1220"><a href="#cb86-1220" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1221"><a href="#cb86-1221" aria-hidden="true" tabindex="-1"></a>gen_data <span class="ot">&lt;-</span> <span class="cf">function</span>(<span class="at">N =</span> <span class="dv">1000</span>, <span class="at">G =</span> <span class="dv">40</span>, <span class="at">K =</span> <span class="dv">10</span>){</span>
<span id="cb86-1222"><a href="#cb86-1222" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1223"><a href="#cb86-1223" aria-hidden="true" tabindex="-1"></a>  Ng <span class="ot">&lt;-</span> N<span class="sc">/</span>G <span class="co"># number of observations per group</span></span>
<span id="cb86-1224"><a href="#cb86-1224" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1225"><a href="#cb86-1225" aria-hidden="true" tabindex="-1"></a>  <span class="co">#=== error term ===#    </span></span>
<span id="cb86-1226"><a href="#cb86-1226" aria-hidden="true" tabindex="-1"></a>  mu <span class="ot">&lt;-</span> <span class="fu">gen_cl_var</span>(G, Ng) <span class="sc">*</span> <span class="dv">3</span></span>
<span id="cb86-1227"><a href="#cb86-1227" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1228"><a href="#cb86-1228" aria-hidden="true" tabindex="-1"></a>  <span class="co">#=== X ===#</span></span>
<span id="cb86-1229"><a href="#cb86-1229" aria-hidden="true" tabindex="-1"></a>  X <span class="ot">&lt;-</span> </span>
<span id="cb86-1230"><a href="#cb86-1230" aria-hidden="true" tabindex="-1"></a>    <span class="fu">lapply</span>(</span>
<span id="cb86-1231"><a href="#cb86-1231" aria-hidden="true" tabindex="-1"></a>      <span class="dv">1</span><span class="sc">:</span>K,</span>
<span id="cb86-1232"><a href="#cb86-1232" aria-hidden="true" tabindex="-1"></a>      <span class="cf">function</span>(x) <span class="fu">gen_cl_var</span>(G, Ng)</span>
<span id="cb86-1233"><a href="#cb86-1233" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb86-1234"><a href="#cb86-1234" aria-hidden="true" tabindex="-1"></a>    <span class="fu">do.call</span>(<span class="st">"cbind"</span>, .) <span class="sc">%&gt;%</span></span>
<span id="cb86-1235"><a href="#cb86-1235" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.table</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb86-1236"><a href="#cb86-1236" aria-hidden="true" tabindex="-1"></a>    <span class="fu">setnames</span>(<span class="fu">names</span>(.), <span class="fu">paste0</span>(<span class="st">"x"</span>, <span class="fu">seq_len</span>(K)))</span>
<span id="cb86-1237"><a href="#cb86-1237" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1238"><a href="#cb86-1238" aria-hidden="true" tabindex="-1"></a>  group <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">1</span><span class="sc">:</span>G, <span class="at">each =</span> Ng)</span>
<span id="cb86-1239"><a href="#cb86-1239" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1240"><a href="#cb86-1240" aria-hidden="true" tabindex="-1"></a>  data <span class="ot">&lt;-</span></span>
<span id="cb86-1241"><a href="#cb86-1241" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cbind</span>(X, mu, group) <span class="sc">%&gt;%</span> </span>
<span id="cb86-1242"><a href="#cb86-1242" aria-hidden="true" tabindex="-1"></a>    <span class="co">#=== convert x1 and x3 to unif ===#</span></span>
<span id="cb86-1243"><a href="#cb86-1243" aria-hidden="true" tabindex="-1"></a>    .[, <span class="st">`</span><span class="at">:=</span><span class="st">`</span>(</span>
<span id="cb86-1244"><a href="#cb86-1244" aria-hidden="true" tabindex="-1"></a>      <span class="at">x1 =</span> <span class="fu">pnorm</span>(x1, <span class="at">sd =</span> <span class="fu">sqrt</span>(<span class="dv">2</span>)),</span>
<span id="cb86-1245"><a href="#cb86-1245" aria-hidden="true" tabindex="-1"></a>      <span class="at">x2 =</span> <span class="fu">pnorm</span>(x2, <span class="at">sd =</span> <span class="fu">sqrt</span>(<span class="dv">2</span>)) <span class="sc">*</span> <span class="dv">2</span>,</span>
<span id="cb86-1246"><a href="#cb86-1246" aria-hidden="true" tabindex="-1"></a>      <span class="at">x3 =</span> <span class="fu">pnorm</span>(x3, <span class="at">sd =</span> <span class="fu">sqrt</span>(<span class="dv">2</span>)) <span class="sc">*</span> <span class="dv">5</span></span>
<span id="cb86-1247"><a href="#cb86-1247" aria-hidden="true" tabindex="-1"></a>    )] <span class="sc">%&gt;%</span> </span>
<span id="cb86-1248"><a href="#cb86-1248" aria-hidden="true" tabindex="-1"></a>    .[, T <span class="sc">:</span><span class="er">=</span> <span class="fu">runif</span>(N) <span class="sc">&lt;</span> ((<span class="fl">0.5</span><span class="sc">+</span>x1)<span class="sc">/</span><span class="dv">2</span>)] <span class="sc">%&gt;%</span> </span>
<span id="cb86-1249"><a href="#cb86-1249" aria-hidden="true" tabindex="-1"></a>    .[, theta_x <span class="sc">:</span><span class="er">=</span> <span class="dv">3</span> <span class="sc">*</span> x1 <span class="sc">+</span> x2 <span class="sc">*</span> x3 <span class="sc">/</span> <span class="dv">2</span>] <span class="sc">%&gt;%</span></span>
<span id="cb86-1250"><a href="#cb86-1250" aria-hidden="true" tabindex="-1"></a>    .[, g_x <span class="sc">:</span><span class="er">=</span> x3 <span class="sc">+</span> <span class="dv">2</span> <span class="sc">*</span> x1 <span class="sc">*</span> x4] <span class="sc">%&gt;%</span></span>
<span id="cb86-1251"><a href="#cb86-1251" aria-hidden="true" tabindex="-1"></a>    .[, det_Y <span class="sc">:</span><span class="er">=</span> theta_x <span class="sc">*</span> T <span class="sc">+</span> g_x] <span class="sc">%&gt;%</span></span>
<span id="cb86-1252"><a href="#cb86-1252" aria-hidden="true" tabindex="-1"></a>    .[, Y <span class="sc">:</span><span class="er">=</span> det_Y <span class="sc">+</span> mu] <span class="sc">%&gt;%</span> </span>
<span id="cb86-1253"><a href="#cb86-1253" aria-hidden="true" tabindex="-1"></a>    .[, id <span class="sc">:</span><span class="er">=</span> <span class="dv">1</span><span class="sc">:</span>.N]</span>
<span id="cb86-1254"><a href="#cb86-1254" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1255"><a href="#cb86-1255" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(data)</span>
<span id="cb86-1256"><a href="#cb86-1256" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb86-1257"><a href="#cb86-1257" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-1258"><a href="#cb86-1258" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1259"><a href="#cb86-1259" aria-hidden="true" tabindex="-1"></a>Here is a dataset that is generated according to the DGP.</span>
<span id="cb86-1260"><a href="#cb86-1260" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1261"><a href="#cb86-1261" aria-hidden="true" tabindex="-1"></a>::: {.column-margin}</span>
<span id="cb86-1262"><a href="#cb86-1262" aria-hidden="true" tabindex="-1"></a>Unfold the Code chunk above to see how <span class="in">`gen_data()`</span> is defined.</span>
<span id="cb86-1263"><a href="#cb86-1263" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb86-1264"><a href="#cb86-1264" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1267"><a href="#cb86-1267" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-1268"><a href="#cb86-1268" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">428943</span>)</span>
<span id="cb86-1269"><a href="#cb86-1269" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">gen_data</span>() </span>
<span id="cb86-1270"><a href="#cb86-1270" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-1271"><a href="#cb86-1271" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1272"><a href="#cb86-1272" aria-hidden="true" tabindex="-1"></a>@fig-viz-cluster-data presents the values of individual outcomes ($Y$), error ($\mu$), and outcomes less errors ($\theta(X)\cdot T + g(X)$) and their respective means by cluster (<span class="in">`group`</span>). You can observe that data are clustered by <span class="in">`group`</span>.</span>
<span id="cb86-1273"><a href="#cb86-1273" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1276"><a href="#cb86-1276" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-1277"><a href="#cb86-1277" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb86-1278"><a href="#cb86-1278" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: Visualization of clustered data </span></span>
<span id="cb86-1279"><a href="#cb86-1279" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-viz-cluster-data</span></span>
<span id="cb86-1280"><a href="#cb86-1280" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1281"><a href="#cb86-1281" aria-hidden="true" tabindex="-1"></a>plot_data <span class="ot">&lt;-</span></span>
<span id="cb86-1282"><a href="#cb86-1282" aria-hidden="true" tabindex="-1"></a>  data[, .(id, Y, det_Y, mu, group)] <span class="sc">%&gt;%</span> </span>
<span id="cb86-1283"><a href="#cb86-1283" aria-hidden="true" tabindex="-1"></a>  <span class="fu">melt</span>(<span class="at">id.var =</span> <span class="fu">c</span>(<span class="st">"id"</span>, <span class="st">"group"</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb86-1284"><a href="#cb86-1284" aria-hidden="true" tabindex="-1"></a>  .[, type <span class="sc">:</span><span class="er">=</span> <span class="fu">fcase</span>(</span>
<span id="cb86-1285"><a href="#cb86-1285" aria-hidden="true" tabindex="-1"></a>    variable <span class="sc">==</span> <span class="st">"Y"</span>, <span class="st">"Y"</span>,</span>
<span id="cb86-1286"><a href="#cb86-1286" aria-hidden="true" tabindex="-1"></a>    variable <span class="sc">==</span> <span class="st">"det_Y"</span>, <span class="st">"Y - Error"</span>,</span>
<span id="cb86-1287"><a href="#cb86-1287" aria-hidden="true" tabindex="-1"></a>    variable <span class="sc">==</span> <span class="st">"mu"</span>, <span class="st">"Error"</span></span>
<span id="cb86-1288"><a href="#cb86-1288" aria-hidden="true" tabindex="-1"></a>  )] <span class="sc">%&gt;%</span> </span>
<span id="cb86-1289"><a href="#cb86-1289" aria-hidden="true" tabindex="-1"></a>  .[, type <span class="sc">:</span><span class="er">=</span> <span class="fu">factor</span>(type, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Y"</span>, <span class="st">"Error"</span>, <span class="st">"Y - Error"</span>))]</span>
<span id="cb86-1290"><a href="#cb86-1290" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1291"><a href="#cb86-1291" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> plot_data) <span class="sc">+</span></span>
<span id="cb86-1292"><a href="#cb86-1292" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">y =</span> value, <span class="at">x =</span> <span class="fu">factor</span>(group)), <span class="at">size =</span> <span class="fl">0.3</span>) <span class="sc">+</span></span>
<span id="cb86-1293"><a href="#cb86-1293" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(</span>
<span id="cb86-1294"><a href="#cb86-1294" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> plot_data[, .(<span class="at">value =</span> <span class="fu">mean</span>(value)), <span class="at">by =</span> .(group, type)],</span>
<span id="cb86-1295"><a href="#cb86-1295" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">y =</span> value, <span class="at">x =</span> <span class="fu">factor</span>(group)),</span>
<span id="cb86-1296"><a href="#cb86-1296" aria-hidden="true" tabindex="-1"></a>    <span class="at">color =</span> <span class="st">"red"</span></span>
<span id="cb86-1297"><a href="#cb86-1297" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb86-1298"><a href="#cb86-1298" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"group"</span>) <span class="sc">+</span></span>
<span id="cb86-1299"><a href="#cb86-1299" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(type <span class="sc">~</span> .) <span class="sc">+</span></span>
<span id="cb86-1300"><a href="#cb86-1300" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb86-1301"><a href="#cb86-1301" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">6</span>, <span class="at">angle =</span> <span class="dv">90</span>)</span>
<span id="cb86-1302"><a href="#cb86-1302" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb86-1303"><a href="#cb86-1303" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() </span>
<span id="cb86-1304"><a href="#cb86-1304" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1305"><a href="#cb86-1305" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-1306"><a href="#cb86-1306" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1307"><a href="#cb86-1307" aria-hidden="true" tabindex="-1"></a>We now run MC simulations. We use the same test data for all the iterations so that we can find the standard error estimates averaged over iterations. The following function trains CFs with and without clustering by <span class="in">`group`</span>, estimate $\theta(X)$ at the test data along with the estimate of the variance of $\hat{\theta}(X)$. </span>
<span id="cb86-1308"><a href="#cb86-1308" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1311"><a href="#cb86-1311" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-1312"><a href="#cb86-1312" aria-hidden="true" tabindex="-1"></a><span class="co">#=== use fixed test data ===#</span></span>
<span id="cb86-1313"><a href="#cb86-1313" aria-hidden="true" tabindex="-1"></a>test_data <span class="ot">&lt;-</span> <span class="fu">gen_data</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb86-1314"><a href="#cb86-1314" aria-hidden="true" tabindex="-1"></a>  .[, test_id <span class="sc">:</span><span class="er">=</span> <span class="dv">1</span><span class="sc">:</span>.N]</span>
<span id="cb86-1315"><a href="#cb86-1315" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1316"><a href="#cb86-1316" aria-hidden="true" tabindex="-1"></a>run_MC_sim <span class="ot">&lt;-</span> <span class="cf">function</span>(i){</span>
<span id="cb86-1317"><a href="#cb86-1317" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1318"><a href="#cb86-1318" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(i)</span>
<span id="cb86-1319"><a href="#cb86-1319" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1320"><a href="#cb86-1320" aria-hidden="true" tabindex="-1"></a>  train_data <span class="ot">&lt;-</span> <span class="fu">gen_data</span>()</span>
<span id="cb86-1321"><a href="#cb86-1321" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1322"><a href="#cb86-1322" aria-hidden="true" tabindex="-1"></a>  cf_trained_no_cluster <span class="ot">&lt;-</span></span>
<span id="cb86-1323"><a href="#cb86-1323" aria-hidden="true" tabindex="-1"></a>    <span class="fu">causal_forest</span>(</span>
<span id="cb86-1324"><a href="#cb86-1324" aria-hidden="true" tabindex="-1"></a>      <span class="at">X =</span> <span class="fu">select</span>(train_data, <span class="fu">starts_with</span>(<span class="st">"x"</span>)),</span>
<span id="cb86-1325"><a href="#cb86-1325" aria-hidden="true" tabindex="-1"></a>      <span class="at">Y =</span> train_data[, Y],</span>
<span id="cb86-1326"><a href="#cb86-1326" aria-hidden="true" tabindex="-1"></a>      <span class="at">W =</span> train_data[, T]</span>
<span id="cb86-1327"><a href="#cb86-1327" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb86-1328"><a href="#cb86-1328" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1329"><a href="#cb86-1329" aria-hidden="true" tabindex="-1"></a>  thta_hat_no_cl <span class="ot">&lt;-</span> </span>
<span id="cb86-1330"><a href="#cb86-1330" aria-hidden="true" tabindex="-1"></a>    <span class="fu">predict</span>(</span>
<span id="cb86-1331"><a href="#cb86-1331" aria-hidden="true" tabindex="-1"></a>      cf_trained_no_cluster, </span>
<span id="cb86-1332"><a href="#cb86-1332" aria-hidden="true" tabindex="-1"></a>      <span class="fu">select</span>(test_data, <span class="fu">starts_with</span>(<span class="st">"x"</span>)),</span>
<span id="cb86-1333"><a href="#cb86-1333" aria-hidden="true" tabindex="-1"></a>      <span class="at">estimate.variance =</span> <span class="cn">TRUE</span></span>
<span id="cb86-1334"><a href="#cb86-1334" aria-hidden="true" tabindex="-1"></a>    )  <span class="sc">%&gt;%</span> </span>
<span id="cb86-1335"><a href="#cb86-1335" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.table</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb86-1336"><a href="#cb86-1336" aria-hidden="true" tabindex="-1"></a>    .[, theta_true <span class="sc">:</span><span class="er">=</span> test_data<span class="sc">$</span>theta_x] <span class="sc">%&gt;%</span> </span>
<span id="cb86-1337"><a href="#cb86-1337" aria-hidden="true" tabindex="-1"></a>    .[, id <span class="sc">:</span><span class="er">=</span> test_data<span class="sc">$</span>id] <span class="sc">%&gt;%</span> </span>
<span id="cb86-1338"><a href="#cb86-1338" aria-hidden="true" tabindex="-1"></a>    .[, type <span class="sc">:</span><span class="er">=</span> <span class="st">"No clustering"</span>]</span>
<span id="cb86-1339"><a href="#cb86-1339" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1340"><a href="#cb86-1340" aria-hidden="true" tabindex="-1"></a>  cf_trained_with_cluster <span class="ot">&lt;-</span></span>
<span id="cb86-1341"><a href="#cb86-1341" aria-hidden="true" tabindex="-1"></a>    <span class="fu">causal_forest</span>(</span>
<span id="cb86-1342"><a href="#cb86-1342" aria-hidden="true" tabindex="-1"></a>      <span class="at">X =</span> <span class="fu">select</span>(train_data, <span class="fu">starts_with</span>(<span class="st">"x"</span>)),</span>
<span id="cb86-1343"><a href="#cb86-1343" aria-hidden="true" tabindex="-1"></a>      <span class="at">Y =</span> train_data[, Y],</span>
<span id="cb86-1344"><a href="#cb86-1344" aria-hidden="true" tabindex="-1"></a>      <span class="at">W =</span> train_data[, T],</span>
<span id="cb86-1345"><a href="#cb86-1345" aria-hidden="true" tabindex="-1"></a>      <span class="at">cluster =</span> train_data[, group]</span>
<span id="cb86-1346"><a href="#cb86-1346" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb86-1347"><a href="#cb86-1347" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1348"><a href="#cb86-1348" aria-hidden="true" tabindex="-1"></a>  thta_hat_with_cl <span class="ot">&lt;-</span> </span>
<span id="cb86-1349"><a href="#cb86-1349" aria-hidden="true" tabindex="-1"></a>    <span class="fu">predict</span>(</span>
<span id="cb86-1350"><a href="#cb86-1350" aria-hidden="true" tabindex="-1"></a>      cf_trained_with_cluster, </span>
<span id="cb86-1351"><a href="#cb86-1351" aria-hidden="true" tabindex="-1"></a>      <span class="fu">select</span>(test_data, <span class="fu">starts_with</span>(<span class="st">"x"</span>)),</span>
<span id="cb86-1352"><a href="#cb86-1352" aria-hidden="true" tabindex="-1"></a>      <span class="at">estimate.variance =</span> <span class="cn">TRUE</span></span>
<span id="cb86-1353"><a href="#cb86-1353" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">%&gt;%</span> </span>
<span id="cb86-1354"><a href="#cb86-1354" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.table</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb86-1355"><a href="#cb86-1355" aria-hidden="true" tabindex="-1"></a>    .[, theta_true <span class="sc">:</span><span class="er">=</span> test_data<span class="sc">$</span>theta_x] <span class="sc">%&gt;%</span> </span>
<span id="cb86-1356"><a href="#cb86-1356" aria-hidden="true" tabindex="-1"></a>    .[, id <span class="sc">:</span><span class="er">=</span> test_data<span class="sc">$</span>id] <span class="sc">%&gt;%</span> </span>
<span id="cb86-1357"><a href="#cb86-1357" aria-hidden="true" tabindex="-1"></a>    .[, type <span class="sc">:</span><span class="er">=</span> <span class="st">"With clustering"</span>]</span>
<span id="cb86-1358"><a href="#cb86-1358" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1359"><a href="#cb86-1359" aria-hidden="true" tabindex="-1"></a>  return_data <span class="ot">&lt;-</span> </span>
<span id="cb86-1360"><a href="#cb86-1360" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rbind</span>(thta_hat_no_cl, thta_hat_with_cl) <span class="sc">%&gt;%</span> </span>
<span id="cb86-1361"><a href="#cb86-1361" aria-hidden="true" tabindex="-1"></a>    .[, sim <span class="sc">:</span><span class="er">=</span> i]</span>
<span id="cb86-1362"><a href="#cb86-1362" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1363"><a href="#cb86-1363" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(return_data)</span>
<span id="cb86-1364"><a href="#cb86-1364" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb86-1365"><a href="#cb86-1365" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1366"><a href="#cb86-1366" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-1367"><a href="#cb86-1367" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1368"><a href="#cb86-1368" aria-hidden="true" tabindex="-1"></a>Let's repeat the experiment 200 times.</span>
<span id="cb86-1369"><a href="#cb86-1369" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1372"><a href="#cb86-1372" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-1373"><a href="#cb86-1373" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false </span></span>
<span id="cb86-1374"><a href="#cb86-1374" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">428943</span>)</span>
<span id="cb86-1375"><a href="#cb86-1375" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1376"><a href="#cb86-1376" aria-hidden="true" tabindex="-1"></a>mc_results <span class="ot">&lt;-</span></span>
<span id="cb86-1377"><a href="#cb86-1377" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(</span>
<span id="cb86-1378"><a href="#cb86-1378" aria-hidden="true" tabindex="-1"></a>    <span class="fu">seq_len</span>(<span class="dv">200</span>),</span>
<span id="cb86-1379"><a href="#cb86-1379" aria-hidden="true" tabindex="-1"></a>    run_MC_sim</span>
<span id="cb86-1380"><a href="#cb86-1380" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb86-1381"><a href="#cb86-1381" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-1382"><a href="#cb86-1382" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1385"><a href="#cb86-1385" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-1386"><a href="#cb86-1386" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb86-1387"><a href="#cb86-1387" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo:  false</span></span>
<span id="cb86-1388"><a href="#cb86-1388" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">428943</span>)</span>
<span id="cb86-1389"><a href="#cb86-1389" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1390"><a href="#cb86-1390" aria-hidden="true" tabindex="-1"></a>mc_results <span class="ot">&lt;-</span></span>
<span id="cb86-1391"><a href="#cb86-1391" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mclapply</span>(</span>
<span id="cb86-1392"><a href="#cb86-1392" aria-hidden="true" tabindex="-1"></a>    <span class="fu">seq_len</span>(<span class="dv">200</span>),</span>
<span id="cb86-1393"><a href="#cb86-1393" aria-hidden="true" tabindex="-1"></a>    run_MC_sim,</span>
<span id="cb86-1394"><a href="#cb86-1394" aria-hidden="true" tabindex="-1"></a>    <span class="at">mc.cores =</span> <span class="fu">detectCores</span>() <span class="sc">/</span> <span class="dv">4</span> <span class="sc">*</span> <span class="dv">3</span></span>
<span id="cb86-1395"><a href="#cb86-1395" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span> </span>
<span id="cb86-1396"><a href="#cb86-1396" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rbindlist</span>()</span>
<span id="cb86-1397"><a href="#cb86-1397" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1398"><a href="#cb86-1398" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(mc_results, here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"LectureNotes/Data/mc_cf_cluster.rds"</span>))</span>
<span id="cb86-1399"><a href="#cb86-1399" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-1400"><a href="#cb86-1400" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1403"><a href="#cb86-1403" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-1404"><a href="#cb86-1404" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb86-1405"><a href="#cb86-1405" aria-hidden="true" tabindex="-1"></a>mc_results <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"Data/mc_cf_cluster.rds"</span>)</span>
<span id="cb86-1406"><a href="#cb86-1406" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-1407"><a href="#cb86-1407" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1408"><a href="#cb86-1408" aria-hidden="true" tabindex="-1"></a>@fig-cate-group-mse shows the distribution of MSE of CATE estimation by CF with and without clustering by <span class="in">`group`</span>. As you can see, clustering is making things worse in this instance.</span>
<span id="cb86-1409"><a href="#cb86-1409" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1412"><a href="#cb86-1412" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-1413"><a href="#cb86-1413" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb86-1414"><a href="#cb86-1414" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: MSE of CATE estimation by CF with and without clustering by group</span></span>
<span id="cb86-1415"><a href="#cb86-1415" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-cate-group-mse</span></span>
<span id="cb86-1416"><a href="#cb86-1416" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1417"><a href="#cb86-1417" aria-hidden="true" tabindex="-1"></a>mc_results <span class="sc">%&gt;%</span> </span>
<span id="cb86-1418"><a href="#cb86-1418" aria-hidden="true" tabindex="-1"></a>  .[, .(<span class="at">mse =</span> <span class="fu">mean</span>((theta_true <span class="sc">-</span> predictions)<span class="sc">^</span><span class="dv">2</span>)), by <span class="ot">=</span> .(sim, type)] <span class="sc">%&gt;%</span> </span>
<span id="cb86-1419"><a href="#cb86-1419" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="at">data =</span> .) <span class="sc">+</span> </span>
<span id="cb86-1420"><a href="#cb86-1420" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_density</span>(<span class="fu">aes</span>(<span class="at">x =</span> mse, <span class="at">fill =</span> type), <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb86-1421"><a href="#cb86-1421" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>()</span>
<span id="cb86-1422"><a href="#cb86-1422" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-1423"><a href="#cb86-1423" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1424"><a href="#cb86-1424" aria-hidden="true" tabindex="-1"></a>@fig-sd-test shows the distribution of the ratio of the standard deviation of $\hat{\theta}(X)$ to the mean of the estimate of the standard deviation of $\hat{\theta}(X)$. Ratio less than 1 indicates that the estimate of the standard deviation of $\hat{\theta}(X)$ is under-estimating the true standard deviation of $\hat{\theta}(X)$.</span>
<span id="cb86-1425"><a href="#cb86-1425" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1428"><a href="#cb86-1428" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb86-1429"><a href="#cb86-1429" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb86-1430"><a href="#cb86-1430" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: Ratio of the estimated standard error of CATE to the standard deviation of CATE</span></span>
<span id="cb86-1431"><a href="#cb86-1431" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-sd-test</span></span>
<span id="cb86-1432"><a href="#cb86-1432" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1433"><a href="#cb86-1433" aria-hidden="true" tabindex="-1"></a>mc_results <span class="sc">%&gt;%</span> </span>
<span id="cb86-1434"><a href="#cb86-1434" aria-hidden="true" tabindex="-1"></a>  .[, .(<span class="at">sd =</span> <span class="fu">sd</span>(predictions), <span class="at">sd_est =</span> <span class="fu">mean</span>(<span class="fu">sqrt</span>(variance.estimates))), by <span class="ot">=</span> .(id, type)] <span class="sc">%&gt;%</span> </span>
<span id="cb86-1435"><a href="#cb86-1435" aria-hidden="true" tabindex="-1"></a>  .[, ratio <span class="sc">:</span><span class="er">=</span> sd_est<span class="sc">/</span>sd] <span class="sc">%&gt;%</span> </span>
<span id="cb86-1436"><a href="#cb86-1436" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="at">data =</span> .) <span class="sc">+</span> </span>
<span id="cb86-1437"><a href="#cb86-1437" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_density</span>(<span class="fu">aes</span>(<span class="at">x =</span> ratio, <span class="at">fill =</span> type), <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb86-1438"><a href="#cb86-1438" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>()</span>
<span id="cb86-1439"><a href="#cb86-1439" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb86-1440"><a href="#cb86-1440" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1441"><a href="#cb86-1441" aria-hidden="true" tabindex="-1"></a><span class="fu">## Hyper-parameter tuning</span></span>
<span id="cb86-1442"><a href="#cb86-1442" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1443"><a href="#cb86-1443" aria-hidden="true" tabindex="-1"></a>You can write your own codes to conduct cross-validation just like any other ML methods as we saw in @sec-cross-validation (cross-validation is a general approach that can be applied to any ML methods) to tune hyper-parameters. However, the <span class="in">`grf`</span> package offers internal tuning options based on out-of-bag error (see @sec-rf-implementation for what out-of-bag prediction is).</span>
<span id="cb86-1444"><a href="#cb86-1444" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1445"><a href="#cb86-1445" aria-hidden="true" tabindex="-1"></a>You can define which hyper-parameters to tune using <span class="in">`tune.parameters`</span> option. For example, <span class="in">`tune.parameters = c("min.node.size", "mtry")`</span> would tune <span class="in">`min.node.size`</span> and <span class="in">`mtry`</span>.</span>
<span id="cb86-1446"><a href="#cb86-1446" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1447"><a href="#cb86-1447" aria-hidden="true" tabindex="-1"></a>The cross-validation procedure takes the following steps:</span>
<span id="cb86-1448"><a href="#cb86-1448" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1449"><a href="#cb86-1449" aria-hidden="true" tabindex="-1"></a>:::{.callout-note}</span>
<span id="cb86-1450"><a href="#cb86-1450" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1451"><a href="#cb86-1451" aria-hidden="true" tabindex="-1"></a><span class="fu">## Cross-validation steps</span></span>
<span id="cb86-1452"><a href="#cb86-1452" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1453"><a href="#cb86-1453" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>Draw <span class="in">`tune.num.reps`</span> numbers of random points in the space of possible values of parameters specified by <span class="in">`tune.parameters`</span></span>
<span id="cb86-1454"><a href="#cb86-1454" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>For each of the <span class="in">`tune.num.reps`</span> sets of parameter values, <span class="in">`tune.num.trees`</span> of trees are built and out-of-bag errors are calculated</span>
<span id="cb86-1455"><a href="#cb86-1455" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>Regress the error estimates on parameter values using smoothing functions </span>
<span id="cb86-1456"><a href="#cb86-1456" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>Redraw <span class="in">`tune.num.draws`</span> numbers of random points in the space of possible values of parameters</span>
<span id="cb86-1457"><a href="#cb86-1457" aria-hidden="true" tabindex="-1"></a><span class="ss">+ </span>Pick the set of parameter values that minimizes out-of-bag error. </span>
<span id="cb86-1458"><a href="#cb86-1458" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1459"><a href="#cb86-1459" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb86-1460"><a href="#cb86-1460" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1461"><a href="#cb86-1461" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1462"><a href="#cb86-1462" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1463"><a href="#cb86-1463" aria-hidden="true" tabindex="-1"></a><span class="fu">## References {.unnumbered}</span></span>
<span id="cb86-1464"><a href="#cb86-1464" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1465"><a href="#cb86-1465" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;div</span> <span class="er">id</span><span class="ot">=</span><span class="st">"refs"</span><span class="kw">&gt;&lt;/div&gt;</span></span>
<span id="cb86-1466"><a href="#cb86-1466" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1467"><a href="#cb86-1467" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1468"><a href="#cb86-1468" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-1469"><a href="#cb86-1469" aria-hidden="true" tabindex="-1"></a></span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->



</body></html>